<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>csb</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th bgcolor="#70b0f0" class="navbar-select"
          >&nbsp;&nbsp;&nbsp;Home&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >CSB v1.2.3.617</th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        Package&nbsp;csb
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="csb-module.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== PACKAGE DESCRIPTION ==================== -->
<h1 class="epydoc">Package csb</h1><p class="nomargin-top"><span class="codelink"><a href="csb-pysrc.html">source&nbsp;code</a></span></p>
<p>CSB is a high-level, object-oriented library used to solve problems in
  the field of Computational Structural Biology.</p>
  <h1 class="heading">Introduction</h1>
    <p>The library is composed of a set of highly branched python packages 
    (namespaces). Some of the packages are meant to be directly used by the
    clients (core library), while others are utility modules and take part 
    in the development of the library:</p>
    <ol start="1">
      <li>
        Core class library -- object-oriented, granular, with an emphasis 
        on design and clean interfaces. A Sequence is not a string, and a 
        Structure is not a dict or list. Naming conventions matter.
      </li>
      <li>
        Application framework -- executable console applications 
        (&quot;protocols&quot;), which consume objects from the core 
        library. The framework ensures that each CSB application is also 
        reusable and can be instantiated as a regular python object without
        any ugly side effects (sys.exit() and friends). See <a 
        href="csb.apps-module.html" class="link">csb.apps</a> for more 
        details.
      </li>
      <li>
        Test framework -- built on top of the standard unittest as a thin 
        wrapping layer. Provides some sugar like transparent management of 
        test data files, and modular test execution. <a 
        href="csb.test-module.html" class="link">csb.test</a> will give you
        all the details.
      </li>
    </ol>
    <p>The core library is roughly composed of:</p>
    <ul>
      <li>
        bioinformatics API: <a href="csb.bio-module.html" 
        class="link">csb.bio</a>, which includes stuff like <a 
        href="csb.bio.io-module.html" class="link">csb.bio.io</a>, <a 
        href="csb.bio.structure-module.html" 
        class="link">csb.bio.structure</a>, <a 
        href="csb.bio.sequence-module.html" 
        class="link">csb.bio.sequence</a>, <a 
        href="csb.bio.hmm-module.html" class="link">csb.bio.hmm</a>
      </li>
      <li>
        statistics API: <a href="csb.statistics-module.html" 
        class="link">csb.statistics</a>, <a href="csb.numeric-module.html" 
        class="link">csb.numeric</a>
      </li>
      <li>
        utilities - <a href="csb.io-module.html" class="link">csb.io</a>, 
        <a href="csb.core-module.html" class="link">csb.core</a>
      </li>
    </ul>
  <h1 class="heading">Getting started</h1>
    <p>Perhaps one of the most frequently used parts of the library is the 
    <a href="csb.bio.structure-module.html" 
    class="link">csb.bio.structure</a> module, which provides the <a 
    href="csb.bio.structure.Structure-class.html" 
    class="link">Structure</a>, <a 
    href="csb.bio.structure.Chain-class.html" class="link">Chain</a>, <a 
    href="csb.bio.structure.Residue-class.html" class="link">Residue</a> 
    and <a href="csb.bio.structure.Atom-class.html" class="link">Atom</a> 
    objects. You could easily build a <a 
    href="csb.bio.structure.Structure-class.html" 
    class="link">Structure</a> from scratch, but a far more common scenario
    is parsing a structure from a PDB file using one of the <a 
    href="csb.bio.io.wwpdb.AbstractStructureParser-class.html" 
    class="link">AbstractStructureParser</a>s. All bio IO objects, 
    including the StructureParser factory, are defined in <a 
    href="csb.bio.io-module.html" class="link">csb.bio.io</a> and 
    sub-packages:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span><span class="py-keyword">from</span> csb.bio.io.wwpdb <span class="py-keyword">import</span> StructureParser
<span class="py-prompt">&gt;&gt;&gt; </span>p = StructureParser(<span class="py-string">&quot;/some/file/pdb1x80.ent&quot;</span>)
<span class="py-prompt">&gt;&gt;&gt; </span>s = p.parse_structure()
<span class="py-prompt">&gt;&gt;&gt; </span><span class="py-keyword">print</span>(s)
<span class="py-output">&lt;Structure: 1x80, 2 chains&gt;</span></pre>
    <p>The last statement will return a <a 
    href="csb.bio.structure.Structure-class.html" 
    class="link">csb.bio.structure.Structure</a> instance, which is a 
    composite hierarchical object:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span><span class="py-keyword">for</span> chain_id <span class="py-keyword">in</span> s.chains:
<span class="py-output">        chain = s.chains[chain_id]</span>
<span class="py-output">        for residue in chain.residues:</span>
<span class="py-output">            for atom_id in residue.atoms:</span>
<span class="py-output">                atom = residue.atoms[atom_id]</span>
<span class="py-output">                print(atom.vector)</span></pre>
    <p>Some of the inner objects in this hierarchy behave just like 
    dictionaries (but are not):</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span>s.chains[<span class="py-string">'A'</span>]        <span class="py-comment"># access chain A by ID</span>
<span class="py-output">&lt;Chain A: Protein&gt;</span>
<span class="py-output"></span><span class="py-prompt">&gt;&gt;&gt; </span>s[<span class="py-string">'A'</span>]               <span class="py-comment"># the same</span>
<span class="py-output">&lt;Chain A: Protein&gt;</span></pre>
    <p>Others behave like collections:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span>chain.residues[10]               <span class="py-comment"># 1-based access to the residues in the chain</span>
<span class="py-output">&lt;ProteinResidue [10]: PRO 10&gt;</span>
<span class="py-output"></span><span class="py-prompt">&gt;&gt;&gt; </span>chain[10]                        <span class="py-comment"># 0-based, list-like access</span>
<span class="py-output">&lt;ProteinResidue [11]: GLY 11&gt;</span></pre>
    <p>But all entities are iterable because they inherit the 
    <code>items</code> iterator from <a 
    href="csb.bio.structure.AbstractEntity-class.html" 
    class="link">AbstractEntity</a>. The above loop can be shortened:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span><span class="py-keyword">for</span> chain <span class="py-keyword">in</span> s.items:
<span class="py-output">        for residue in chain.items:</span>
<span class="py-output">            for atom in residue.items:</span>
<span class="py-output">                print(atom.vector)</span></pre>
    <p>or even more:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span><span class="py-keyword">from</span> csb.bio.structure <span class="py-keyword">import</span> Atom
<span class="py-prompt">&gt;&gt;&gt; </span><span class="py-keyword">for</span> atom <span class="py-keyword">in</span> s.components(klass=Atom):
<span class="py-output">        print(atom.vector)</span></pre>
    <p>You may also be interested in extracting a sub-chain from this 
    structure:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span>s.chains[<span class="py-string">'B'</span>].subregion(3, 20)    <span class="py-comment"># from positions 3 to 20, inclusive</span>
<span class="py-output">&lt;Chain B: Protein&gt;</span></pre>
    <p>or modifying it in some way, for example, in order to append a new 
    residue, try:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span><span class="py-keyword">from</span> csb.bio.structure <span class="py-keyword">import</span> ProteinResidue
<span class="py-prompt">&gt;&gt;&gt; </span><span class="py-keyword">from</span> csb.bio.sequence <span class="py-keyword">import</span> ProteinAlphabet
<span class="py-prompt">&gt;&gt;&gt; </span>residue = ProteinResidue(401, ProteinAlphabet.ALA)
<span class="py-prompt">&gt;&gt;&gt; </span>s.chains[<span class="py-string">'A'</span>].residues.append(residue)</pre>
    <p>Finally, you would probably want to save your structure back to a 
    PDB file:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span>s.to_pdb(<span class="py-string">'/some/file/name.pdb'</span>)</pre>
  <h1 class="heading">Where to go from here</h1>
    <p>If you want to dive into statistics, you could peek inside <a 
    href="csb.statistics-module.html" class="link">csb.statistics</a> and 
    its sub-packages. For example, <a href="csb.statistics.pdf-module.html"
    class="link">csb.statistics.pdf</a> contains a collection of <a 
    href="csb.statistics.pdf.AbstractDensity-class.html" 
    class="link">probability density objects</a>, like <a 
    href="csb.statistics.pdf.Normal-class.html" class="link">Gaussian</a> 
    or <a href="csb.statistics.pdf.Gamma-class.html" 
    class="link">Gamma</a>.</p>
    <p>But chances are you would first like to try reading some files, so 
    you could start exploring <a href="csb.bio.io-module.html" 
    class="link">csb.bio.io</a> right now. As we have already seen, <a 
    href="csb.bio.io.wwpdb-module.html" class="link" 
    onclick="show_private();">csb.bio.io.wwpdb</a> provides PDB <a 
    href="csb.bio.structure.Structure-class.html" 
    class="link">Structure</a> parsers, for example <a 
    href="csb.bio.io.wwpdb.RegularStructureParser-class.html" 
    class="link">csb.bio.io.wwpdb.RegularStructureParser</a> and <a 
    href="csb.bio.io.wwpdb.LegacyStructureParser-class.html" 
    class="link">csb.bio.io.wwpdb.LegacyStructureParser</a>.</p>
    <p><a href="csb.bio.io.fasta-module.html" class="link" 
    onclick="show_private();">csb.bio.io.fasta</a> is all about reading 
    FASTA <a href="csb.bio.sequence.AbstractSequence-class.html" 
    class="link">Sequence</a>s and <a 
    href="csb.bio.sequence.AbstractAlignment-class.html" 
    class="link">SequenceAlignment</a>s. Be sure to check out <a 
    href="csb.bio.io.fasta.SequenceParser-class.html" 
    class="link">csb.bio.io.fasta.SequenceParser</a>, <a 
    href="csb.bio.io.fasta.SequenceAlignmentReader-class.html" 
    class="link">csb.bio.io.fasta.SequenceAlignmentReader</a> and <a 
    href="csb.bio.io.fasta.StructureAlignmentFactory-class.html" 
    class="link">csb.bio.io.fasta.StructureAlignmentFactory</a>.</p>
    <p>If you are working with HHpred (<a 
    href="csb.bio.hmm.ProfileHMM-class.html" class="link">ProfileHMM</a>s, 
    <a href="csb.bio.hmm.HHpredHit-class.html" 
    class="link">HHpredHit</a>s), then <a 
    href="csb.bio.io.hhpred-module.html" class="link" 
    onclick="show_private();">csb.bio.io.hhpred</a> is for you. This 
    package provides <a href="csb.bio.io.hhpred.HHProfileParser-class.html"
    class="link">csb.bio.io.hhpred.HHProfileParser</a> and <a 
    href="csb.bio.io.hhpred.HHOutputParser-class.html" 
    class="link">csb.bio.io.hhpred.HHOutputParser</a>, which are used to 
    read *.hhm and *.hhr files.</p>
    <p>Finally, if you want to make some nice plots with matplotlib, you 
    may like the clean object-oriented interface of our <a 
    href="csb.io.plots.Chart-class.html" class="link">Chart</a>. See <a 
    href="csb.io.plots-module.html" class="link">csb.io.plots</a> and maybe
    also <a href="csb.io.tsv-module.html" class="link">csb.io.tsv</a> to 
    get started.</p>
  <h1 class="heading">Development</h1>
    <p>When contributing code to CSB, please take into account the 
    following:</p>
    <ol start="1">
      <li>
        New features or bug fixes should always be accompanied by test 
        cases. Also, always run the complete test suite before committing. 
        For more details on this topic, see <a href="csb.test-module.html" 
        class="link">csb.test</a>.
      </li>
      <li>
        The source code of CSB must be cross-platform and cross-interpreter
        compatible. <a href="csb.core-module.html" 
        class="link">csb.core</a> and <a href="csb.io-module.html" 
        class="link">csb.io</a> will give you all necessary details on how 
        to use the CSB compatibility layer.
      </li>
    </ol>
  <h1 class="heading">License</h1>
    <p>CSB is open source and distributed under OSI-approved MIT 
    license:</p>
<pre class="literalblock">
   Copyright (c) 2012 Michael Habeck
   
   Permission is hereby granted, free of charge, to any person obtaining
   a copy of this software and associated documentation files (the
   &quot;Software&quot;), to deal in the Software without restriction, including
   without limitation the rights to use, copy, modify, merge, publish,
   distribute, sublicense, and/or sell copies of the Software, and to
   permit persons to whom the Software is furnished to do so, subject to
   the following conditions:
   
   The above copyright notice and this permission notice shall be
   included in all copies or substantial portions of the Software.
   
   THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND,
   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
   IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
   CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
   TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
   SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
</pre>

<hr />
<div class="fields">      <p><strong>Version:</strong>
        1.2.3.617
      </p>
</div><!-- ==================== SUBMODULES ==================== -->
<a name="section-Submodules"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td align="left" colspan="2" class="table-header">
    <span class="table-header">Submodules</span></td>
</tr>
  <tr><td class="summary">
  <ul class="nomargin">
    <li> <strong class="uidlink"><a href="csb.apps-module.html">csb.apps</a></strong>: <em class="summary">Root package for all executable CSB client programs.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.bfit-module.html">csb.apps.bfit</a></strong>: <em class="summary">Python application for robust structure superposition of two 
        structures.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.bfite-module.html">csb.apps.bfite</a></strong>: <em class="summary">Python application for robust structure superposition of an 
        ensemble of structures.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm-module.html">csb.apps.buildhmm</a></strong>: <em class="summary">Build an HMM from a FASTA sequence.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag-module.html">csb.apps.csfrag</a></strong>: <em class="summary">CSfrag: build a dynamic library of analogous fragments, given a 
        list of assigned chemical shifts.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.embd-module.html">csb.apps.embd</a></strong>: <em class="summary">Sharpening of EM maps by non-negative blind deconvolution.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.helloworld-module.html">csb.apps.helloworld</a></strong>: <em class="summary">This is a CSB HelloWorld dummy application.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag-module.html">csb.apps.hhfrag</a></strong>: <em class="summary">HHfrag: build a dynamic variable-length fragment library for 
        protein structure prediction with Rosetta AbInitio.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhsearch-module.html">csb.apps.hhsearch</a></strong>: <em class="summary">Python bindings for the HHsearch program.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.precision-module.html">csb.apps.precision</a></strong>: <em class="summary">Measure the precision and coverage of a fragment library stored in 
        Rosetta NNmake format.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.promix-module.html">csb.apps.promix</a></strong>: <em class="summary">ProMix: Take a protein structure ensemble and find a mixture of 
        rigid segments or a mixture of conformers.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio-module.html">csb.bio</a></strong>: <em class="summary">Core bioinformatics abstractions and I/O.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments-module.html">csb.bio.fragments</a></strong>: <em class="summary">APIs for working with protein structure fragments and libraries.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.rosetta-module.html">csb.bio.fragments.rosetta</a></strong>: <em class="summary">Rosetta fragment libraries.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm-module.html">csb.bio.hmm</a></strong>: <em class="summary">HHpred and Hidden Markov Model APIs.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.pseudocounts-module.html">csb.bio.hmm.pseudocounts</a></strong>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io-module.html">csb.bio.io</a></strong>: <em class="summary">I/O for various biological file formats.</em>
    <ul>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.clans-module.html" onclick="show_private();">csb.bio.io.clans</a></strong>: <em class="summary">Classes for parsing/manipulating/writing CLANS (by Tancred Frickey)
        files</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.cs-module.html" onclick="show_private();">csb.bio.io.cs</a></strong>: <em class="summary">Simple NMR STAR chemical shift readers.</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.dssp-module.html" onclick="show_private();">csb.bio.io.dssp</a></strong>: <em class="summary">DSSP Parser</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.fasta-module.html" onclick="show_private();">csb.bio.io.fasta</a></strong>: <em class="summary">FASTA format sequence I/O.</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.hhpred-module.html" onclick="show_private();">csb.bio.io.hhpred</a></strong>: <em class="summary">HHpred-related format parsers.</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.mrc-module.html" onclick="show_private();">csb.bio.io.mrc</a></strong>: <em class="summary">Cryo-EM density map I/O</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.noe-module.html">csb.bio.io.noe</a></strong>: <em class="summary">Simple XEASY and Sparky peak list parsers.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.procheck-module.html">csb.bio.io.procheck</a></strong>: <em class="summary">Procheck parser</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.psipred-module.html">csb.bio.io.psipred</a></strong>: <em class="summary">PSIPRED Parser</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.svg-module.html">csb.bio.io.svg</a></strong>: <em class="summary">Build SVG diagrams from various csb objects.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.vasco-module.html">csb.bio.io.vasco</a></strong>: <em class="summary">VASCO Chemical Shift format parser.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.whatif-module.html">csb.bio.io.whatif</a></strong>: <em class="summary">Simple WhatIf/WhatCheck Summary parser</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.wwpdb-module.html" onclick="show_private();">csb.bio.io.wwpdb</a></strong>: <em class="summary">PDB structure parsers, format builders and database providers.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr-module.html">csb.bio.nmr</a></strong>: <em class="summary">NMR related objects.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence-module.html">csb.bio.sequence</a></strong>: <em class="summary">Sequence and sequence alignment APIs.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.alignment-module.html">csb.bio.sequence.alignment</a></strong>: <em class="summary">Collection of sequence alignment algorithms.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure-module.html">csb.bio.structure</a></strong>: <em class="summary">3D and secondary structure APIs.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.utils-module.html">csb.bio.utils</a></strong>: <em class="summary">Computational utility functions.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.build-module.html">csb.build</a></strong>: <em class="summary">CSB build related tools and programs.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.core-module.html">csb.core</a></strong>: <em class="summary">Generic containers, data structures and language extensions.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.io-module.html">csb.io</a></strong>: <em class="summary">Generic I/O utility objects.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.io.plots-module.html">csb.io.plots</a></strong>: <em class="summary">Plotting facilities, based on Python's MPL library.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv-module.html">csb.io.tsv</a></strong>: <em class="summary">Read, query and update textual tables via flexible SQL interface.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.numeric-module.html">csb.numeric</a></strong>: <em class="summary">Low level numeric / math utility functions.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.numeric.integrators-module.html">csb.numeric.integrators</a></strong>: <em class="summary">provides various integration schemes and an abstract gradient 
        class.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics-module.html">csb.statistics</a></strong>: <em class="summary">Statistics root package.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.ars-module.html">csb.statistics.ars</a></strong>: <em class="summary">Adaptive Rejection Sampling (ARS)</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.maxent-module.html">csb.statistics.maxent</a></strong>: <em class="summary">A Maximum-Entropy model for backbone torsion angles.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.mixtures-module.html">csb.statistics.mixtures</a></strong>: <em class="summary">Mixture models for multi-dimensional data.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf-module.html">csb.statistics.pdf</a></strong>: <em class="summary">Probability density functions.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.parameterized-module.html">csb.statistics.pdf.parameterized</a></strong>: <em class="summary">Probability density functions with support for shared and computed 
        parameters.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.rand-module.html">csb.statistics.rand</a></strong>: <em class="summary">Random number generators</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers-module.html">csb.statistics.samplers</a></strong>: <em class="summary">Defines abstract samplers.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc-module.html">csb.statistics.samplers.mc</a></strong>: <em class="summary">Abstract Monte Carlo samplers.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain-module.html">csb.statistics.samplers.mc.multichain</a></strong>: <em class="summary">Implements several extended-ensemble Monte Carlo sampling 
        algorithms.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator-module.html">csb.statistics.samplers.mc.neqsteppropagator</a></strong>: <em class="summary">Propagator class employing stepwise trajectories as used in the 
        NCMC algorithm (Nilmeier et al., &quot;Nonequilibrium candidate 
        Monte Carlo is an efficient tool for equilibrium simulation&quot;, 
        PNAS 2011)</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.propagators-module.html">csb.statistics.samplers.mc.propagators</a></strong>: <em class="summary">Provides various deterministic and stochastic propagators.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.singlechain-module.html">csb.statistics.samplers.mc.singlechain</a></strong>: <em class="summary">Various Monte Carlo equilibrium sampling algorithms, which simulate
        only one Markov chain.</em>    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture-module.html">csb.statistics.scalemixture</a></strong>: <em class="summary">Approximation of a distribution as a mixture of gaussians with a 
        zero mean but different sigmas</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statmech-module.html">csb.statmech</a></strong>: <em class="summary">Methods for statistical mechanics</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statmech.ensembles-module.html">csb.statmech.ensembles</a></strong>: <em class="summary">Statistical Ensembles</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statmech.wham-module.html">csb.statmech.wham</a></strong>: <em class="summary">Estimate the free energy and density of states from tempered 
        ensembles using histogram re-weighting.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.test-module.html">csb.test</a></strong>: <em class="summary">This is a top level package, hosting the entire CSB test framework.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.test.app-module.html">csb.test.app</a></strong>: <em class="summary">CSB Test Runner app.</em>    </li>
    </ul>
    </li>
  </ul></td></tr>
</table>

<br />
<!-- ==================== CLASSES ==================== -->
<a name="section-Classes"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td align="left" colspan="2" class="table-header">
    <span class="table-header">Classes</span></td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.Version-class.html" class="summary-name">Version</a><br />
      CSB version number.
    </td>
  </tr>
</table>
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th bgcolor="#70b0f0" class="navbar-select"
          >&nbsp;&nbsp;&nbsp;Home&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >CSB v1.2.3.617</th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Sat Jan 18 14:12:13 2014
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
