

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Installation</title>
    
    <link rel="stylesheet" href="../_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.6.5a',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/gemini.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="Quick start" href="quick_start.html" />
    <link rel="prev" title="GEMINI: a flexible framework for exploring genome variation" href="../index.html" />


<script src="../_static/labibi.js"></script>



<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', '']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="quick_start.html" title="Quick start"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="GEMINI: a flexible framework for exploring genome variation"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">gemini v0.6.5a</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  
  <div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="automated-installation">
<h2>Automated installation<a class="headerlink" href="#automated-installation" title="Permalink to this headline">¶</a></h2>
<p>GEMINI contains an automated installation script which installs
GEMINI along with required Python dependencies, third party software
and data files.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>wget https://raw.github.com/arq5x/gemini/master/gemini/scripts/gemini_install.py
</pre></div>
</div>
<p>If <tt class="docutils literal"><span class="pre">wget</span></tt> isn&#8217;t available, you can use <tt class="docutils literal"><span class="pre">curl</span></tt> instead:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>curl <span class="se">\</span>
https://raw.github.com/arq5x/gemini/master/gemini/scripts/gemini_install.py <span class="se">\</span>
&gt; gemini_install.py
</pre></div>
</div>
<p>Once you have downloaded the above installation script, you can proceed as follows:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>python gemini_install.py /usr/local /usr/local/share/gemini
<span class="nv">$ </span><span class="nb">export </span><span class="nv">PATH</span><span class="o">=</span><span class="nv">$PATH</span>:/usr/local/gemini/bin
<span class="c"># it would be wise to add the above line to your ``.bashrc`` or ``.bash_profile``</span>
</pre></div>
</div>
<p>This installs the GEMINI executable as <tt class="docutils literal"><span class="pre">/usr/local/bin/gemini</span></tt>,
other required third party dependencies in <tt class="docutils literal"><span class="pre">/usr/local/bin</span></tt>, and
associated data files in <tt class="docutils literal"><span class="pre">/usr/local/share/gemini</span></tt>.  <strong>Please note that this
is merely an example: one can place the GEMINI executables and annotation files
in any directories one wishes.</strong></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The automated installtion script typically requires ~30 minutes, primarily
owing to the time required to download the GEMINI genome annotation files.
Also, please note that the annotation files <strong>requires ~15Gb of storage</strong>,
so please ensure that the data directory (<tt class="docutils literal"><span class="pre">/usr/local/share/gemini</span></tt> in
the example above) has sufficient space.</p>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p><strong>Some tips and tricks for installation issues:</strong></p>
<ol class="last arabic simple">
<li>Some older versions of wget have certificate problems with GitHub
files. If you run into this problem, you can alternatively download
the install script using``wget &#8211;no-check-certificates`` or <tt class="docutils literal"><span class="pre">curl</span> <span class="pre">-O</span></tt>.</li>
<li>The installation script is idempotent and you can re-run it multiple
times without any issues. If you experience internet connectivity or
other transient errors during installation, a re-run can often solve
the problem (fingers crossed).</li>
<li>If you&#8217;re installing behind a proxy you&#8217;ll need to set proxy information in
a <a class="reference external" href="http://docs.continuum.io/conda/intro.html#configuration">condarc file</a> and also set <tt class="docutils literal"><span class="pre">all_proxy</span></tt>, <tt class="docutils literal"><span class="pre">http_proxy</span></tt> and <tt class="docutils literal"><span class="pre">https_proxy</span></tt>
in your <tt class="docutils literal"><span class="pre">.bashrc</span></tt> file. See <a class="reference external" href="https://groups.google.com/d/msg/gemini-variation/WMQiMDeW6aE/6oa8nS1NorUJ">this mailing list discussion</a> for more information.</li>
</ol>
</div>
<div class="section" id="dependencies">
<h3>Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h3>
<p>The installer requires:</p>
<blockquote>
<div><ul class="simple">
<li>Python 2.7.x</li>
<li>git</li>
<li>a working C / C++ compiler such as gcc</li>
<li>zlib (including headers)</li>
</ul>
</div></blockquote>
<p>These dependencies already exist on most UNIX/LINUX/OS X machines.  However,
on minimalist machines <strong>such as fresh Amazon Cloud instances</strong>, one may have
to install these dependencies before running the automated installer. In the
case of Amazon instances, the following command should take care of all of the
above dependencies:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>sudo yum -y install python27 git gcc gcc-c++ zlib-devel
</pre></div>
</div>
</div>
<div class="section" id="installing-without-root-access">
<h3>Installing without root access.<a class="headerlink" href="#installing-without-root-access" title="Permalink to this headline">¶</a></h3>
<p>As many users do not have root or sudo access, the automated installer also
has options to install in &#8220;non-root&#8221; environments:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>python gemini_install.py ~/gemini ~/gemini --nosudo
</pre></div>
</div>
</div>
<div class="section" id="updating-your-path-to-find-the-gemini-executable">
<h3>Updating your PATH to find the GEMINI executable<a class="headerlink" href="#updating-your-path-to-find-the-gemini-executable" title="Permalink to this headline">¶</a></h3>
<p>At this point, you will have a self-contained installation of GEMINI,
including both the software and its associated genome annotations. However,
if you have done a custom install in a &#8220;non-root&#8221; enviornment, you will
first need to update your <tt class="docutils literal"><span class="pre">PATH</span></tt> environment variable to include the path
to the bin directory that you just created by running the automated installer.</p>
<p>For example, if, as above, you placed you custom install in <tt class="docutils literal"><span class="pre">~/gemini</span></tt>, you
would need to update your <tt class="docutils literal"><span class="pre">PATH</span></tt> as follows. It would be wise to also add this
to your <tt class="docutils literal"><span class="pre">.bashrc</span></tt> or <tt class="docutils literal"><span class="pre">.bash_profile</span></tt>:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span><span class="nb">export </span><span class="nv">PATH</span><span class="o">=</span><span class="nv">$PATH</span>:~/gemini/bin
</pre></div>
</div>
<p>Note that this change will only last for the life of your current terminal
session.  To make this more permanent, update your <tt class="docutils literal"><span class="pre">.bash_profile</span></tt> so that
this change is made each time you login.</p>
<p>If successful, you should be able to run the following command from anywhere
on your system:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini -v
gemini 0.3.0b
</pre></div>
</div>
</div>
<div class="section" id="running-unit-tests">
<h3>Running unit tests<a class="headerlink" href="#running-unit-tests" title="Permalink to this headline">¶</a></h3>
<p>If successfully installed, you should be able to change
directories into the <tt class="docutils literal"><span class="pre">gemini</span></tt> directory within the
directory into which you installed the GEMINI source code
and run a script of unit tests that will ensure that GEMINI
is running appropriately on your system. For example, if, as
above, you installed the GEMINI executables to <tt class="docutils literal"><span class="pre">~/gemini</span></tt>, you
would issue the following commands to run the unit tests:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span><span class="nb">cd</span> ~/gemini/gemini
<span class="nv">$ </span>bash master-test.sh
</pre></div>
</div>
</div>
<div class="section" id="updating-the-gemini-executables-and-annotations">
<h3>Updating the GEMINI executables and annotations<a class="headerlink" href="#updating-the-gemini-executables-and-annotations" title="Permalink to this headline">¶</a></h3>
<p>Once installed with the automated installer, it is easy to upgrade the
GEMINI programs and associated data files to the latest released
version with:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini update
</pre></div>
</div>
<p>There are also flags to update to the latest development version of the code or
to only update data files without updating the source:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini update --devel
<span class="nv">$ </span>gemini update --dataonly
</pre></div>
</div>
</div>
</div>
<div class="section" id="software-dependencies">
<h2>Software dependencies<a class="headerlink" href="#software-dependencies" title="Permalink to this headline">¶</a></h2>
<p>GEMINI depends upon several widely-used genomics command line software as well
as multiple Python packages.  We recognize that the dependency stack is quite
deep and are working on ways to minimize dependencies in the interest of the
most streamlined installation process possible.  Nonetheless, the following are
core dependencies:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Python 2.7.x</li>
<li><a class="reference external" href="https://github.com/arq5x/grabix">grabix</a></li>
<li><a class="reference external" href="http://sourceforge.net/projects/samtools/files/">samtools</a></li>
<li><a class="reference external" href="http://sourceforge.net/projects/samtools/files/">tabix</a></li>
<li><a class="reference external" href="https://code.google.com/p/bedtools/">bedtools</a></li>
<li><a class="reference external" href="http://pythonhosted.org/pybedtools/main.html#installing-pybedtools">pybedtools</a></li>
</ol>
</div></blockquote>
</div>
<div class="section" id="manual-installation">
<h2>Manual installation<a class="headerlink" href="#manual-installation" title="Permalink to this headline">¶</a></h2>
<p>Once the above dependencies have been installed, one can begin installing
<tt class="docutils literal"><span class="pre">GEMINI</span></tt> itself. To install you should download the latest source code from
GitHub, either by going to:</p>
<div class="highlight-python"><pre>http://github.com/arq5x/gemini</pre>
</div>
<p>and clicking on &#8220;Downloads&#8221;, or by cloning the git repository with:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>git clone https://github.com/arq5x/gemini.git
</pre></div>
</div>
<p>Once you have the source code, run:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span><span class="nb">cd </span>gemini
<span class="nv">$ </span>sudo python setup.py install
</pre></div>
</div>
<p>to install it. If you don&#8217;t have permission to install it in the default
directory, you can simply build the source in-place and use the package
from the git repository:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>python setup.py build_ext --inplace
</pre></div>
</div>
</div>
<div class="section" id="installing-annotation-files">
<h2>Installing annotation files<a class="headerlink" href="#installing-annotation-files" title="Permalink to this headline">¶</a></h2>
<p>One of the more appealing features in <tt class="docutils literal"><span class="pre">GEMINI</span></tt> is that it automatically
annotates variants in a VCF file with several genome annotations.  However,
you must first install these data files on your system. It&#8217;s easy enough &#8212;
you just need to run the following script and tell it in which what full path
you&#8217;d like to install the necessary data files. The recommended path is
<tt class="docutils literal"><span class="pre">/usr/local/share</span></tt>, but you can install the data files wherever you want.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>python gemini/install-data.py /usr/local/share/
</pre></div>
</div>
</div>
<div class="section" id="using-previously-installed-annotation-files">
<h2>Using previously installed annotation files<a class="headerlink" href="#using-previously-installed-annotation-files" title="Permalink to this headline">¶</a></h2>
<p>If you have installed GEMINI with the annotation files on a server and you can NFS mount
the annotation files, you can tell a local install of GEMINI where those annotation files
are by making the file ~/.gemini/gemini-config.yaml:</p>
<div class="highlight-python"><pre>annotation_dir: /path/to/nfs_mounted/gemini/data
versions:
  GRCh37-gms-mappability.vcf.gz: 2
  hg19.rmsk.bed.gz: 2</pre>
</div>
</div>
<div class="section" id="running-the-testing-suite">
<h2>Running the testing suite<a class="headerlink" href="#running-the-testing-suite" title="Permalink to this headline">¶</a></h2>
<p>GEMINI comes with a full test suite to make sure that everything has installed
correctly on your system.  We <strong>strongly</strong> encourage you to run these tests.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bash master-test.sh
</pre></div>
</div>
<p><cite>GEMINI</cite> depends upon external tools to predict the functional consequence of variants in a VCF file.
We currently support annotations produced by both <a class="reference external" href="http://snpeff.sourceforge.net/">SnpEff</a>
and <a class="reference external" href="http://www.ensembl.org/info/docs/variation/vep/index.html">VEP</a>.
Recommended instructions for annotating existing VCF files with these tools are available here.
In addition, we have attempted to standardize the terms used to describe the functional consequence of a given variant,
as each annotation tool uses different vocabulary.</p>
<p>The variant consequence columns in the variant table are populated either by <cite>snpEff</cite> or <cite>VEP</cite> as defined by the user using the <cite>-t</cite> option while running gemini load
(To populate these columns the input VCF file should have been annotated either by <cite>snpEff</cite> or <cite>VEP</cite>):</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini load -v my.vcf -t VEP -d my.db
<span class="nv">$ </span>gemini load -v my.vcf -t snpEFF -d my.db
</pre></div>
</div>
<p>By default the following columns in the variant table would be set to null:</p>
<ul class="simple">
<li>anno_id</li>
<li>gene</li>
<li>affected_gene</li>
<li>affected_transcript</li>
<li>affected_exon</li>
<li>is_exonic</li>
<li>is_lof</li>
<li>is_coding</li>
<li>codon_change</li>
<li>aa_change</li>
<li>aa_length</li>
<li>biotype</li>
<li>most_severe_impact</li>
<li>impact_severity</li>
<li>polyphen_pred</li>
<li>polyphen_score</li>
<li>sift_pred</li>
<li>sift_score</li>
</ul>
<p>The table below shows the alternate <cite>GEMINI</cite> terms for the consequences from <cite>snpEff</cite> and <cite>VEP</cite>, for SQL queries.
The last column represents the severity terms associated with the impacts:</p>
<table border="1" class="docutils">
<colgroup>
<col width="29%" />
<col width="22%" />
<col width="34%" />
<col width="15%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Gemini terms</th>
<th class="head">snpEff terms</th>
<th class="head">VEP terms</th>
<th class="head">Impact severity</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>splice_acceptor</td>
<td>SPLICE_SITE_ACCEPTOR</td>
<td>splice_acceptor_variant</td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>splice_donor</td>
<td>SPLICE_SITE_DONOR</td>
<td>splice_donor_variant</td>
<td>HIGH</td>
</tr>
<tr class="row-even"><td>stop_gain</td>
<td>STOP_GAINED</td>
<td>stop_gained</td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>stop_loss</td>
<td>STOP_LOST</td>
<td>stop_lost</td>
<td>HIGH</td>
</tr>
<tr class="row-even"><td>frame_shift</td>
<td>FRAME_SHIFT</td>
<td>frameshift_variant</td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>start_loss</td>
<td>START_LOST</td>
<td>null</td>
<td>HIGH</td>
</tr>
<tr class="row-even"><td>exon_deleted</td>
<td>EXON_DELETED</td>
<td>null</td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>non_synonymous_start</td>
<td>NON_SYNONYMOUS_START</td>
<td>null</td>
<td>HIGH</td>
</tr>
<tr class="row-even"><td>transcript_codon_change</td>
<td>null</td>
<td>initiator_codon_variant</td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>non_syn_coding</td>
<td>NON_SYNONYMOUS_CODING</td>
<td>missense_variant</td>
<td>MED</td>
</tr>
<tr class="row-even"><td>inframe_codon_gain</td>
<td>CODON_INSERTION</td>
<td>inframe_insertion</td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>inframe_codon_loss</td>
<td>CODON_DELETION</td>
<td>inframe_deletion</td>
<td>MED</td>
</tr>
<tr class="row-even"><td>inframe_codon_change</td>
<td>CODON_CHANGE</td>
<td>null</td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>codon_change_del</td>
<td>CODON_CHANGE_PLUS_CODON_DELETION</td>
<td>null</td>
<td>MED</td>
</tr>
<tr class="row-even"><td>codon_change_ins</td>
<td>CODON_CHANGE_PLUS_CODON_INSERTION</td>
<td>null</td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>UTR_5_del</td>
<td>UTR_5_DELETED</td>
<td>null</td>
<td>MED</td>
</tr>
<tr class="row-even"><td>UTR_3_del</td>
<td>UTR_3_DELETED</td>
<td>null</td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>other_splice_variant</td>
<td>null</td>
<td>splice_region_variant</td>
<td>MED</td>
</tr>
<tr class="row-even"><td>mature_miRNA</td>
<td>null</td>
<td>mature_miRNA_variant</td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>regulatory_region</td>
<td>null</td>
<td>regulatory_region_variant</td>
<td>MED</td>
</tr>
<tr class="row-even"><td>TF_binding_site</td>
<td>null</td>
<td>TF_binding_site_variant</td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>regulatory_region_ablation</td>
<td>null</td>
<td>regulatory_region_ablation</td>
<td>MED</td>
</tr>
<tr class="row-even"><td>regulatory_region_amplification</td>
<td>null</td>
<td>regulatory_region_amplification</td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>TFBS_ablation</td>
<td>null</td>
<td>TFBS_ablation</td>
<td>MED</td>
</tr>
<tr class="row-even"><td>TFBS_amplification</td>
<td>null</td>
<td>TFBS_amplification</td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>synonymous_stop</td>
<td>SYNONYMOUS_STOP</td>
<td>stop_retained_variant</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>synonymous_coding</td>
<td>SYNONYMOUS_CODING</td>
<td>synonymous_variant</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>UTR_5_prime</td>
<td>UTR_5_PRIME</td>
<td>5_prime_UTR_variant</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>UTR_3_prime</td>
<td>UTR_3_PRIME</td>
<td>3_prime_UTR_variant</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>intron</td>
<td>INTRON</td>
<td>intron_variant</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>CDS</td>
<td>CDS</td>
<td>coding_sequence_variant</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>upstream</td>
<td>UPSTREAM</td>
<td>upstream_gene_variant</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>downstream</td>
<td>DOWNSTREAM</td>
<td>downstream_gene_variant</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>intergenic</td>
<td>INTERGENIC, INTERGENIC_CONSERVED</td>
<td>intergenic_variant</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>intragenic</td>
<td>INTRAGENIC</td>
<td>null</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>gene</td>
<td>GENE</td>
<td>null</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>transcript</td>
<td>TRANSCRIPT</td>
<td>null</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>exon</td>
<td>EXON</td>
<td>null</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>start_gain</td>
<td>START_GAINED</td>
<td>null</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>synonymous_start</td>
<td>SYNONYMOUS_START</td>
<td>null</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>intron_conserved</td>
<td>INTRON_CONSERVED</td>
<td>null</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>nc_transcript</td>
<td>null</td>
<td>nc_transcript_variant</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>NMD_transcript</td>
<td>null</td>
<td>NMD_transcript_variant</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>incomplete_terminal_codon</td>
<td>null</td>
<td>incomplete_terminal_codon_variant</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>nc_exon</td>
<td>null</td>
<td>non_coding_exon_variant</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>transcript_ablation</td>
<td>null</td>
<td>transcript_ablation</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>transcript_amplification</td>
<td>null</td>
<td>transcript_amplification</td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>feature elongation</td>
<td>null</td>
<td>feature elongation</td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>feature truncation</td>
<td>null</td>
<td>feature truncation</td>
<td>LOW</td>
</tr>
</tbody>
</table>
<p><em>Note: &#8220;null&#8221; refers to the absence of the corresponding term in the alternate database</em></p>
</div>
</div>




  

<div id="disqus_thread"></div>
 <script type="text/javascript">
     /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
     var disqus_shortname = 'geminidocs'; // required: replace example with your forum shortname

     /* * * DON'T EDIT BELOW THIS LINE * * */
     (function() {
         var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
         dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
         (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
     })();
 </script>
 <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
 <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>






          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/gemini.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Installation</a><ul>
<li><a class="reference internal" href="#automated-installation">Automated installation</a><ul>
<li><a class="reference internal" href="#dependencies">Dependencies</a></li>
<li><a class="reference internal" href="#installing-without-root-access">Installing without root access.</a></li>
<li><a class="reference internal" href="#updating-your-path-to-find-the-gemini-executable">Updating your PATH to find the GEMINI executable</a></li>
<li><a class="reference internal" href="#running-unit-tests">Running unit tests</a></li>
<li><a class="reference internal" href="#updating-the-gemini-executables-and-annotations">Updating the GEMINI executables and annotations</a></li>
</ul>
</li>
<li><a class="reference internal" href="#software-dependencies">Software dependencies</a></li>
<li><a class="reference internal" href="#manual-installation">Manual installation</a></li>
<li><a class="reference internal" href="#installing-annotation-files">Installing annotation files</a></li>
<li><a class="reference internal" href="#using-previously-installed-annotation-files">Using previously installed annotation files</a></li>
<li><a class="reference internal" href="#running-the-testing-suite">Running the testing suite</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../index.html"
                        title="previous chapter"><strong>GEMINI</strong>: <em>a flexible framework for exploring genome variation</em></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="quick_start.html"
                        title="next chapter">Quick start</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/installation.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="quick_start.html" title="Quick start"
             >next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="GEMINI: a flexible framework for exploring genome variation"
             >previous</a> |</li>
        <li><a href="../index.html">gemini v0.6.5a</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2012,2013.
      Last updated on Jan 05, 2014.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>


<div id="editor-trap">
      <h3>Edit and improve this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/arq5x/gemini/blob/master/docs/content/installation.rst">
          Installation
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/arq5x/gemini/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', UA-39150563-1]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>