<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>Class Hierarchy</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="csb-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th bgcolor="#70b0f0" class="navbar-select"
          >&nbsp;&nbsp;&nbsp;Trees&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >CSB v1.2.2.614</th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">&nbsp;</td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="class-tree.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<center><b>
 [ <a href="module-tree.html">Module Hierarchy</a>
 | <a href="class-tree.html">Class Hierarchy</a> ]
</b></center><br />
<h1 class="epydoc">Class Hierarchy</h1>
<ul class="nomargin-top">
    <li> <strong class="uidlink"><a href="csb.bio.io.procheck.ProcheckParser-class.html">csb.bio.io.procheck.ProcheckParser</a></strong>:
      <em class="summary">Simple Prochceck Summary parser</em>
    </li>
    <li> <strong class="uidlink">StringIO.StringIO</strong>:
      <em class="summary">class StringIO([buffer])</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.io.MemoryStream-class.html">csb.io.MemoryStream</a></strong>:
      <em class="summary">In-memory stream object.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">object</strong>:
      <em class="summary">The most base type</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.A3MSequenceIterator-class.html">csb.bio.io.fasta.A3MSequenceIterator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.ars.ARS-class.html">csb.statistics.ars.ARS</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.AbstractAlignment-class.html">csb.bio.sequence.AbstractAlignment</a></strong>:
      <em class="summary">Base class for all alignment objects.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.A3MAlignment-class.html">csb.bio.sequence.A3MAlignment</a></strong>:
      <em class="summary">A specific type of multiple alignment, which provides some 
        operations relative to a master sequence (the first entry in the 
        alignment).</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.SequenceAlignment-class.html">csb.bio.sequence.SequenceAlignment</a></strong>:
      <em class="summary">Multiple sequence alignment.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.HHpredHitAlignment-class.html">csb.bio.hmm.HHpredHitAlignment</a></strong>:
      <em class="summary">Represents a query-template alignment in an HHpred result.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.StructureAlignment-class.html">csb.bio.sequence.StructureAlignment</a></strong>:
      <em class="summary">Multiple structure alignment.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.alignment.AbstractAlignmentAlgorithm-class.html">csb.bio.sequence.alignment.AbstractAlignmentAlgorithm</a></strong>:
      <em class="summary">Base class for all sequence alignment algorithms.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.alignment.GlobalAlignmentAlgorithm-class.html">csb.bio.sequence.alignment.GlobalAlignmentAlgorithm</a></strong>:
      <em class="summary">Needleman-Wunsch global sequence alignment.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.RobustResidueMapper.GlobalAligner-class.html">csb.bio.io.wwpdb.RobustResidueMapper.GlobalAligner</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.alignment.LocalAlignmentAlgorithm-class.html">csb.bio.sequence.alignment.LocalAlignmentAlgorithm</a></strong>:
      <em class="summary">Smith-Waterman local sequence alignment.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.AbstractContainer-class.html">csb.core.AbstractContainer</a></strong>:
      <em class="summary">Defines the behavior of a high-level object, which can hold an 
        array of elements.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.core.AbstractNIContainer-class.html">csb.core.AbstractNIContainer</a></strong>:
      <em class="summary">Same as the <a href="csb.core.AbstractContainer-class.html" 
        class="link">AbstractContainer</a>, but provides access to the 
        child elements through 
        <code>AbstractNativeIndexer._getinternal</code> instead of the 
        standard __getitem__.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Chain-class.html">csb.bio.structure.Chain</a></strong>:
      <em class="summary">Represents a polymeric chain.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Ensemble-class.html">csb.bio.structure.Ensemble</a></strong>:
      <em class="summary">Represents an ensemble of multiple <a 
        href="csb.bio.structure.Structure-class.html" 
        class="link">Structure</a> models.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Residue-class.html">csb.bio.structure.Residue</a></strong>:
      <em class="summary">Base class representing a single residue.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.NucleicResidue-class.html">csb.bio.structure.NucleicResidue</a></strong>:
      <em class="summary">Represents a single nucleotide residue.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.ProteinResidue-class.html">csb.bio.structure.ProteinResidue</a></strong>:
      <em class="summary">Represents a single amino acid residue.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.UnknownResidue-class.html">csb.bio.structure.UnknownResidue</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Structure-class.html">csb.bio.structure.Structure</a></strong>:
      <em class="summary">Represents a single model of a PDB 3-Dimensional molecular 
        structure.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.Target-class.html">csb.bio.fragments.Target</a></strong>:
      <em class="summary">Represents a protein structure prediction target.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ChemShiftTarget-class.html">csb.bio.fragments.ChemShiftTarget</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.Container-class.html">csb.core.Container</a></strong>:
      <em class="summary">Generic implementation of <a 
        href="csb.core.AbstractContainer-class.html" 
        class="link">AbstractContainer</a>.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.rosetta.RosettaFragmentMap-class.html">csb.bio.fragments.rosetta.RosettaFragmentMap</a></strong>:
      <em class="summary">Represents a Rosetta fragment library.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.AbstractDensity-class.html">csb.statistics.pdf.AbstractDensity</a></strong>:
      <em class="summary">Defines the interface and common operations for all probability 
        density functions.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.BaseDensity-class.html">csb.statistics.pdf.BaseDensity</a></strong>:
      <em class="summary">Base abstract class for all PDFs, which operate on simple float or 
        array-of-float parameters.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.Dirichlet-class.html">csb.statistics.pdf.Dirichlet</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.Gamma-class.html">csb.statistics.pdf.Gamma</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.GammaPrior-class.html">csb.statistics.scalemixture.GammaPrior</a></strong>:
      <em class="summary">Gamma prior on mixture weights including a weak gamma prior on its 
        parameter.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.GeneralizedInverseGaussian-class.html">csb.statistics.pdf.GeneralizedInverseGaussian</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.GeneralizedNormal-class.html">csb.statistics.pdf.GeneralizedNormal</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.GumbelMinimum-class.html">csb.statistics.pdf.GumbelMinimum</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.GumbelMaximum-class.html">csb.statistics.pdf.GumbelMaximum</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.InverseGamma-class.html">csb.statistics.pdf.InverseGamma</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.InvGammaPrior-class.html">csb.statistics.scalemixture.InvGammaPrior</a></strong>:
      <em class="summary">Inverse Gamma prior on mixture weights including a weak gamma prior
        on its parameter.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.InverseGaussian-class.html">csb.statistics.pdf.InverseGaussian</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.Laplace-class.html">csb.statistics.pdf.Laplace</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.maxent.MaxentModel-class.html">csb.statistics.maxent.MaxentModel</a></strong>:
      <em class="summary">Fourier expansion of a biangular log-probability density</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.Normal-class.html">csb.statistics.pdf.Normal</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.MultivariateGaussian-class.html">csb.statistics.pdf.MultivariateGaussian</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.ScaleMixture-class.html">csb.statistics.scalemixture.ScaleMixture</a></strong>:
      <em class="summary">Robust probabilistic superposition and comparison of protein 
        structures Martin Mechelke and Michael Habeck</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.parameterized.ParameterizedDensity-class.html">csb.statistics.pdf.parameterized.ParameterizedDensity</a></strong>:
      <em class="summary">Base abstract class for all PDFs, which operate on simple or 
        computed (chained) parameters.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.AbstractEntity-class.html">csb.bio.structure.AbstractEntity</a></strong>:
      <em class="summary">Base class for all protein structure entities.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Atom-class.html">csb.bio.structure.Atom</a></strong>:
      <em class="summary">Represents a single atom in space.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.DisorderedAtom-class.html">csb.bio.structure.DisorderedAtom</a></strong>:
      <em class="summary">A wobbling atom, which has alternative locations.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Chain-class.html">csb.bio.structure.Chain</a></strong>:
      <em class="summary">Represents a polymeric chain.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Ensemble-class.html">csb.bio.structure.Ensemble</a></strong>:
      <em class="summary">Represents an ensemble of multiple <a 
        href="csb.bio.structure.Structure-class.html" 
        class="link">Structure</a> models.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Residue-class.html">csb.bio.structure.Residue</a></strong>:
      <em class="summary">Base class representing a single residue.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.NucleicResidue-class.html">csb.bio.structure.NucleicResidue</a></strong>:
      <em class="summary">Represents a single nucleotide residue.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.ProteinResidue-class.html">csb.bio.structure.ProteinResidue</a></strong>:
      <em class="summary">Represents a single amino acid residue.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.UnknownResidue-class.html">csb.bio.structure.UnknownResidue</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Structure-class.html">csb.bio.structure.Structure</a></strong>:
      <em class="summary">Represents a single model of a PDB 3-Dimensional molecular 
        structure.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.AbstractEstimator-class.html">csb.statistics.pdf.AbstractEstimator</a></strong>:
      <em class="summary">Density parameter estimation strategy.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.DirichletEstimator-class.html">csb.statistics.pdf.DirichletEstimator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.GammaMLEstimator-class.html">csb.statistics.pdf.GammaMLEstimator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.GammaScaleMAP-class.html">csb.statistics.scalemixture.GammaScaleMAP</a></strong>:
      <em class="summary">MAP estimator of the scales</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.GammaScaleSampler-class.html">csb.statistics.scalemixture.GammaScaleSampler</a></strong>:
      <em class="summary">Sample the scalies given the data</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.GaussianMLEstimator-class.html">csb.statistics.pdf.GaussianMLEstimator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.GenNormalBruteForceEstimator-class.html">csb.statistics.pdf.GenNormalBruteForceEstimator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.GumbelMaxMomentsEstimator-class.html">csb.statistics.pdf.GumbelMaxMomentsEstimator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.GumbelMinMomentsEstimator-class.html">csb.statistics.pdf.GumbelMinMomentsEstimator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.InvGammaScaleMAP-class.html">csb.statistics.scalemixture.InvGammaScaleMAP</a></strong>:
      <em class="summary">MAP estimator of the scales</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.InvGammaScaleSampler-class.html">csb.statistics.scalemixture.InvGammaScaleSampler</a></strong>:
      <em class="summary">Sample the scales given the data</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.InverseGaussianMLEstimator-class.html">csb.statistics.pdf.InverseGaussianMLEstimator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.LaplaceMLEstimator-class.html">csb.statistics.pdf.LaplaceMLEstimator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.MultivariateGaussianMLEstimator-class.html">csb.statistics.pdf.MultivariateGaussianMLEstimator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.NullEstimator-class.html">csb.statistics.pdf.NullEstimator</a></strong>:
      <em class="summary">Does not estimate anything.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.ScaleMixturePriorEstimator-class.html">csb.statistics.scalemixture.ScaleMixturePriorEstimator</a></strong>:
      <em class="summary">Prior on the scales of a <a 
        href="csb.statistics.scalemixture.ScaleMixture-class.html" 
        class="link">ScaleMixture</a>, which determines how the scales are 
        estimated.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.GammaPosteriorMAP-class.html">csb.statistics.scalemixture.GammaPosteriorMAP</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.GammaPosteriorSampler-class.html">csb.statistics.scalemixture.GammaPosteriorSampler</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.InvGammaPosteriorMAP-class.html">csb.statistics.scalemixture.InvGammaPosteriorMAP</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.InvGammaPosteriorSampler-class.html">csb.statistics.scalemixture.InvGammaPosteriorSampler</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.numeric.integrators.AbstractGradient-class.html">csb.numeric.integrators.AbstractGradient</a></strong>:
      <em class="summary">Abstract gradient class.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.Gradient-class.html">csb.statistics.samplers.mc.multichain.Gradient</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.AbstractHeatWorkJacobianLogger-class.html">csb.statistics.samplers.mc.neqsteppropagator.AbstractHeatWorkJacobianLogger</a></strong>:
      <em class="summary">Abstract class defining the interface for objects keeping track of 
        and accumulating heat, work and the Jacobian during a 
        nonequilibrium trajectory.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.TrivialHeatWorkJacobianLogger-class.html">csb.statistics.samplers.mc.neqsteppropagator.TrivialHeatWorkJacobianLogger</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.AbstractIndexer-class.html">csb.core.AbstractIndexer</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.core.BaseCollectionContainer-class.html">csb.core.BaseCollectionContainer</a></strong>:
      <em class="summary">Base class which defines the behavior of a read-only collection 
        container.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.core.CollectionContainer-class.html">csb.core.CollectionContainer</a></strong>:
      <em class="summary">Write-enabled Collection Container.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.ChainResiduesCollection-class.html">csb.bio.structure.ChainResiduesCollection</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.DisorderedAtom-class.html">csb.bio.structure.DisorderedAtom</a></strong>:
      <em class="summary">A wobbling atom, which has alternative locations.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.EnsembleModelsCollection-class.html">csb.bio.structure.EnsembleModelsCollection</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.HMMLayersCollection-class.html">csb.bio.hmm.HMMLayersCollection</a></strong>:
      <em class="summary">Provides consecutive, 1-based access to all of the layers in the 
        profile.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.SecondaryStructure-class.html">csb.bio.structure.SecondaryStructure</a></strong>:
      <em class="summary">Describes the secondary structure of a chain.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.TorsionAnglesCollection-class.html">csb.bio.structure.TorsionAnglesCollection</a></strong>:
      <em class="summary">Describes a collection of torsion angles.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.Trajectory-class.html">csb.statistics.samplers.mc.Trajectory</a></strong>:
      <em class="summary">Ordered collection of states, representing a phase-space 
        trajectory.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.NonequilibriumTrajectory-class.html">csb.statistics.samplers.mc.neqsteppropagator.NonequilibriumTrajectory</a></strong>:
      <em class="summary">Trajectory holding additional information about energy difference 
        the Jacobian.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.PerturbationResult-class.html">csb.statistics.samplers.mc.neqsteppropagator.PerturbationResult</a></strong>:
      <em class="summary">Instances hold the result of a perturbation.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.EnsembleState-class.html">csb.statistics.samplers.EnsembleState</a></strong>:
      <em class="summary">Defines an Ensemble Monte Carlo state; it is a read-only collection
        of State objects.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.MCCollection-class.html">csb.statistics.samplers.mc.MCCollection</a></strong>:
      <em class="summary">Collection of single-chain samplers.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.ReadOnlyCollectionContainer-class.html">csb.core.ReadOnlyCollectionContainer</a></strong>:
      <em class="summary">This is a write-once container, which is filled with items only at 
        object construction.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.ClansEntryCollection-class.html">csb.bio.io.clans.ClansEntryCollection</a></strong>:
      <em class="summary">Read-only container for holding (and internally managing) <a 
        href="csb.bio.io.clans.ClansEntry-class.html" 
        class="link">ClansEntry</a> instances.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.ClansSeqgroupCollection-class.html">csb.bio.io.clans.ClansSeqgroupCollection</a></strong>:
      <em class="summary">Read-only container for holding (and internally managing) <a 
        href="csb.bio.io.clans.ClansSeqgroup-class.html" 
        class="link">ClansSeqgroup</a> instances.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.SequenceCollection-class.html">csb.bio.sequence.SequenceCollection</a></strong>:
      <em class="summary">Represents a list of <a 
        href="csb.bio.sequence.AbstractSequence-class.html" 
        class="link">AbstractSequence</a>s.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.BaseDictionaryContainer-class.html">csb.core.BaseDictionaryContainer</a></strong>:
      <em class="summary">Base class which defines the behavior of a read only key-value 
        collection container.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.AlignmentRowsTable-class.html">csb.bio.sequence.AlignmentRowsTable</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.DictionaryContainer-class.html">csb.core.DictionaryContainer</a></strong>:
      <em class="summary">Write-enabled Dictionary Container.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.vasco.ChemicalShiftContainer-class.html">csb.bio.io.vasco.ChemicalShiftContainer</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.EmissionTable-class.html">csb.bio.hmm.EmissionTable</a></strong>:
      <em class="summary">Represents a lookup table of emission probabilities.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.HMMLayer-class.html">csb.bio.hmm.HMMLayer</a></strong>:
      <em class="summary">Provides a dictionary-like catalog of the available states at this 
        layer.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.ResidueAtomsTable-class.html">csb.bio.structure.ResidueAtomsTable</a></strong>:
      <em class="summary">Represents a collection of atoms.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.StructureChainsTable-class.html">csb.bio.structure.StructureChainsTable</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.TransitionTable-class.html">csb.bio.hmm.TransitionTable</a></strong>:
      <em class="summary">Represents a lookup table of transitions that are possible from 
        within a given state.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.ReadOnlyDictionaryContainer-class.html">csb.core.ReadOnlyDictionaryContainer</a></strong>:
      <em class="summary">This is a write-once container, which is filled with items only at 
        object construction.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.numeric.integrators.AbstractIntegrator-class.html">csb.numeric.integrators.AbstractIntegrator</a></strong>:
      <em class="summary">Abstract integrator class.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.numeric.integrators.LeapFrog-class.html">csb.numeric.integrators.LeapFrog</a></strong>:
      <em class="summary">Leap Frog integration scheme implementation that calculates 
        position and momenta at equal times.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.numeric.integrators.FastLeapFrog-class.html">csb.numeric.integrators.FastLeapFrog</a></strong>:
      <em class="summary">Leap Frog integration scheme implementation that calculates 
        position and momenta at unequal times by concatenating the momentum
        updates of two successive integration steps.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.numeric.integrators.VelocityVerlet-class.html">csb.numeric.integrators.VelocityVerlet</a></strong>:
      <em class="summary">Velocity Verlet integration scheme implementation.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.parameterized.AbstractParameter-class.html">csb.statistics.pdf.parameterized.AbstractParameter</a></strong>:
      <em class="summary">Abstract parameterization, which can exist independently or be 
        coupled to other parameters upon request.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.parameterized.Parameter-class.html">csb.statistics.pdf.parameterized.Parameter</a></strong>:
      <em class="summary">Default parameter implementation which accepts float values only.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.parameterized.NonVirtualParameter-class.html">csb.statistics.pdf.parameterized.NonVirtualParameter</a></strong>:
      <em class="summary">A float <a 
        href="csb.statistics.pdf.parameterized.Parameter-class.html" 
        class="link">Parameter</a> that is explicitly non-computed and 
        cannot be bound to another <a 
        href="csb.statistics.pdf.parameterized.Parameter-class.html" 
        class="link">Parameter</a>.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.AbstractPerturbation-class.html">csb.statistics.samplers.mc.neqsteppropagator.AbstractPerturbation</a></strong>:
      <em class="summary">Describes an abstract system perturbation</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.ReducedHamiltonianPerturbation-class.html">csb.statistics.samplers.mc.neqsteppropagator.ReducedHamiltonianPerturbation</a></strong>:
      <em class="summary">System perturbation by changing the reduced Hamiltonian</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.AbstractPerturbationParam-class.html">csb.statistics.samplers.mc.neqsteppropagator.AbstractPerturbationParam</a></strong>:
      <em class="summary">Subclasses hold informations required for different kinds of system
        perturbation</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.AbstractPropagation-class.html">csb.statistics.samplers.mc.neqsteppropagator.AbstractPropagation</a></strong>:
      <em class="summary">Describes an abstract system propagation</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.AbstractMCPropagation-class.html">csb.statistics.samplers.mc.neqsteppropagator.AbstractMCPropagation</a></strong>:
      <em class="summary">System propagation by some MC algorithm.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.HMCPropagation-class.html">csb.statistics.samplers.mc.neqsteppropagator.HMCPropagation</a></strong>:
      <em class="summary">System propagation by HMC</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.AbstractMDPropagation-class.html">csb.statistics.samplers.mc.neqsteppropagator.AbstractMDPropagation</a></strong>:
      <em class="summary">System propagation by some MD algorithm</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.PlainMDPropagation-class.html">csb.statistics.samplers.mc.neqsteppropagator.PlainMDPropagation</a></strong>:
      <em class="summary">System propagation by plain, microcanonical MD</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.AbstractPropagationParam-class.html">csb.statistics.samplers.mc.neqsteppropagator.AbstractPropagationParam</a></strong>:
      <em class="summary">Subclasses hold informations required for different kinds of system
        propagation</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.HMCPropagationParam-class.html">csb.statistics.samplers.mc.neqsteppropagator.HMCPropagationParam</a></strong>:
      <em class="summary">Holds all required information for propagating a system by HMC.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.MDPropagationParam-class.html">csb.statistics.samplers.mc.neqsteppropagator.MDPropagationParam</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.PlainMDPropagationParam-class.html">csb.statistics.samplers.mc.neqsteppropagator.PlainMDPropagationParam</a></strong>:
      <em class="summary">Holds all required information for propagating a system by MD.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.AbstractPropagationResult-class.html">csb.statistics.samplers.mc.AbstractPropagationResult</a></strong>:
      <em class="summary">Abstract class providing the interface for the result of a 
        deterministic or stochastic propagation of a state.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.PropagationResult-class.html">csb.statistics.samplers.mc.PropagationResult</a></strong>:
      <em class="summary">Describes the result of a deterministic or stochastic propagation 
        of a state.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.Trajectory-class.html">csb.statistics.samplers.mc.Trajectory</a></strong>:
      <em class="summary">Ordered collection of states, representing a phase-space 
        trajectory.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.NonequilibriumTrajectory-class.html">csb.statistics.samplers.mc.neqsteppropagator.NonequilibriumTrajectory</a></strong>:
      <em class="summary">Trajectory holding additional information about energy difference 
        the Jacobian.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.PerturbationResult-class.html">csb.statistics.samplers.mc.neqsteppropagator.PerturbationResult</a></strong>:
      <em class="summary">Instances hold the result of a perturbation.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.propagators.AbstractPropagator-class.html">csb.statistics.samplers.mc.propagators.AbstractPropagator</a></strong>:
      <em class="summary">Abstract propagator class.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.propagators.AbstractMCPropagator-class.html">csb.statistics.samplers.mc.propagators.AbstractMCPropagator</a></strong>:
      <em class="summary">Provides the interface for MC trajectory generators.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.propagators.AbstractNCMCPropagator-class.html">csb.statistics.samplers.mc.propagators.AbstractNCMCPropagator</a></strong>:
      <em class="summary">Draws a number of samples from a PDF using the <a 
        href="csb.statistics.samplers.mc.singlechain.AbstractNCMCSampler-class.html"
        class="link">AbstractNCMCSampler</a>.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.propagators.HMCPropagator-class.html">csb.statistics.samplers.mc.propagators.HMCPropagator</a></strong>:
      <em class="summary">Draws a number of samples from a PDF using the <a 
        href="csb.statistics.samplers.mc.singlechain.HMCSampler-class.html"
        class="link">HMCSampler</a> and returns them as a <a 
        href="csb.statistics.samplers.mc.Trajectory-class.html" 
        class="link">Trajectory</a>.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.propagators.RWMCPropagator-class.html">csb.statistics.samplers.mc.propagators.RWMCPropagator</a></strong>:
      <em class="summary">Draws a number of samples from a PDF using the <a 
        href="csb.statistics.samplers.mc.singlechain.RWMCSampler-class.html"
        class="link">RWMCSampler</a> and returns them as a <a 
        href="csb.statistics.samplers.mc.Trajectory-class.html" 
        class="link">Trajectory</a>.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.propagators.MDPropagator-class.html">csb.statistics.samplers.mc.propagators.MDPropagator</a></strong>:
      <em class="summary">Molecular Dynamics propagator.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.propagators.ThermostattedMDPropagator-class.html">csb.statistics.samplers.mc.propagators.ThermostattedMDPropagator</a></strong>:
      <em class="summary">Thermostatted Molecular Dynamics propagator.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.NonequilibriumStepPropagator-class.html">csb.statistics.samplers.mc.neqsteppropagator.NonequilibriumStepPropagator</a></strong>:
      <em class="summary">Propagator class which propagates a system using NCMC-like stepwise
        trajectories</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.AbstractResidueMapper-class.html">csb.bio.io.wwpdb.AbstractResidueMapper</a></strong>:
      <em class="summary">Defines the base interface of all residue mappers, used to align 
        PDB ATOM records to the real (SEQRES) sequence of a chain.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.CombinedResidueMapper-class.html">csb.bio.io.wwpdb.CombinedResidueMapper</a></strong>:
      <em class="summary">The best of both worlds: attempts to map the residues using <a 
        href="csb.bio.io.wwpdb.FastResidueMapper-class.html" 
        class="link">FastResidueMapper</a>, but upon failure secures 
        success by switching to <a 
        href="csb.bio.io.wwpdb.RobustResidueMapper-class.html" 
        class="link">RobustResidueMapper</a>.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.FastResidueMapper-class.html">csb.bio.io.wwpdb.FastResidueMapper</a></strong>:
      <em class="summary">RegExp-based residue mapper.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.RobustResidueMapper-class.html">csb.bio.io.wwpdb.RobustResidueMapper</a></strong>:
      <em class="summary">Exhaustive residue mapper, which uses Needleman-Wunsch global 
        alignment.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.AbstractSampler-class.html">csb.statistics.samplers.AbstractSampler</a></strong>:
      <em class="summary">Abstract interface for sampling algorithms.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.AbstractMC-class.html">csb.statistics.samplers.mc.AbstractMC</a></strong>:
      <em class="summary">Abstract Monte Carlo sampler class.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.AbstractEnsembleMC-class.html">csb.statistics.samplers.mc.multichain.AbstractEnsembleMC</a></strong>:
      <em class="summary">Abstract class for Monte Carlo sampling algorithms simulating 
        several ensembles.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.AbstractExchangeMC-class.html">csb.statistics.samplers.mc.multichain.AbstractExchangeMC</a></strong>:
      <em class="summary">Abstract class for Monte Carlo sampling algorithms employing some 
        replica exchange method.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.AbstractRENS-class.html">csb.statistics.samplers.mc.multichain.AbstractRENS</a></strong>:
      <em class="summary">Abstract Replica Exchange with Nonequilibrium Switches (RENS, 
        Ballard &amp; Jarzynski 2009) class.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.AbstractStepRENS-class.html">csb.statistics.samplers.mc.multichain.AbstractStepRENS</a></strong>:
      <em class="summary">Replica Exchange with Nonequilibrium Switches (RENS, Ballard &amp; 
        Jarzynski 2009) with stepwise trajectories as described in Nilmeier
        et al., &quot;Nonequilibrium candidate Monte Carlo is an efficient 
        tool for equilibrium simulation&quot;, PNAS 2011.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.HMCStepRENS-class.html">csb.statistics.samplers.mc.multichain.HMCStepRENS</a></strong>:
      <em class="summary">Replica Exchange with Nonequilibrium Switches (RENS, Ballard &amp; 
        Jarzynski 2009) with stepwise trajectories as described in Nilmeier
        et al., &quot;Nonequilibrium candidate Monte Carlo is an efficient 
        tool for equilibrium simulation&quot;, PNAS 2011.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.MDRENS-class.html">csb.statistics.samplers.mc.multichain.MDRENS</a></strong>:
      <em class="summary">Replica Exchange with Nonequilibrium Switches (RENS, Ballard &amp; 
        Jarzynski 2009) with Molecular Dynamics (MD) trajectories.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.ThermostattedMDRENS-class.html">csb.statistics.samplers.mc.multichain.ThermostattedMDRENS</a></strong>:
      <em class="summary">Replica Exchange with Nonequilibrium Switches (RENS, Ballard &amp; 
        Jarzynski, 2009) with Andersen-thermostatted Molecular Dynamics 
        (MD) trajectories.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.ReplicaExchangeMC-class.html">csb.statistics.samplers.mc.multichain.ReplicaExchangeMC</a></strong>:
      <em class="summary">Replica Exchange (RE, Swendsen &amp; Yang 1986) implementation.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.singlechain.AbstractSingleChainMC-class.html">csb.statistics.samplers.mc.singlechain.AbstractSingleChainMC</a></strong>:
      <em class="summary">Abstract class for Monte Carlo sampling algorithms simulating only 
        one ensemble.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.singlechain.AbstractNCMCSampler-class.html">csb.statistics.samplers.mc.singlechain.AbstractNCMCSampler</a></strong>:
      <em class="summary">Implementation of the NCMC sampling algorithm (Nilmeier et al., 
        &quot;Nonequilibrium candidate Monte Carlo is an efficient tool for
        equilibrium simulation&quot;, PNAS 2011) for sampling from one 
        ensemble only.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.singlechain.HMCSampler-class.html">csb.statistics.samplers.mc.singlechain.HMCSampler</a></strong>:
      <em class="summary">Hamilton Monte Carlo (HMC, also called Hybrid Monte Carlo by the 
        inventors, Duane, Kennedy, Pendleton, Duncan 1987).</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.singlechain.RWMCSampler-class.html">csb.statistics.samplers.mc.singlechain.RWMCSampler</a></strong>:
      <em class="summary">Random Walk Metropolis Monte Carlo implementation (Metropolis, 
        Rosenbluth, Teller, Teller 1953; Hastings, 1970).</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.alignment.AbstractScoringMatrix-class.html">csb.bio.sequence.alignment.AbstractScoringMatrix</a></strong>:
      <em class="summary">Defines a pairwise sequence alignment scoring function.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.alignment.IdentityMatrix-class.html">csb.bio.sequence.alignment.IdentityMatrix</a></strong>:
      <em class="summary">Simple identity-based scoring matrix.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.alignment.SimilarityMatrix-class.html">csb.bio.sequence.alignment.SimilarityMatrix</a></strong>:
      <em class="summary">Similarity scoring matrix.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.AbstractSequence-class.html">csb.bio.sequence.AbstractSequence</a></strong>:
      <em class="summary">Base abstract class for all Sequence objects.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.ChainSequence-class.html">csb.bio.sequence.ChainSequence</a></strong>:
      <em class="summary">Sequence view for <a href="csb.bio.structure.Chain-class.html" 
        class="link">csb.bio.structure.Chain</a> objects.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.SparseChainSequence-class.html">csb.bio.io.wwpdb.SparseChainSequence</a></strong>:
      <em class="summary">Sequence view for reference (SEQRES) or sparse (ATOM) PDB chains.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.RichSequence-class.html">csb.bio.sequence.RichSequence</a></strong>:
      <em class="summary">Sequence implementation, which converts the sequence into a list of
        <a href="csb.bio.sequence.ResidueInfo-class.html" 
        class="link">ResidueInfo</a> objects.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.Sequence-class.html">csb.bio.sequence.Sequence</a></strong>:
      <em class="summary">High-performance sequence object.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.AbstractState-class.html">csb.statistics.samplers.AbstractState</a></strong>:
      <em class="summary">Represents a point in phase-space.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.EnsembleState-class.html">csb.statistics.samplers.EnsembleState</a></strong>:
      <em class="summary">Defines an Ensemble Monte Carlo state; it is a read-only collection
        of State objects.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.State-class.html">csb.statistics.samplers.State</a></strong>:
      <em class="summary">Represents a point in phase-space.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.AbstractStepStatistics-class.html">csb.statistics.samplers.mc.neqsteppropagator.AbstractStepStatistics</a></strong>:
      <em class="summary">Abstract class defining a minimal interface for objects allowing to
        store statistics of what happens in <a 
        href="csb.statistics.samplers.mc.neqsteppropagator.Step-class.html"
        class="link">Step</a> instances.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.DummyStepStatistics-class.html">csb.statistics.samplers.mc.neqsteppropagator.DummyStepStatistics</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.AbstractStructureParser-class.html">csb.bio.io.wwpdb.AbstractStructureParser</a></strong>:
      <em class="summary">A base PDB structure format-aware parser.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.LegacyStructureParser-class.html">csb.bio.io.wwpdb.LegacyStructureParser</a></strong>:
      <em class="summary">This is a customized PDB parser, which is designed to read both 
        sequence and atom data from the ATOM section.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.RegularStructureParser-class.html">csb.bio.io.wwpdb.RegularStructureParser</a></strong>:
      <em class="summary">This is the de facto PDB parser, which is designed to read SEQRES 
        and ATOM sections separately, and them map them.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.PDBHeaderParser-class.html">csb.bio.io.wwpdb.PDBHeaderParser</a></strong>:
      <em class="summary">Ultra fast PDB HEADER parser.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.AbstractSwapCommunicator-class.html">csb.statistics.samplers.mc.multichain.AbstractSwapCommunicator</a></strong>:
      <em class="summary">Holds all the information which needs to be communicated between 
        distinct swap substeps.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.RENSSwapCommunicator-class.html">csb.statistics.samplers.mc.multichain.RENSSwapCommunicator</a></strong>:
      <em class="summary">Holds all the information which needs to be communicated between 
        distinct RENS swap substeps.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.RESwapCommunicator-class.html">csb.statistics.samplers.mc.multichain.RESwapCommunicator</a></strong>:
      <em class="summary">Holds all the information which needs to be communicated between 
        distinct RE swap substeps.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.AbstractSwapParameterInfo-class.html">csb.statistics.samplers.mc.multichain.AbstractSwapParameterInfo</a></strong>:
      <em class="summary">Subclass instances hold all parameters necessary for performing a 
        swap between two given samplers.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.RESwapParameterInfo-class.html">csb.statistics.samplers.mc.multichain.RESwapParameterInfo</a></strong>:
      <em class="summary">Holds parameters for a standard Replica Exchange swap.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.AbstractRENSSwapParameterInfo-class.html">csb.statistics.samplers.mc.multichain.AbstractRENSSwapParameterInfo</a></strong>:
      <em class="summary">Holds parameters for a RENS swap.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.HMCStepRENSSwapParameterInfo-class.html">csb.statistics.samplers.mc.multichain.HMCStepRENSSwapParameterInfo</a></strong>:
      <em class="summary">Holds all required information for performing HMCStepRENS swaps.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.MDRENSSwapParameterInfo-class.html">csb.statistics.samplers.mc.multichain.MDRENSSwapParameterInfo</a></strong>:
      <em class="summary">Holds parameters for a MDRENS swap.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.ThermostattedMDRENSSwapParameterInfo-class.html">csb.statistics.samplers.mc.multichain.ThermostattedMDRENSSwapParameterInfo</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.AbstractSwapScheme-class.html">csb.statistics.samplers.mc.multichain.AbstractSwapScheme</a></strong>:
      <em class="summary">Provides the interface for classes defining schemes according to 
        which swaps in Replica Exchange-like simulations are performed.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.AlternatingAdjacentSwapScheme-class.html">csb.statistics.samplers.mc.multichain.AlternatingAdjacentSwapScheme</a></strong>:
      <em class="summary">Provides a swapping scheme in which tries exchanges between 
        neighbours only following the scheme 1 &lt;-&gt; 2, 3 &lt;-&gt; 4, 
        ...</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.AbstractSystemInfo-class.html">csb.statistics.samplers.mc.neqsteppropagator.AbstractSystemInfo</a></strong>:
      <em class="summary">Subclasses hold all information describing a current system setup 
        (Hamiltonian, boundaries, ...)</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.HamiltonianSysInfo-class.html">csb.statistics.samplers.mc.neqsteppropagator.HamiltonianSysInfo</a></strong>:
      <em class="summary">Holds information describing a system by its Hamiltonian only.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.test.AbstractTestBuilder-class.html">csb.test.AbstractTestBuilder</a></strong>:
      <em class="summary">This is a base class, defining a test loader which exposes the 
        <code>loadTests</code> method.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.test.AnyTestBuilder-class.html">csb.test.AnyTestBuilder</a></strong>:
      <em class="summary">Build a test suite of cases, marked as either unit, functional or 
        regression tests.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.test.CustomTestBuilder-class.html">csb.test.CustomTestBuilder</a></strong>:
      <em class="summary">Build a test suite of cases, marked as custom tests.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.test.FunctionalTestBuilder-class.html">csb.test.FunctionalTestBuilder</a></strong>:
      <em class="summary">Build a test suite of cases, marked as functional tests.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.test.RegressionTestBuilder-class.html">csb.test.RegressionTestBuilder</a></strong>:
      <em class="summary">Build a test suite of cases, marked as regression tests.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.test.UnitTestBuilder-class.html">csb.test.UnitTestBuilder</a></strong>:
      <em class="summary">Build a test suite of cases, marked as unit tests.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statmech.wham.AbstractWHAM-class.html">csb.statmech.wham.AbstractWHAM</a></strong>:
      <em class="summary">Abstract base class</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statmech.wham.NonparametricWHAM-class.html">csb.statmech.wham.NonparametricWHAM</a></strong>:
      <em class="summary">Implementation of the nonparametric WHAM outlined in Habeck 2012, 
        in which histograms are reduced to delta peaks, this allows to use 
        energies samples at different orders of magnitude, improving the 
        accuracy of the DOS estimates.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statmech.wham.WHAM-class.html">csb.statmech.wham.WHAM</a></strong>:
      <em class="summary">Implementation of the original WHAM methods based on histograms.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.alignment.AlignmentResult-class.html">csb.bio.sequence.alignment.AlignmentResult</a></strong>:
      <em class="summary">Represents a pairwise sequence alignment result.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.AppRunner-class.html">csb.apps.AppRunner</a></strong>:
      <em class="summary">A base abstract class for all application runners.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.bfite.AppRunner-class.html">csb.apps.bfite.AppRunner</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhsearch.AppRunner-class.html">csb.apps.hhsearch.AppRunner</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.bfit.AppRunner-class.html">csb.apps.bfit.AppRunner</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm.AppRunner-class.html">csb.apps.buildhmm.AppRunner</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.helloworld.AppRunner-class.html">csb.apps.helloworld.AppRunner</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag.AppRunner-class.html">csb.apps.hhfrag.AppRunner</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.promix.AppRunner-class.html">csb.apps.promix.AppRunner</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.AppRunner-class.html">csb.apps.csfrag.AppRunner</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.precision.AppRunner-class.html">csb.apps.precision.AppRunner</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.embd.AppRunner-class.html">csb.apps.embd.AppRunner</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.Application-class.html">csb.apps.Application</a></strong>:
      <em class="summary">Base CSB application class.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.bfite.BFitApp-class.html">csb.apps.bfite.BFitApp</a></strong>:
      <em class="summary">Python application for robust structure superposition of two 
        protein structures</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.bfit.BFitApp-class.html">csb.apps.bfit.BFitApp</a></strong>:
      <em class="summary">Python application for robust structure superposition of two 
        protein structures</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm.BuildProfileApp-class.html">csb.apps.buildhmm.BuildProfileApp</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.CSfragApp-class.html">csb.apps.csfrag.CSfragApp</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.embd.DeconvolutionApp-class.html">csb.apps.embd.DeconvolutionApp</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag.HHfragApp-class.html">csb.apps.hhfrag.HHfragApp</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhsearch.HHsearchApp-class.html">csb.apps.hhsearch.HHsearchApp</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.helloworld.HelloWorldApp-class.html">csb.apps.helloworld.HelloWorldApp</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.precision.PrecisionApp-class.html">csb.apps.precision.PrecisionApp</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.promix.ProMixApp-class.html">csb.apps.promix.ProMixApp</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.ArgHandler-class.html">csb.apps.ArgHandler</a></strong>:
      <em class="summary">Command line argument handler.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.AssignmentFactory-class.html">csb.bio.fragments.AssignmentFactory</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.AsyncParseResult-class.html">csb.bio.io.wwpdb.AsyncParseResult</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.AsyncStructureParser-class.html">csb.bio.io.wwpdb.AsyncStructureParser</a></strong>:
      <em class="summary">Wraps StructureParser in an asynchronous call.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.AtomConnectivity-class.html">csb.bio.nmr.AtomConnectivity</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.test.Attributes-class.html">csb.test.Attributes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.plots.Backends-class.html">csb.io.plots.Backends</a></strong>:
      <em class="summary">Enumeration of chart backends.</em>
    </li>
    <li> <strong class="uidlink">exceptions.BaseException</strong>:
      <em class="summary">Common base class for all exceptions</em>
    <ul>
    <li> <strong class="uidlink">exceptions.Exception</strong>:
      <em class="summary">Common base class for all non-exit exceptions.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.AppExit-class.html">csb.apps.AppExit</a></strong>:
      <em class="summary">Used to signal an immediate application exit condition (e.g.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm.BuildError-class.html">csb.apps.buildhmm.BuildError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm.BuildArgError-class.html">csb.apps.buildhmm.BuildArgError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm.BuildIOError-class.html">csb.apps.buildhmm.BuildIOError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm.NoOutputError-class.html">csb.apps.buildhmm.NoOutputError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.DuplicateEntryError-class.html">csb.bio.io.clans.DuplicateEntryError</a></strong>:
      <em class="summary">Raised during <a 
        href="csb.bio.io.clans.Clans-class.html#_update_index" class="link"
        onclick="show_private();">Clans._update_index</a> if two entries 
        are identical in name, sequence, and coordinates.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.DuplicateEntryNameError-class.html">csb.bio.io.clans.DuplicateEntryNameError</a></strong>:
      <em class="summary">Raised during <a href="csb.bio.io.clans.Clans-class.html#get_entry"
        class="link">Clans.get_entry</a> if two entries have identical 
        names.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.InvalidOperation-class.html">csb.bio.structure.InvalidOperation</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.numeric.MatrixInitError-class.html">csb.numeric.MatrixInitError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.MissingBlockError-class.html">csb.bio.io.clans.MissingBlockError</a></strong>:
      <em class="summary">Raised if an expected tag is not found during parsing of a CLANS 
        file.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.ProcessError-class.html">csb.io.ProcessError</a></strong>:
      <em class="summary">Raised on <a href="csb.io.Shell-class.html#run" 
        class="link">Shell.run()</a> failures.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.io.TimeoutError-class.html">csb.io.TimeoutError</a></strong>:
      <em class="summary">Raised on <a href="csb.io.Shell-class.html#run" 
        class="link">Shell.run()</a> timeouts.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">exceptions.StandardError</strong>:
      <em class="summary">Base class for all standard Python exceptions that do not represent
        interpreter exiting.</em>
    <ul>
    <li> <strong class="uidlink"><a href="exceptions.AssertionError-class.html">exceptions.AssertionError</a></strong>:
      <em class="summary">Assertion failed.</em>
    </li>
    <li> <strong class="uidlink">exceptions.AttributeError</strong>:
      <em class="summary">Attribute not found.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.core.EnumMemberError-class.html">csb.core.EnumMemberError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.ParameterNotFoundError-class.html">csb.statistics.pdf.ParameterNotFoundError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.UnobservableStateError-class.html">csb.bio.hmm.UnobservableStateError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">exceptions.ImportError</strong>:
      <em class="summary">Import can't find module, or can't find name in module.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.test.InvalidNamespaceError-class.html">csb.test.InvalidNamespaceError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">exceptions.LookupError</strong>:
      <em class="summary">Base class for lookup errors.</em>
    <ul>
    <li> <strong class="uidlink">exceptions.IndexError</strong>:
      <em class="summary">Sequence index out of range.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.core.CollectionIndexError-class.html">csb.core.CollectionIndexError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.EntityIndexError-class.html">csb.bio.structure.EntityIndexError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.LayerIndexError-class.html">csb.bio.hmm.LayerIndexError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.PositionError-class.html">csb.bio.sequence.PositionError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.ColumnPositionError-class.html">csb.bio.sequence.ColumnPositionError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.SequencePositionError-class.html">csb.bio.sequence.SequencePositionError</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">exceptions.KeyError</strong>:
      <em class="summary">Mapping key not found.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.DuplicateSequenceError-class.html">csb.bio.sequence.DuplicateSequenceError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.EmissionExistsError-class.html">csb.bio.hmm.EmissionExistsError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.EntityNotSupportedError-class.html">csb.bio.nmr.EntityNotSupportedError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.InvalidColumnError-class.html">csb.io.tsv.InvalidColumnError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.InvalidKeyError-class.html">csb.core.InvalidKeyError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.core.DuplicateKeyError-class.html">csb.core.DuplicateKeyError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.DuplicateAtomIDError-class.html">csb.bio.structure.DuplicateAtomIDError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.DuplicateChainIDError-class.html">csb.bio.structure.DuplicateChainIDError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.DuplicateModelIDError-class.html">csb.bio.structure.DuplicateModelIDError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.DuplicateResidueIDError-class.html">csb.bio.structure.DuplicateResidueIDError</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.ItemNotFoundError-class.html">csb.core.ItemNotFoundError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.EntityNotFoundError-class.html">csb.bio.structure.EntityNotFoundError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.AtomNotFoundError-class.html">csb.bio.structure.AtomNotFoundError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.ChainNotFoundError-class.html">csb.bio.structure.ChainNotFoundError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.StateNotFoundError-class.html">csb.bio.hmm.StateNotFoundError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.TransitionNotFoundError-class.html">csb.bio.hmm.TransitionNotFoundError</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.alignment.ResidueNotFoundError-class.html">csb.bio.sequence.alignment.ResidueNotFoundError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.SequenceNotFoundError-class.html">csb.bio.sequence.SequenceNotFoundError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.StateExistsError-class.html">csb.bio.hmm.StateExistsError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.StructureNotFoundError-class.html">csb.bio.io.wwpdb.StructureNotFoundError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.TransitionExistsError-class.html">csb.bio.hmm.TransitionExistsError</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">exceptions.NameError</strong>:
      <em class="summary">Name not found globally.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.test.InvalidNamespaceError-class.html">csb.test.InvalidNamespaceError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">exceptions.RuntimeError</strong>:
      <em class="summary">Unspecified run-time error.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ClusterDivergingError-class.html">csb.bio.fragments.ClusterDivergingError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.build.RevisionError-class.html">csb.build.RevisionError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">exceptions.TypeError</strong>:
      <em class="summary">Inappropriate argument type.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.DimensionError-class.html">csb.statistics.samplers.DimensionError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.IncompatibleEstimatorError-class.html">csb.statistics.pdf.IncompatibleEstimatorError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">exceptions.ValueError</strong>:
      <em class="summary">Inappropriate argument value (of correct type).</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.AlignmentArgumentLengthError-class.html">csb.bio.structure.AlignmentArgumentLengthError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.precision.ArgumentError-class.html">csb.apps.precision.ArgumentError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.precision.ArgumentIOError-class.html">csb.apps.precision.ArgumentIOError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.ArgumentError-class.html">csb.apps.csfrag.ArgumentError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.ArgumentIOError-class.html">csb.apps.csfrag.ArgumentIOError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag.ArgumentError-class.html">csb.apps.hhfrag.ArgumentError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag.ArgumentIOError-class.html">csb.apps.hhfrag.ArgumentIOError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Broken3DStructureError-class.html">csb.bio.structure.Broken3DStructureError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.Missing3DStructureError-class.html">csb.bio.structure.Missing3DStructureError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.BrokenSecStructureError-class.html">csb.bio.structure.BrokenSecStructureError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.UnknownSecStructureError-class.html">csb.bio.structure.UnknownSecStructureError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.cs.ChemShiftFormatError-class.html">csb.bio.io.cs.ChemShiftFormatError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ClusterExhaustedError-class.html">csb.bio.fragments.ClusterExhaustedError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ClusterEmptyError-class.html">csb.bio.fragments.ClusterEmptyError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.dssp.DSSPParseError-class.html">csb.bio.io.dssp.DSSPParseError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.mrc.DensityMapFormatError-class.html">csb.bio.io.mrc.DensityMapFormatError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.EnumValueError-class.html">csb.core.EnumValueError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.hhpred.HHProfileFormatError-class.html">csb.bio.io.hhpred.HHProfileFormatError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.hhpred.HHOutputFormatError-class.html">csb.bio.io.hhpred.HHOutputFormatError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.hhpred.StructureFormatError-class.html">csb.bio.io.hhpred.StructureFormatError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.HMMArgumentError-class.html">csb.bio.hmm.HMMArgumentError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.InvalidCommandError-class.html">csb.io.InvalidCommandError</a></strong>:
      <em class="summary">Raised when <a href="csb.io.Shell-class.html#run" 
        class="link">Shell.run()</a> encounters an OSError.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.InvalidOperationError-class.html">csb.apps.csfrag.InvalidOperationError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag.InvalidOperationError-class.html">csb.apps.hhfrag.InvalidOperationError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.InvalidResidueError-class.html">csb.bio.nmr.InvalidResidueError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.PDBParseError-class.html">csb.bio.io.wwpdb.PDBParseError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.HeaderFormatError-class.html">csb.bio.io.wwpdb.HeaderFormatError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.SecStructureFormatError-class.html">csb.bio.io.wwpdb.SecStructureFormatError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.StructureFormatError-class.html">csb.bio.io.wwpdb.StructureFormatError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.InvalidEntryIDError-class.html">csb.bio.io.wwpdb.InvalidEntryIDError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.ResidueMappingError-class.html">csb.bio.io.wwpdb.ResidueMappingError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.UnknownPDBResidueError-class.html">csb.bio.io.wwpdb.UnknownPDBResidueError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.psipred.PSIPredParseError-class.html">csb.bio.io.psipred.PSIPredParseError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.ParameterValueError-class.html">csb.statistics.pdf.ParameterValueError</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.EstimationFailureError-class.html">csb.statistics.pdf.EstimationFailureError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.parameterized.ParameterValueError-class.html">csb.statistics.pdf.parameterized.ParameterValueError</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.parameterized.ParameterizationError-class.html">csb.statistics.pdf.parameterized.ParameterizationError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.noe.PeakListFormatError-class.html">csb.bio.io.noe.PeakListFormatError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.SequenceError-class.html">csb.bio.sequence.SequenceError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.SequenceFormatError-class.html">csb.bio.io.fasta.SequenceFormatError</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.UnknownTagError-class.html">csb.bio.io.clans.UnknownTagError</a></strong>:
      <em class="summary">Raised if an unknown tag is encountered while parsing a CLANS file.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.UnsupportedTypeError-class.html">csb.io.tsv.UnsupportedTypeError</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.noe.BasePeakListReader-class.html">csb.bio.io.noe.BasePeakListReader</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.noe.SparkyPeakListReader-class.html">csb.bio.io.noe.SparkyPeakListReader</a></strong>:
      <em class="summary">Sparky NOE peak list parser.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.noe.XeasyPeakListReader-class.html">csb.bio.io.noe.XeasyPeakListReader</a></strong>:
      <em class="summary">XEASY NOE peak list parser.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.BaseSequenceParser-class.html">csb.bio.io.fasta.BaseSequenceParser</a></strong>:
      <em class="summary">FASTA parser template.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.SequenceParser-class.html">csb.bio.io.fasta.SequenceParser</a></strong>:
      <em class="summary">Standard FASTA parser.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.PDBSequenceParser-class.html">csb.bio.io.fasta.PDBSequenceParser</a></strong>:
      <em class="summary">PDB FASTA parser.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.BenchmarkAdapter-class.html">csb.bio.fragments.BenchmarkAdapter</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.build.BuildTypes-class.html">csb.build.BuildTypes</a></strong>:
      <em class="summary">Enumeration of build types.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.mrc.ByteOrder-class.html">csb.bio.io.mrc.ByteOrder</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.CSfrag-class.html">csb.apps.csfrag.CSfrag</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.plots.Chart-class.html">csb.io.plots.Chart</a></strong>:
      <em class="summary">Simple and clean facade to Matplotlib's plotting API.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ChemShiftAssignmentFactory-class.html">csb.bio.fragments.ChemShiftAssignmentFactory</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.ChemShiftInfo-class.html">csb.bio.nmr.ChemShiftInfo</a></strong>:
      <em class="summary">Chemical shift struct.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.cs.ChemShiftReader-class.html">csb.bio.io.cs.ChemShiftReader</a></strong>:
      <em class="summary">Simple NMR STAR v2 chemical shift reader.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.cs.ChemShift3Reader-class.html">csb.bio.io.cs.ChemShift3Reader</a></strong>:
      <em class="summary">Simple NMR STAR v3 chemical shift reader.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.ChemShiftScoringModel-class.html">csb.bio.nmr.ChemShiftScoringModel</a></strong>:
      <em class="summary">Chemical shift similarity scoring model.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.ChemicalShiftNetwork-class.html">csb.bio.nmr.ChemicalShiftNetwork</a></strong>:
      <em class="summary">Describes a network of covalently connected, chemical shift visible
        nuclei.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.SQLiteRepository.ChunkedReader-class.html">csb.io.tsv.SQLiteRepository.ChunkedReader</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.Clans-class.html">csb.bio.io.clans.Clans</a></strong>:
      <em class="summary">Class for holding and manipulating data from one CLANS file.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.ClansEntry-class.html">csb.bio.io.clans.ClansEntry</a></strong>:
      <em class="summary">Class holding the data of one CLANS sequence entry.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.ClansFileBuilder-class.html">csb.bio.io.clans.ClansFileBuilder</a></strong>:
      <em class="summary">Base abstract files for building a file in CLANS format.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.ClansFileWriter-class.html">csb.bio.io.clans.ClansFileWriter</a></strong>:
      <em class="summary">Class for serializing a <a href="csb.bio.io.clans.Clans-class.html"
        class="link">Clans</a> instance to a file in CLANS format.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.ClansParams-class.html">csb.bio.io.clans.ClansParams</a></strong>:
      <em class="summary">Class for handling <a href="csb.bio.io.clans.Clans-class.html" 
        class="link">Clans</a> parameters.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.ClansParser-class.html">csb.bio.io.clans.ClansParser</a></strong>:
      <em class="summary">CLANS file format aware parser.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.ClansSeqgroup-class.html">csb.bio.io.clans.ClansSeqgroup</a></strong>:
      <em class="summary">Class holding the data of one CLANS group (seqgroup).</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ClusterNode-class.html">csb.bio.fragments.ClusterNode</a></strong>:
      <em class="summary">Cluster node.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ClusterRep-class.html">csb.bio.fragments.ClusterRep</a></strong>:
      <em class="summary">Cluster's representative (centroid) node.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.clans.Color-class.html">csb.bio.io.clans.Color</a></strong>:
      <em class="summary">RGB color handling class.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.ColumnInfo-class.html">csb.bio.sequence.ColumnInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.ColumnInfo-class.html">csb.io.tsv.ColumnInfo</a></strong>:
      <em class="summary">Holder object for column metadata.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.CompositeEntityIterator-class.html">csb.bio.structure.CompositeEntityIterator</a></strong>:
      <em class="summary">Iterates over composite <a 
        href="csb.bio.structure.AbstractEntity-class.html" 
        class="link">AbstractEntity</a> hierarchies.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.structure.ConfinedEntityIterator-class.html">csb.bio.structure.ConfinedEntityIterator</a></strong>:
      <em class="summary">Iterates over composite <a 
        href="csb.bio.structure.AbstractEntity-class.html" 
        class="link">AbstractEntity</a> hierarchies, but terminates the 
        traversal of a branch once a specific node type is encountered.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.test.Config-class.html">csb.test.Config</a></strong>:
      <em class="summary">General CSB Test Config.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.BenchmarkAdapter.Connection-class.html">csb.bio.fragments.BenchmarkAdapter.Connection</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.test.Console-class.html">csb.test.Console</a></strong>:
      <em class="summary">Build and run all tests of the specified namespace and kind.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.build.Console-class.html">csb.build.Console</a></strong>:
      <em class="summary">CSB Build Bot.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.ContactMap-class.html">csb.bio.nmr.ContactMap</a></strong>:
      <em class="summary">Describes a protein contact map.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.ContactMapComparisonInfo-class.html">csb.bio.nmr.ContactMapComparisonInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhsearch.Context-class.html">csb.apps.hhsearch.Context</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.hhsearch.HHTask-class.html">csb.apps.hhsearch.HHTask</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag.SliceContext-class.html">csb.apps.hhfrag.SliceContext</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.Cumulative-class.html">csb.statistics.Cumulative</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.dssp.DSSPParser-class.html">csb.bio.io.dssp.DSSPParser</a></strong>:
      <em class="summary">Simple DSSP Secondary Structure Parser.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.DataRow-class.html">csb.io.tsv.DataRow</a></strong>:
      <em class="summary">Represents a table data row.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.embd.Deconvolution-class.html">csb.apps.embd.Deconvolution</a></strong>:
      <em class="summary">Blind deconvolution for n-dimensional images.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.mrc.DensityInfo-class.html">csb.bio.io.mrc.DensityInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.mrc.DensityMapReader-class.html">csb.bio.io.mrc.DensityMapReader</a></strong>:
      <em class="summary">Binary MRC density map reader.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.mrc.DensityMapWriter-class.html">csb.bio.io.mrc.DensityMapWriter</a></strong>:
      <em class="summary">Binary MRC density map writer.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.EVDParameters-class.html">csb.bio.hmm.EVDParameters</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.EntryID-class.html">csb.bio.io.wwpdb.EntryID</a></strong>:
      <em class="summary">Represents a PDB Chain identifier.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.DegenerateID-class.html">csb.bio.io.wwpdb.DegenerateID</a></strong>:
      <em class="summary">Looks like a <a href="csb.bio.io.wwpdb.StandardID-class.html" 
        class="link">StandardID</a>, except that the accession number may 
        have arbitrary length.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.SeqResID-class.html">csb.bio.io.wwpdb.SeqResID</a></strong>:
      <em class="summary">Same as a <a href="csb.bio.io.wwpdb.StandardID-class.html" 
        class="link">StandardID</a>, but contains an additional underscore 
        between te accession number and the chain identifier.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.StandardID-class.html">csb.bio.io.wwpdb.StandardID</a></strong>:
      <em class="summary">Standard PDB ID in the following form: xxxxY, where xxxx is the 
        accession number (lower case) and Y is an optional chain 
        identifier.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.EntryReader-class.html">csb.io.EntryReader</a></strong>:
      <em class="summary">Generic flatfile reader.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.EntryWriter-class.html">csb.io.EntryWriter</a></strong>:
      <em class="summary">A simple stream writer.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.Enum-class.html">csb.core.Enum</a></strong>:
      <em class="summary">A collection of efficient static methods for working with <a 
        href="csb.core.enum-class.html" class="link">enum</a> classes.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.EnumBase-class.html">csb.core.EnumBase</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.core.enum-class.html">csb.core.enum</a></strong>:
      <em class="summary">Base class for all enumeration types.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.AlignmentFormats-class.html">csb.bio.sequence.AlignmentFormats</a></strong>:
      <em class="summary">Enumeration of multiple sequence alignment formats</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.AlignmentTypes-class.html">csb.bio.sequence.AlignmentTypes</a></strong>:
      <em class="summary">Enumeration of alignment strategies</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.AngleUnits-class.html">csb.bio.structure.AngleUnits</a></strong>:
      <em class="summary">Torsion angle unit types</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.ChemElements-class.html">csb.bio.structure.ChemElements</a></strong>:
      <em class="summary">Periodic table elements</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.NucleicAlphabet-class.html">csb.bio.sequence.NucleicAlphabet</a></strong>:
      <em class="summary">Nucleic sequence alphabet</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.ProteinAlphabet-class.html">csb.bio.sequence.ProteinAlphabet</a></strong>:
      <em class="summary">Protein sequence alphabet</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.ScoreUnits-class.html">csb.bio.hmm.ScoreUnits</a></strong>:
      <em class="summary">Enumeration of HMM emission and transition score units</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.SecStructures-class.html">csb.bio.structure.SecStructures</a></strong>:
      <em class="summary">Secondary structure types</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.SequenceTypes-class.html">csb.bio.sequence.SequenceTypes</a></strong>:
      <em class="summary">Enumeration of sequence types</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.States-class.html">csb.bio.hmm.States</a></strong>:
      <em class="summary">Enumeration of HMM state types</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.StdProteinAlphabet-class.html">csb.bio.sequence.StdProteinAlphabet</a></strong>:
      <em class="summary">Standard protein sequence alphabet</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.UnknownAlphabet-class.html">csb.bio.sequence.UnknownAlphabet</a></strong>:
      <em class="summary">Unknown sequence alphabet</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.EnumItem-class.html">csb.core.EnumItem</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.ars.Envelope-class.html">csb.statistics.ars.Envelope</a></strong>:
      <em class="summary">Envelope function for adaptive rejection sampling.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.ExitCodes-class.html">csb.apps.ExitCodes</a></strong>:
      <em class="summary">Exit code constants.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.bfite.ExitCodes-class.html">csb.apps.bfite.ExitCodes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.promix.ExitCodes-class.html">csb.apps.promix.ExitCodes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.precision.ExitCodes-class.html">csb.apps.precision.ExitCodes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm.ExitCodes-class.html">csb.apps.buildhmm.ExitCodes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.ExitCodes-class.html">csb.apps.csfrag.ExitCodes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag.ExitCodes-class.html">csb.apps.hhfrag.ExitCodes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhsearch.ExitCodes-class.html">csb.apps.hhsearch.ExitCodes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.bfit.ExitCodes-class.html">csb.apps.bfit.ExitCodes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.embd.ExitCodes-class.html">csb.apps.embd.ExitCodes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.helloworld.ExitCodes-class.html">csb.apps.helloworld.ExitCodes</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.Expression-class.html">csb.io.tsv.Expression</a></strong>:
      <em class="summary">Metadata container: represents a table select or update expression.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.FileBuilder-class.html">csb.bio.io.wwpdb.FileBuilder</a></strong>:
      <em class="summary">Base abstract files for all structure file formatters.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.PDBFileBuilder-class.html">csb.bio.io.wwpdb.PDBFileBuilder</a></strong>:
      <em class="summary">PDB file format builder.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.PDBEnsembleFileBuilder-class.html">csb.bio.io.wwpdb.PDBEnsembleFileBuilder</a></strong>:
      <em class="summary">Supports serialization of NMR ensembles.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.Filters-class.html">csb.bio.nmr.Filters</a></strong>:
      <em class="summary">Pre-built atom filters for <a 
        href="csb.bio.nmr.ContactMap-class.html" 
        class="link">ContactMap</a>s.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.FragmentCluster-class.html">csb.bio.fragments.FragmentCluster</a></strong>:
      <em class="summary">Provides clustering/filtering of the fragments, covering a common 
        residue in the target.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.FragmentMatch-class.html">csb.bio.fragments.FragmentMatch</a></strong>:
      <em class="summary">Base class, representing a match between a fragment and its target.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.Assignment-class.html">csb.bio.fragments.Assignment</a></strong>:
      <em class="summary">Represents a match between a fragment and its target.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ChemShiftAssignment-class.html">csb.bio.fragments.ChemShiftAssignment</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.FragmentTypes-class.html">csb.bio.fragments.FragmentTypes</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.mixtures.GaussianMixture-class.html">csb.statistics.mixtures.GaussianMixture</a></strong>:
      <em class="summary">Gaussian mixture model for multi-dimensional data.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.mixtures.AbstractStructureMixture-class.html">csb.statistics.mixtures.AbstractStructureMixture</a></strong>:
      <em class="summary">Abstract mixture model for protein structure ensembles.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.mixtures.ConformerMixture-class.html">csb.statistics.mixtures.ConformerMixture</a></strong>:
      <em class="summary">Gaussian mixture model for protein structure ensembles using a set 
        of conformers</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.mixtures.SegmentMixture-class.html">csb.statistics.mixtures.SegmentMixture</a></strong>:
      <em class="summary">Gaussian mixture model for protein structure ensembles using a set 
        of segments</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.precision.GlobalInfo-class.html">csb.apps.precision.GlobalInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.hhpred.HHMFileBuilder-class.html">csb.bio.io.hhpred.HHMFileBuilder</a></strong>:
      <em class="summary">Builder for HHpred's hhm files.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.hhpred.HHOutputParser-class.html">csb.bio.io.hhpred.HHOutputParser</a></strong>:
      <em class="summary">Parser for HHsearch result (*.hhr) files.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.hhpred.HHProfileParser-class.html">csb.bio.io.hhpred.HHProfileParser</a></strong>:
      <em class="summary">A class that is HHpred HMM format aware.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag.HHfrag-class.html">csb.apps.hhfrag.HHfrag</a></strong>:
      <em class="summary">The HHfrag dynamic fragment detection protocol.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.HHpredHit-class.html">csb.bio.hmm.HHpredHit</a></strong>:
      <em class="summary">Represents a single HHsearch hit.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.HHpredHitList-class.html">csb.bio.hmm.HHpredHitList</a></strong>:
      <em class="summary">Represents a collection of <a 
        href="csb.bio.hmm.HHpredHit-class.html" 
        class="link">HHpredHit</a>s.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhsearch.HHsearch-class.html">csb.apps.hhsearch.HHsearch</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.mrc.HeaderInfo-class.html">csb.bio.io.mrc.HeaderInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.InterpolationFactory-class.html">csb.statistics.samplers.mc.multichain.InterpolationFactory</a></strong>:
      <em class="summary">Produces interpolations for functions changed during 
        non-equilibrium trajectories.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.numeric.InvertibleMatrix-class.html">csb.numeric.InvertibleMatrix</a></strong>:
      <em class="summary">Matrix object which is intended to save time in MCMC sampling 
        algorithms involving repeated integration of Hamiltonian equations 
        of motion and frequent draws from multivariate normal distributions
        involving mass matrices as covariance matrices.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.Label-class.html">csb.bio.nmr.Label</a></strong>:
      <em class="summary">Utility class for working with chemical shift labels.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.precision.LibrarySuperimposer-class.html">csb.apps.precision.LibrarySuperimposer</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.ars.LogProb-class.html">csb.statistics.ars.LogProb</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.ARSPosteriorAlpha-class.html">csb.statistics.scalemixture.ARSPosteriorAlpha</a></strong>:
      <em class="summary">This class represents the posterior distribution of the alpha 
        parameter of the Gamma and Inverse Gamma prior, and allows sampling
        using adaptive rejection sampling <a 
        href="csb.statistics.ars.ARS-class.html" class="link">ARS</a>.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.ars.Gauss-class.html">csb.statistics.ars.Gauss</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.propagators.Looper-class.html">csb.statistics.samplers.mc.propagators.Looper</a></strong>:
      <em class="summary">Implements an iterable list with a ring-like topology, that is, if 
        the iterator points on the last element, next() returns the first 
        element.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.MDParam-class.html">csb.statistics.samplers.mc.neqsteppropagator.MDParam</a></strong>:
      <em class="summary">Holds all required information for calculating a MD trajectory.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.HMCPropagationParam-class.html">csb.statistics.samplers.mc.neqsteppropagator.HMCPropagationParam</a></strong>:
      <em class="summary">Holds all required information for propagating a system by HMC.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.MDPropagationParam-class.html">csb.statistics.samplers.mc.neqsteppropagator.MDPropagationParam</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.PlainMDPropagationParam-class.html">csb.statistics.samplers.mc.neqsteppropagator.PlainMDPropagationParam</a></strong>:
      <em class="summary">Holds all required information for propagating a system by MD.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.MatchInfo-class.html">csb.apps.csfrag.MatchInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.MatchTable-class.html">csb.apps.csfrag.MatchTable</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.maxent.MaxentPosterior-class.html">csb.statistics.maxent.MaxentPosterior</a></strong>:
      <em class="summary">Object to hold and calculate the posterior (log)probability given 
        an exponential family model and corresponding data.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.Metrics-class.html">csb.bio.fragments.Metrics</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.NOEPeak-class.html">csb.bio.nmr.NOEPeak</a></strong>:
      <em class="summary">Describes a single NOE peak.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.NOESpectrum-class.html">csb.bio.nmr.NOESpectrum</a></strong>:
      <em class="summary">Describes an NOE spectrum.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhsearch.HHsearch.Options-class.html">csb.apps.hhsearch.HHsearch.Options</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.OutputBuilder-class.html">csb.bio.io.fasta.OutputBuilder</a></strong>:
      <em class="summary">Base sequence/alignment string format builder.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.A3MOutputBuilder-class.html">csb.bio.io.fasta.A3MOutputBuilder</a></strong>:
      <em class="summary">Formats sequences as A3M strings.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.FASTAOutputBuilder-class.html">csb.bio.io.fasta.FASTAOutputBuilder</a></strong>:
      <em class="summary">Formats sequences as standard FASTA strings.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.PIROutputBuilder-class.html">csb.bio.io.fasta.PIROutputBuilder</a></strong>:
      <em class="summary">Formats sequences as PIR FASTA strings, recognized by Modeller.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.rosetta.OutputBuilder-class.html">csb.bio.fragments.rosetta.OutputBuilder</a></strong>:
      <em class="summary">Rosetta fragment file formatter.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.rosetta.ExtendedOutputBuilder-class.html">csb.bio.fragments.rosetta.ExtendedOutputBuilder</a></strong>:
      <em class="summary">Builds non-standard fragment files, which contain the CA 
        coordinates of each residue at the end of each line.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.embd.OutputPathBuilder-class.html">csb.apps.embd.OutputPathBuilder</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.psipred.PSIPredParser-class.html">csb.bio.io.psipred.PSIPredParser</a></strong>:
      <em class="summary">Simple PSI-PRED Secondary Structure Parser.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.psipred.PSIPredResidueInfo-class.html">csb.bio.io.psipred.PSIPredResidueInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.plots.PlotsCollection-class.html">csb.io.plots.PlotsCollection</a></strong>:
      <em class="summary">A list-like collection of all plots in the chart (0-based).</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.Predicate-class.html">csb.io.tsv.Predicate</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.io.tsv.Between-class.html">csb.io.tsv.Between</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.Equals-class.html">csb.io.tsv.Equals</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.Greater-class.html">csb.io.tsv.Greater</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.GreaterOrEquals-class.html">csb.io.tsv.GreaterOrEquals</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.In-class.html">csb.io.tsv.In</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.Lower-class.html">csb.io.tsv.Lower</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.LowerOrEquals-class.html">csb.io.tsv.LowerOrEquals</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.NotEquals-class.html">csb.io.tsv.NotEquals</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.NotIn-class.html">csb.io.tsv.NotIn</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag.PredictionBuilder-class.html">csb.apps.hhfrag.PredictionBuilder</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm.ProfileBuilder-class.html">csb.apps.buildhmm.ProfileBuilder</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm.FASTAProfileBuilder-class.html">csb.apps.buildhmm.FASTAProfileBuilder</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm.PDBProfileBuilder-class.html">csb.apps.buildhmm.PDBProfileBuilder</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.ProfileHMM-class.html">csb.bio.hmm.ProfileHMM</a></strong>:
      <em class="summary">Describes a protein profile Hidden Markov Model.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.ProfileHMMRegion-class.html">csb.bio.hmm.ProfileHMMRegion</a></strong>:
      <em class="summary">A shallow proxy referring to a sub-region of a given Profile HMM.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.ProfileHMMSegment-class.html">csb.bio.hmm.ProfileHMMSegment</a></strong>:
      <em class="summary">Represents a segment (fragment) of a ProfileHMM.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.EmissionProfileSegment-class.html">csb.bio.hmm.EmissionProfileSegment</a></strong>:
      <em class="summary">Represents a segment of the Match state emission probabilities of a
        <a href="csb.bio.hmm.ProfileHMM-class.html" 
        class="link">ProfileHMM</a>.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.ProfileLength-class.html">csb.bio.hmm.ProfileLength</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.Protocol-class.html">csb.statistics.samplers.mc.neqsteppropagator.Protocol</a></strong>:
      <em class="summary">Describes a stepwise protocol as in Nilmeier et al.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.Proxy-class.html">csb.core.Proxy</a></strong>:
      <em class="summary">Base class implementing the proxy pattern.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.io.AutoFlushStream-class.html">csb.io.AutoFlushStream</a></strong>:
      <em class="summary">Wrapper around a buffered stream which automatically calls flush() 
        after each write().</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.TempFile-class.html">csb.io.TempFile</a></strong>:
      <em class="summary">Create a temporary file and take care of deleting it upon object 
        destruction.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.pseudocounts.PseudocountBuilder-class.html">csb.bio.hmm.pseudocounts.PseudocountBuilder</a></strong>:
      <em class="summary">Constructs profile HMMs with pseudocounts.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.REMatchProxy-class.html">csb.core.REMatchProxy</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.RENSTrajInfo-class.html">csb.statistics.samplers.mc.multichain.RENSTrajInfo</a></strong>:
      <em class="summary">Holds information necessary for calculating a RENS trajectory.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr.RandomCoil-class.html">csb.bio.nmr.RandomCoil</a></strong>:
      <em class="summary">Utility class containing all necessary data and methods for 
        computing secondary chemical shifts.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.ReducedHamiltonian-class.html">csb.statistics.samplers.mc.neqsteppropagator.ReducedHamiltonian</a></strong>:
      <em class="summary">Describes a reduced Hamiltonian (Hamiltonian, its position gradient
        and the system temperature)</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.ReplicaHistory-class.html">csb.statistics.samplers.mc.multichain.ReplicaHistory</a></strong>:
      <em class="summary">Replica history object, works with both RE and RENS for the 
        AlternatingAdjacentSwapScheme.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.RepositoryImp-class.html">csb.io.tsv.RepositoryImp</a></strong>:
      <em class="summary">Abstract SQL backend interface.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.io.tsv.SQLiteRepository-class.html">csb.io.tsv.SQLiteRepository</a></strong>:
      <em class="summary">SQLite-based concrete repository implementor.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.dssp.ResidueAssignmentInfo-class.html">csb.bio.io.dssp.ResidueAssignmentInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ResidueAssignmentInfo-class.html">csb.bio.fragments.ResidueAssignmentInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ResidueEventInfo-class.html">csb.bio.fragments.ResidueEventInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.ResidueInfo-class.html">csb.bio.sequence.ResidueInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.ResidueInfo-class.html">csb.bio.io.wwpdb.ResidueInfo</a></strong>:
      <em class="summary">High-performance struct, which functions as a container for 
        unmapped <a href="csb.bio.structure.Atom-class.html" 
        class="link">Atom</a>s.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.rosetta.ResidueInfo-class.html">csb.bio.fragments.rosetta.ResidueInfo</a></strong>:
      <em class="summary">Container struct for a single rosetta fragment residue.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.build.RevisionHandler-class.html">csb.build.RevisionHandler</a></strong>:
      <em class="summary">Determines the current repository revision number of a working 
        copy.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.build.MercurialHandler-class.html">csb.build.MercurialHandler</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.build.SubversionHandler-class.html">csb.build.SubversionHandler</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.build.RevisionInfo-class.html">csb.build.RevisionInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.rosetta.RosettaFragment-class.html">csb.bio.fragments.rosetta.RosettaFragment</a></strong>:
      <em class="summary">Represents a single Rosetta fragment match.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.RosettaFragsetFactory-class.html">csb.bio.fragments.RosettaFragsetFactory</a></strong>:
      <em class="summary">Simplifies the construction of fragment libraries.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.svg.SSCartoonBuilder-class.html">csb.bio.io.svg.SSCartoonBuilder</a></strong>:
      <em class="summary">Creates 2D vector diagrams from <a 
        href="csb.bio.structure.SecondaryStructure-class.html" 
        class="link">SecondaryStructure</a> objects.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.ScaleMixturePrior-class.html">csb.statistics.scalemixture.ScaleMixturePrior</a></strong>:
      <em class="summary">Prior on the scales of a <a 
        href="csb.statistics.scalemixture.ScaleMixture-class.html" 
        class="link">ScaleMixture</a>, which determines how the scales are 
        estimated.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.GammaPrior-class.html">csb.statistics.scalemixture.GammaPrior</a></strong>:
      <em class="summary">Gamma prior on mixture weights including a weak gamma prior on its 
        parameter.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture.InvGammaPrior-class.html">csb.statistics.scalemixture.InvGammaPrior</a></strong>:
      <em class="summary">Inverse Gamma prior on mixture weights including a weak gamma prior
        on its parameter.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.ScoringHelper-class.html">csb.apps.csfrag.ScoringHelper</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhsearch.SecStructureScoring-class.html">csb.apps.hhsearch.SecStructureScoring</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.SecondaryShiftConverter-class.html">csb.apps.csfrag.SecondaryShiftConverter</a></strong>:
      <em class="summary">Helper, which reads assigned shifts from NMR STAR files and 
        calculates corrected secondary shifts.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag.SecondaryShiftReader-class.html">csb.apps.csfrag.SecondaryShiftReader</a></strong>:
      <em class="summary">Reads secondary shifts from files in CSfrag format.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.SecondaryStructureElement-class.html">csb.bio.structure.SecondaryStructureElement</a></strong>:
      <em class="summary">Describes a Secondary Structure Element.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.SequenceAdapter-class.html">csb.bio.sequence.SequenceAdapter</a></strong>:
      <em class="summary">Base wrapper class for <a 
        href="csb.bio.sequence.AbstractSequence-class.html" 
        class="link">AbstractSequence</a> objects.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.AlignedSequenceAdapter-class.html">csb.bio.sequence.AlignedSequenceAdapter</a></strong>:
      <em class="summary">Adapter, which wraps a gapped <a 
        href="csb.bio.sequence.AbstractSequence-class.html" 
        class="link">AbstractSequence</a> object and makes it compatible 
        with the MSA row/entry interface, expected by <a 
        href="csb.bio.sequence.AbstractAlignment-class.html" 
        class="link">AbstractAlignment</a>.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.SequenceAlignmentReader-class.html">csb.bio.io.fasta.SequenceAlignmentReader</a></strong>:
      <em class="summary">Sequence alignment parser.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.SequenceAlphabets-class.html">csb.bio.sequence.SequenceAlphabets</a></strong>:
      <em class="summary">Sequence alphabet enumerations.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.SequenceIndexer-class.html">csb.bio.sequence.SequenceIndexer</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.ColumnIndexer-class.html">csb.bio.sequence.ColumnIndexer</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.UngappedSequenceIndexer-class.html">csb.bio.sequence.UngappedSequenceIndexer</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.Shell-class.html">csb.io.Shell</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.ShellInfo-class.html">csb.io.ShellInfo</a></strong>:
      <em class="summary">Shell command execution info</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.vasco.ShiftInfo-class.html">csb.bio.io.vasco.ShiftInfo</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.singlechain.SimpleProposalCommunicator-class.html">csb.statistics.samplers.mc.singlechain.SimpleProposalCommunicator</a></strong>:
      <em class="summary">This holds all the information needed to calculate the acceptance 
        probability in both the <a 
        href="csb.statistics.samplers.mc.singlechain.RWMCSampler-class.html"
        class="link">RWMCSampler</a> and <a 
        href="csb.statistics.samplers.mc.singlechain.HMCSampler-class.html"
        class="link">HMCSampler</a> classes, that is, only the proposal 
        state.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.singlechain.NCMCProposalCommunicator-class.html">csb.statistics.samplers.mc.singlechain.NCMCProposalCommunicator</a></strong>:
      <em class="summary">Holds all information (that is, the trajectory with heat, work, 
        Hamiltonian difference and jacbian) needed to calculate the 
        acceptance probability in the AbstractNCMCSampler class.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.SingleSwapStatistics-class.html">csb.statistics.samplers.mc.multichain.SingleSwapStatistics</a></strong>:
      <em class="summary">Tracks swap statistics of a single sampler pair.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.SliceHelper-class.html">csb.bio.sequence.SliceHelper</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.State-class.html">csb.bio.hmm.State</a></strong>:
      <em class="summary">Describes a Hidden Markov Model state.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.StateFactory-class.html">csb.bio.hmm.StateFactory</a></strong>:
      <em class="summary">Simplifies the construction of protein profile HMM states.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statmech.ensembles.StatisticalEnsemble-class.html">csb.statmech.ensembles.StatisticalEnsemble</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statmech.ensembles.BoltzmannEnsemble-class.html">csb.statmech.ensembles.BoltzmannEnsemble</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statmech.ensembles.FermiEnsemble-class.html">csb.statmech.ensembles.FermiEnsemble</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statmech.ensembles.CompositeEnsemble-class.html">csb.statmech.ensembles.CompositeEnsemble</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.statmech.ensembles.TsallisEnsemble-class.html">csb.statmech.ensembles.TsallisEnsemble</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator.Step-class.html">csb.statistics.samplers.mc.neqsteppropagator.Step</a></strong>:
      <em class="summary">Defines a step in an NCMC-like stepwise protocol.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.Step-class.html">csb.io.tsv.Step</a></strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.io.tsv.Operator-class.html">csb.io.tsv.Operator</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.Where-class.html">csb.io.tsv.Where</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.dssp.StrideParser-class.html">csb.bio.io.dssp.StrideParser</a></strong>:
      <em class="summary">Simple STRIDE Secondary Structure Parser.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.fasta.StructureAlignmentFactory-class.html">csb.bio.io.fasta.StructureAlignmentFactory</a></strong>:
      <em class="summary">Protein structure alignment parser.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.StructureProvider-class.html">csb.bio.io.wwpdb.StructureProvider</a></strong>:
      <em class="summary">Base class for all PDB data source providers.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.CustomStructureProvider-class.html">csb.bio.io.wwpdb.CustomStructureProvider</a></strong>:
      <em class="summary">A custom PDB data source.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.FileSystemStructureProvider-class.html">csb.bio.io.wwpdb.FileSystemStructureProvider</a></strong>:
      <em class="summary">Simple file system based PDB data source.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.wwpdb.RemoteStructureProvider-class.html">csb.bio.io.wwpdb.RemoteStructureProvider</a></strong>:
      <em class="summary">Retrieves PDB structures from a specified remote URL.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.SuperimposeInfo-class.html">csb.bio.structure.SuperimposeInfo</a></strong>:
      <em class="summary">Describes a structural alignment result.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain.SwapStatistics-class.html">csb.statistics.samplers.mc.multichain.SwapStatistics</a></strong>:
      <em class="summary">Tracks swap statistics for an AbstractExchangeMC subclass instance.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv.Table-class.html">csb.io.tsv.Table</a></strong>:
      <em class="summary">Build and query a TSV Table.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.TargetResidue-class.html">csb.bio.fragments.TargetResidue</a></strong>:
      <em class="summary">Wrapper around <a href="csb.bio.fragments.Target-class.html" 
        class="link">Target</a>'s native residues.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.ChemShiftTargetResidue-class.html">csb.bio.fragments.ChemShiftTargetResidue</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.io.TempFolder-class.html">csb.io.TempFolder</a></strong>:
      <em class="summary">Create a temporary directory which is automatically wiped when the 
        object is closed.</em>
    </li>
    <li> <strong class="uidlink">unittest.case.TestCase</strong>:
      <em class="summary">A class whose instances are single test cases.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.test.Case-class.html">csb.test.Case</a></strong>:
      <em class="summary">Base class, defining a CSB Test Case.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.TorsionAnglePredictor-class.html">csb.bio.fragments.TorsionAnglePredictor</a></strong>:
      <em class="summary">Fragment-based phi/psi angles predictor.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure.TorsionAngles-class.html">csb.bio.structure.TorsionAngles</a></strong>:
      <em class="summary">Describes a collection of phi, psi and omega backbone torsion 
        angles.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.TorsionPredictionInfo-class.html">csb.bio.fragments.TorsionPredictionInfo</a></strong>:
      <em class="summary">Struct container for a single torsion angle prediction.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.TrajectoryBuilder-class.html">csb.statistics.samplers.mc.TrajectoryBuilder</a></strong>:
      <em class="summary">Allows to  build a Trajectory object step by step.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.ShortTrajectoryBuilder-class.html">csb.statistics.samplers.mc.ShortTrajectoryBuilder</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.Transition-class.html">csb.bio.hmm.Transition</a></strong>:
      <em class="summary">Describes a Hidden Markov Model transition between two states.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.TransitionType-class.html">csb.bio.hmm.TransitionType</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.ArgHandler.Type-class.html">csb.apps.ArgHandler.Type</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.apps.embd.Util-class.html">csb.apps.embd.Util</a></strong>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.vasco.VascoStructureParser-class.html">csb.bio.io.vasco.VascoStructureParser</a></strong>:
      <em class="summary">Simple Vasco Parser</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.Version-class.html">csb.Version</a></strong>:
      <em class="summary">CSB version number.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.whatif.WhatCheckParser-class.html">csb.bio.io.whatif.WhatCheckParser</a></strong>:
      <em class="summary">Simple WhatIf/WhatCheck Summary parser</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.noe.XeasyFileBuilder-class.html">csb.bio.io.noe.XeasyFileBuilder</a></strong>:
      <em class="summary">XEASY output format builder.</em>
    </li>
    <li> <strong class="uidlink">threading._Verbose</strong>
    <ul>
    <li> <strong class="uidlink">threading.Thread</strong>
    <ul>
    <li> <strong class="uidlink"><a href="csb.io.plots.Backend-class.html">csb.io.plots.Backend</a></strong>:
      <em class="summary">Abstract class defining the behavior of all Chart GUI backends.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.io.plots.WxBackendImpl-class.html">csb.io.plots.WxBackendImpl</a></strong>:
      <em class="summary">WxPython <a href="csb.io.plots.Backend-class.html" 
        class="link">Backend</a> implementor.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.InterruptibleThread-class.html">csb.core.InterruptibleThread</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">dict</strong>:
      <em class="summary">dict() -&gt; new empty dictionary
dict(mapping) -&gt; new dictionary initialized from a mapping object's
    (key, value) pairs
dict(iterable) -&gt; new dictionary initialized as if via:
    d = {}
    for k, v in iterable:
        d[k] = v
dict(**kwargs) -&gt; new dictionary initialized with the name=value pairs
    in the keyword argument list.</em>
    <ul>
    <li> <strong class="uidlink">collections.OrderedDict</strong>:
      <em class="summary">Dictionary that remembers insertion order</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.core.OrderedDict-class.html">csb.core.OrderedDict</a></strong>
    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">list</strong>:
      <em class="summary">list() -&gt; new empty list list(iterable) -&gt; new list 
        initialized from iterable's items</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.core.Stack-class.html">csb.core.Stack</a></strong>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">property</strong>:
      <em class="summary">property(fget=None, fset=None, fdel=None, doc=None) -&gt; property attribute</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.core.typedproperty-class.html">csb.core.typedproperty</a></strong>:
      <em class="summary">Property decorator for convenient creation of typed, encapsulated 
        fields.</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.validatedproperty-class.html">csb.core.validatedproperty</a></strong>:
      <em class="summary">Property decorator with predefined getter/setter and value checking
        or casting in the setter.</em>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink">type</strong>:
      <em class="summary">type(object) -&gt; the object's type type(name, bases, dict) -&gt; 
        a new type</em>
    <ul>
    <li> <strong class="uidlink"><a href="abc.ABCMeta-class.html">abc.ABCMeta</a></strong>:
      <em class="summary">Metaclass for defining Abstract Base Classes (ABCs).</em>
    </li>
    <li> <strong class="uidlink"><a href="csb.core.EnumMeta-class.html">csb.core.EnumMeta</a></strong>:
      <em class="summary">Metaclass for enum types.</em>
    </li>
    </ul>
    </li>
    </ul>
    </li>
</ul>
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="csb-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th bgcolor="#70b0f0" class="navbar-select"
          >&nbsp;&nbsp;&nbsp;Trees&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >CSB v1.2.2.614</th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Fri Dec 27 22:25:16 2013
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
