<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>csb.bio.io.wwpdb</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="csb-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >CSB v1.2.2.614</th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="csb-module.html">Package&nbsp;csb</a> ::
        <a href="csb.bio-module.html">Package&nbsp;bio</a> ::
        <a href="csb.bio.io-module.html">Package&nbsp;io</a> ::
        Module&nbsp;wwpdb
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="csb.bio.io.wwpdb-module.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== MODULE DESCRIPTION ==================== -->
<h1 class="epydoc">Module wwpdb</h1><p class="nomargin-top"><span class="codelink"><a href="csb.bio.io.wwpdb-pysrc.html">source&nbsp;code</a></span></p>
<p>PDB structure parsers, format builders and database providers.</p>
  <p>The most basic usage is:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span>parser = StructureParser(<span class="py-string">'structure.pdb'</span>)
<span class="py-prompt">&gt;&gt;&gt; </span>parser.parse_structure()
<span class="py-output">&lt;Structure&gt;     # a Structure object (model)</span></pre>
  <p>or if this is an NMR ensemble:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span>parser.parse_models()
<span class="py-output">&lt;Ensemble&gt;      # an Ensemble object (collection of alternative Structure-s)</span></pre>
  <p>This module introduces a family of PDB file parsers. The common 
  interface of all parsers is defined in <a 
  href="csb.bio.io.wwpdb.AbstractStructureParser-class.html" 
  class="link">AbstractStructureParser</a>. This class has several 
  implementations:</p>
  <ul>
    <li>
      <a href="csb.bio.io.wwpdb.RegularStructureParser-class.html" 
      class="link">RegularStructureParser</a> - handles normal PDB files 
      with SEQRES fields
    </li>
    <li>
      <a href="csb.bio.io.wwpdb.LegacyStructureParser-class.html" 
      class="link">LegacyStructureParser</a> - reads structures from legacy
      or malformed PDB files, which are lacking SEQRES records (initializes
      all residues from the ATOMs instead)
    </li>
    <li>
      <a href="csb.bio.io.wwpdb.PDBHeaderParser-class.html" 
      class="link">PDBHeaderParser</a> - reads only the headers of the PDB 
      files and produces structures without coordinates. Useful for reading
      metadata (e.g. accession numbers or just plain SEQRES sequences) with
      minimum overhead
    </li>
  </ul>
  <p>Unless you have a special reason, you should use the <a 
  href="csb.bio.io-module.html#StructureParser" 
  class="link">StructureParser</a> factory, which returns a proper <a 
  href="csb.bio.io.wwpdb.AbstractStructureParser-class.html" 
  class="link">AbstractStructureParser</a> implementation, depending on the
  input PDB file. If the input file looks like a regular PDB file, the 
  factory returns a <a 
  href="csb.bio.io.wwpdb.RegularStructureParser-class.html" 
  class="link">RegularStructureParser</a>, otherwise it instantiates <a 
  href="csb.bio.io.wwpdb.LegacyStructureParser-class.html" 
  class="link">LegacyStructureParser</a>. <a 
  href="csb.bio.io-module.html#StructureParser" 
  class="link">StructureParser</a> is in fact an alias for <a 
  href="csb.bio.io.wwpdb.AbstractStructureParser-class.html#create_parser" 
  class="link">AbstractStructureParser.create_parser</a>.</p>
  <p>Writing your own, customized PDB parser is easy. Suppose that you are 
  trying to parse a PDB-like file which misuses the charge column to store 
  custom info. This will certainly crash <a 
  href="csb.bio.io.wwpdb.RegularStructureParser-class.html" 
  class="link">RegularStructureParser</a> (for good), but you can create 
  your own parser as a workaround. All you need to to is to override the 
  virtual <code>_read_charge</code> hook method:</p>
<pre class="literalblock">
   class CustomParser(RegularStructureParser):
   
       def _read_charge(self, line):
           try:
               return super(CustomParser, self)._read_charge(line)
           except StructureFormatError:
               return None
</pre>
  <p>Another important abstraction in this module is <a 
  href="csb.bio.io.wwpdb.StructureProvider-class.html" 
  class="link">StructureProvider</a>. It has several implementations which 
  can be used to retrieve PDB <a 
  href="csb.bio.structure.Structure-class.html" class="link">Structure</a>s
  from various sources: file system directories, remote URLs, etc. You can 
  easily create your own provider as well. See <a 
  href="csb.bio.io.wwpdb.StructureProvider-class.html" 
  class="link">StructureProvider</a> for details.</p>
  <p>Finally, this module gives you some <a 
  href="csb.bio.io.wwpdb.FileBuilder-class.html" 
  class="link">FileBuilder</a>s, used for text serialization of <a 
  href="csb.bio.structure.Structure-class.html" class="link">Structure</a>s
  and <a href="csb.bio.structure.Ensemble-class.html" 
  class="link">Ensemble</a>s:</p>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span>builder = PDBFileBuilder(stream)
<span class="py-prompt">&gt;&gt;&gt; </span>builder.add_header(structure)
<span class="py-prompt">&gt;&gt;&gt; </span>builder.add_structure(structure)</pre>
  <p>where stream is any Python stream, e.g. an open file or 
  sys.stdout.</p>
  <p>See <a href="csb.bio.structure.Ensemble-class.html" 
  class="link">Ensemble</a> and <a 
  href="csb.bio.structure.Structure-class.html" class="link">Structure</a> 
  from <a href="csb.bio.structure-module.html" 
  class="link">csb.bio.structure</a> for details on these objects.</p>

<!-- ==================== CLASSES ==================== -->
<a name="section-Classes"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td align="left" colspan="2" class="table-header">
    <span class="table-header">Classes</span></td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.AbstractResidueMapper-class.html" class="summary-name">AbstractResidueMapper</a><br />
      Defines the base interface of all residue mappers, used to align 
        PDB ATOM records to the real (SEQRES) sequence of a chain.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.AbstractStructureParser-class.html" class="summary-name">AbstractStructureParser</a><br />
      A base PDB structure format-aware parser.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.AsyncParseResult-class.html" class="summary-name">AsyncParseResult</a>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.AsyncStructureParser-class.html" class="summary-name">AsyncStructureParser</a><br />
      Wraps StructureParser in an asynchronous call.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.CombinedResidueMapper-class.html" class="summary-name">CombinedResidueMapper</a><br />
      The best of both worlds: attempts to map the residues using <a 
        href="csb.bio.io.wwpdb.FastResidueMapper-class.html" 
        class="link">FastResidueMapper</a>, but upon failure secures 
        success by switching to <a 
        href="csb.bio.io.wwpdb.RobustResidueMapper-class.html" 
        class="link">RobustResidueMapper</a>.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.CustomStructureProvider-class.html" class="summary-name">CustomStructureProvider</a><br />
      A custom PDB data source.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.DegenerateID-class.html" class="summary-name">DegenerateID</a><br />
      Looks like a <a href="csb.bio.io.wwpdb.StandardID-class.html" 
        class="link">StandardID</a>, except that the accession number may 
        have arbitrary length.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.EntryID-class.html" class="summary-name">EntryID</a><br />
      Represents a PDB Chain identifier.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.FastResidueMapper-class.html" class="summary-name">FastResidueMapper</a><br />
      RegExp-based residue mapper.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.FileBuilder-class.html" class="summary-name">FileBuilder</a><br />
      Base abstract files for all structure file formatters.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.FileSystemStructureProvider-class.html" class="summary-name">FileSystemStructureProvider</a><br />
      Simple file system based PDB data source.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.HeaderFormatError-class.html" class="summary-name">HeaderFormatError</a>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.InvalidEntryIDError-class.html" class="summary-name">InvalidEntryIDError</a>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.LegacyStructureParser-class.html" class="summary-name">LegacyStructureParser</a><br />
      This is a customized PDB parser, which is designed to read both 
        sequence and atom data from the ATOM section.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.PDBEnsembleFileBuilder-class.html" class="summary-name">PDBEnsembleFileBuilder</a><br />
      Supports serialization of NMR ensembles.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.PDBFileBuilder-class.html" class="summary-name">PDBFileBuilder</a><br />
      PDB file format builder.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.PDBHeaderParser-class.html" class="summary-name">PDBHeaderParser</a><br />
      Ultra fast PDB HEADER parser.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.PDBParseError-class.html" class="summary-name">PDBParseError</a>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.RegularStructureParser-class.html" class="summary-name">RegularStructureParser</a><br />
      This is the de facto PDB parser, which is designed to read SEQRES 
        and ATOM sections separately, and them map them.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.RemoteStructureProvider-class.html" class="summary-name">RemoteStructureProvider</a><br />
      Retrieves PDB structures from a specified remote URL.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.ResidueInfo-class.html" class="summary-name">ResidueInfo</a><br />
      High-performance struct, which functions as a container for 
        unmapped <a href="csb.bio.structure.Atom-class.html" 
        class="link">Atom</a>s.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.ResidueMappingError-class.html" class="summary-name">ResidueMappingError</a>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.RobustResidueMapper-class.html" class="summary-name">RobustResidueMapper</a><br />
      Exhaustive residue mapper, which uses Needleman-Wunsch global 
        alignment.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.SecStructureFormatError-class.html" class="summary-name">SecStructureFormatError</a>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.SeqResID-class.html" class="summary-name">SeqResID</a><br />
      Same as a <a href="csb.bio.io.wwpdb.StandardID-class.html" 
        class="link">StandardID</a>, but contains an additional underscore 
        between te accession number and the chain identifier.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.SparseChainSequence-class.html" class="summary-name">SparseChainSequence</a><br />
      Sequence view for reference (SEQRES) or sparse (ATOM) PDB chains.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.StandardID-class.html" class="summary-name">StandardID</a><br />
      Standard PDB ID in the following form: xxxxY, where xxxx is the 
        accession number (lower case) and Y is an optional chain 
        identifier.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.StructureFormatError-class.html" class="summary-name">StructureFormatError</a>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.StructureNotFoundError-class.html" class="summary-name">StructureNotFoundError</a>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.StructureProvider-class.html" class="summary-name">StructureProvider</a><br />
      Base class for all PDB data source providers.
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb.UnknownPDBResidueError-class.html" class="summary-name">UnknownPDBResidueError</a>
    </td>
  </tr>
</table>
<!-- ==================== FUNCTIONS ==================== -->
<a name="section-Functions"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td align="left" colspan="2" class="table-header">
    <span class="table-header">Functions</span></td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type"><a href="csb.bio.io.wwpdb.AbstractStructureParser-class.html" 
      class="link">AbstractStructureParser</a></span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="csb.bio.io.wwpdb-module.html#StructureParser" class="summary-sig-name">StructureParser</a>(<span class="summary-sig-arg">structure_file</span>,
        <span class="summary-sig-arg">check_ss</span>=<span class="summary-sig-default">False</span>,
        <span class="summary-sig-arg">mapper</span>=<span class="summary-sig-default">None</span>)</span><br />
      A StructureParser factory, which instantiates and returns the proper 
      parser object based on the contents of the PDB file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="csb.bio.io.wwpdb-pysrc.html">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">str</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="csb.bio.io.wwpdb-module.html#find" class="summary-sig-name">find</a>(<span class="summary-sig-arg">id</span>,
        <span class="summary-sig-arg">paths</span>)</span><br />
      Try to discover a PDB file for PDB <code>id</code> in 
      <code>paths</code>.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="csb.bio.io.wwpdb-pysrc.html#find">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type"><a href="csb.bio.structure.Structure-class.html" 
      class="link">Structure</a></span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="csb.bio.io.wwpdb-module.html#get" class="summary-sig-name">get</a>(<span class="summary-sig-arg">accession</span>,
        <span class="summary-sig-arg">model</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">prefix</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">http://www.rcsb.org/pdb/files/pdb</code><code class="variable-quote">'</code></span>)</span><br />
      Download and parse a PDB entry.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="csb.bio.io.wwpdb-pysrc.html#get">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
</table>
<!-- ==================== VARIABLES ==================== -->
<a name="section-Variables"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td align="left" colspan="2" class="table-header">
    <span class="table-header">Variables</span></td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb-module.html#PDB_AMINOACIDS" class="summary-name">PDB_AMINOACIDS</a> = <code title="{'2AS': 'ASP',
 '3AH': 'HIS',
 '5HP': 'GLU',
 'ACL': 'ARG',
 'AGM': 'ARG',
 'AIB': 'ALA',
 'ALA': 'ALA',
 'ALM': 'ALA',
..."><code class="variable-group">{</code><code class="variable-quote">'</code><code class="variable-string">2AS</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">ASP</code><code class="variable-quote">'</code><code class="variable-op">, </code><code class="variable-quote">'</code><code class="variable-string">3AH</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">HIS</code><code class="variable-quote">'</code><code class="variable-op">, </code><code class="variable-quote">'</code><code class="variable-string">5HP</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">GLU</code><code class="variable-quote">'</code><code class="variable-op">, </code><code class="variable-quote">'</code><code class="variable-string">A</code><code class="variable-ellipsis">...</code></code>
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
        <a href="csb.bio.io.wwpdb-module.html#PDB_NUCLEOTIDES" class="summary-name">PDB_NUCLEOTIDES</a> = <code title="{'  M': 'Amino',
 'A': 'Adenine',
 'B': 'NotA',
 'C': 'Cytosine',
 'D': 'NotC',
 'DA': 'Adenine',
 'DC': 'Cytosine',
 'DG': 'Guanine',
..."><code class="variable-group">{</code><code class="variable-quote">'</code><code class="variable-string">  M</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">Amino</code><code class="variable-quote">'</code><code class="variable-op">, </code><code class="variable-quote">'</code><code class="variable-string">A</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">Adenine</code><code class="variable-quote">'</code><code class="variable-op">, </code><code class="variable-quote">'</code><code class="variable-string">B</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">NotA</code><code class="variable-quote">'</code><code class="variable-ellipsis">...</code></code>
    </td>
  </tr>
</table>
<!-- ==================== FUNCTION DETAILS ==================== -->
<a name="section-FunctionDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td align="left" colspan="2" class="table-header">
    <span class="table-header">Function Details</span></td>
</tr>
</table>
<a name="StructureParser"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">StructureParser</span>(<span class="sig-arg">structure_file</span>,
        <span class="sig-arg">check_ss</span>=<span class="sig-default">False</span>,
        <span class="sig-arg">mapper</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="csb.bio.io.wwpdb-pysrc.html">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>A StructureParser factory, which instantiates and returns the proper 
  parser object based on the contents of the PDB file.</p>
  <p>If the file contains a SEQRES section, <a 
  href="csb.bio.io.wwpdb.RegularStructureParser-class.html" 
  class="link">RegularStructureParser</a> is returned, otherwise <a 
  href="csb.bio.io.wwpdb.LegacyStructureParser-class.html" 
  class="link">LegacyStructureParser</a> is instantiated. In the latter 
  case LegacyStructureParser will read the sequence data directly from the 
  ATOMs.</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>structure_file</code></strong> (str) - the PDB file to parse</li>
        <li><strong class="pname"><code>check_ss</code></strong> (bool) - if True, secondary structure errors in the file will cause <a 
          href="csb.bio.io.wwpdb.SecStructureFormatError-class.html" 
          class="link">SecStructureFormatError</a> exceptions</li>
        <li><strong class="pname"><code>mapper</code></strong> (<a href="csb.bio.io.wwpdb.AbstractResidueMapper-class.html" 
          class="link">AbstractResidueMapper</a>) - residue mapper, used to align ATOM records to SEQRES. If None, 
          use the default (<a 
          href="csb.bio.io.wwpdb.CombinedResidueMapper-class.html" 
          class="link">CombinedResidueMapper</a>)</li>
    </ul></dd>
    <dt>Returns: <a href="csb.bio.io.wwpdb.AbstractStructureParser-class.html" 
      class="link">AbstractStructureParser</a></dt>
  </dl>
</td></tr></table>
</div>
<a name="find"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">find</span>(<span class="sig-arg">id</span>,
        <span class="sig-arg">paths</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="csb.bio.io.wwpdb-pysrc.html#find">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Try to discover a PDB file for PDB <code>id</code> in 
  <code>paths</code>.</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>id</code></strong> (str) - PDB ID of the entry</li>
        <li><strong class="pname"><code>paths</code></strong> (list of str) - a list of directories to scan</li>
    </ul></dd>
    <dt>Returns: str</dt>
        <dd>path and file name on success, None otherwise</dd>
  </dl>
</td></tr></table>
</div>
<a name="get"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get</span>(<span class="sig-arg">accession</span>,
        <span class="sig-arg">model</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">prefix</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">http://www.rcsb.org/pdb/files/pdb</code><code class="variable-quote">'</code></span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="csb.bio.io.wwpdb-pysrc.html#get">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Download and parse a PDB entry.</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>accession</code></strong> (str) - accession number of the entry</li>
        <li><strong class="pname"><code>model</code></strong> (str) - model identifier</li>
        <li><strong class="pname"><code>prefix</code></strong> (str) - download URL prefix</li>
    </ul></dd>
    <dt>Returns: <a href="csb.bio.structure.Structure-class.html" 
      class="link">Structure</a></dt>
        <dd>object representation of the selected model</dd>
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== VARIABLES DETAILS ==================== -->
<a name="section-VariablesDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td align="left" colspan="2" class="table-header">
    <span class="table-header">Variables Details</span></td>
</tr>
</table>
<a name="PDB_AMINOACIDS"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <h3 class="epydoc">PDB_AMINOACIDS</h3>
  
  <dl class="fields">
  </dl>
  <dl class="fields">
    <dt>Value:</dt>
      <dd><table><tr><td><pre class="variable">
<code class="variable-group">{</code><code class="variable-quote">'</code><code class="variable-string">2AS</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">ASP</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">3AH</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">HIS</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">5HP</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">GLU</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">ACL</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">ARG</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">AGM</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">ARG</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">AIB</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">ALA</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">ALA</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">ALA</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">ALM</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">ALA</code><code class="variable-quote">'</code><code class="variable-op">,</code>
<code class="variable-ellipsis">...</code>
</pre></td></tr></table>
</dd>
  </dl>
</td></tr></table>
</div>
<a name="PDB_NUCLEOTIDES"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <h3 class="epydoc">PDB_NUCLEOTIDES</h3>
  
  <dl class="fields">
  </dl>
  <dl class="fields">
    <dt>Value:</dt>
      <dd><table><tr><td><pre class="variable">
<code class="variable-group">{</code><code class="variable-quote">'</code><code class="variable-string">  M</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">Amino</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">A</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">Adenine</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">B</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">NotA</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">C</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">Cytosine</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">D</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">NotC</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">DA</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">Adenine</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">DC</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">Cytosine</code><code class="variable-quote">'</code><code class="variable-op">,</code>
 <code class="variable-quote">'</code><code class="variable-string">DG</code><code class="variable-quote">'</code><code class="variable-op">: </code><code class="variable-quote">'</code><code class="variable-string">Guanine</code><code class="variable-quote">'</code><code class="variable-op">,</code>
<code class="variable-ellipsis">...</code>
</pre></td></tr></table>
</dd>
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="csb-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >CSB v1.2.2.614</th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Fri Dec 27 22:25:17 2013
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
