<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>Module Hierarchy</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="csb-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th bgcolor="#70b0f0" class="navbar-select"
          >&nbsp;&nbsp;&nbsp;Trees&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >CSB v1.2.2.614</th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">&nbsp;</td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="module-tree.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<center><b>
 [ <a href="module-tree.html">Module Hierarchy</a>
 | <a href="class-tree.html">Class Hierarchy</a> ]
</b></center><br />
<h1 class="epydoc">Module Hierarchy</h1>
<ul class="nomargin-top">
    <li> <strong class="uidlink"><a href="csb-module.html">csb</a></strong>: <em class="summary">CSB is a high-level, object-oriented library used to solve problems
        in the field of Computational Structural Biology.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps-module.html">csb.apps</a></strong>: <em class="summary">Root package for all executable CSB client programs.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.apps.bfit-module.html">csb.apps.bfit</a></strong>: <em class="summary">Python application for robust structure superposition of two 
        structures.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.bfite-module.html">csb.apps.bfite</a></strong>: <em class="summary">Python application for robust structure superposition of an 
        ensemble of structures.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.buildhmm-module.html">csb.apps.buildhmm</a></strong>: <em class="summary">Build an HMM from a FASTA sequence.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.csfrag-module.html">csb.apps.csfrag</a></strong>: <em class="summary">CSfrag: build a dynamic library of analogous fragments, given a 
        list of assigned chemical shifts.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.embd-module.html">csb.apps.embd</a></strong>: <em class="summary">Sharpening of EM maps by non-negative blind deconvolution.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.helloworld-module.html">csb.apps.helloworld</a></strong>: <em class="summary">This is a CSB HelloWorld dummy application.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhfrag-module.html">csb.apps.hhfrag</a></strong>: <em class="summary">HHfrag: build a dynamic variable-length fragment library for 
        protein structure prediction with Rosetta AbInitio.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.hhsearch-module.html">csb.apps.hhsearch</a></strong>: <em class="summary">Python bindings for the HHsearch program.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.precision-module.html">csb.apps.precision</a></strong>: <em class="summary">Measure the precision and coverage of a fragment library stored in 
        Rosetta NNmake format.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.apps.promix-module.html">csb.apps.promix</a></strong>: <em class="summary">ProMix: Take a protein structure ensemble and find a mixture of 
        rigid segments or a mixture of conformers.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio-module.html">csb.bio</a></strong>: <em class="summary">Core bioinformatics abstractions and I/O.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments-module.html">csb.bio.fragments</a></strong>: <em class="summary">APIs for working with protein structure fragments and libraries.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.fragments.rosetta-module.html">csb.bio.fragments.rosetta</a></strong>: <em class="summary">Rosetta fragment libraries.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.hmm-module.html">csb.bio.hmm</a></strong>: <em class="summary">HHpred and Hidden Markov Model APIs.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.hmm.pseudocounts-module.html">csb.bio.hmm.pseudocounts</a></strong>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io-module.html">csb.bio.io</a></strong>: <em class="summary">I/O for various biological file formats.</em>
    <ul>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.clans-module.html" onclick="show_private();">csb.bio.io.clans</a></strong>: <em class="summary">Classes for parsing/manipulating/writing CLANS (by Tancred Frickey)
        files</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.cs-module.html" onclick="show_private();">csb.bio.io.cs</a></strong>: <em class="summary">Simple NMR STAR chemical shift readers.</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.dssp-module.html" onclick="show_private();">csb.bio.io.dssp</a></strong>: <em class="summary">DSSP Parser</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.fasta-module.html" onclick="show_private();">csb.bio.io.fasta</a></strong>: <em class="summary">FASTA format sequence I/O.</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.hhpred-module.html" onclick="show_private();">csb.bio.io.hhpred</a></strong>: <em class="summary">HHpred-related format parsers.</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.mrc-module.html" onclick="show_private();">csb.bio.io.mrc</a></strong>: <em class="summary">Cryo-EM density map I/O</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.noe-module.html">csb.bio.io.noe</a></strong>: <em class="summary">Simple XEASY and Sparky peak list parsers.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.procheck-module.html">csb.bio.io.procheck</a></strong>: <em class="summary">Procheck parser</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.psipred-module.html">csb.bio.io.psipred</a></strong>: <em class="summary">PSIPRED Parser</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.svg-module.html">csb.bio.io.svg</a></strong>: <em class="summary">Build SVG diagrams from various csb objects.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.vasco-module.html">csb.bio.io.vasco</a></strong>: <em class="summary">VASCO Chemical Shift format parser.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.io.whatif-module.html">csb.bio.io.whatif</a></strong>: <em class="summary">Simple WhatIf/WhatCheck Summary parser</em>    </li>
    <li class="private"> <strong class="uidlink"><a href="csb.bio.io.wwpdb-module.html" onclick="show_private();">csb.bio.io.wwpdb</a></strong>: <em class="summary">PDB structure parsers, format builders and database providers.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.nmr-module.html">csb.bio.nmr</a></strong>: <em class="summary">NMR related objects.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.sequence-module.html">csb.bio.sequence</a></strong>: <em class="summary">Sequence and sequence alignment APIs.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.bio.sequence.alignment-module.html">csb.bio.sequence.alignment</a></strong>: <em class="summary">Collection of sequence alignment algorithms.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.bio.structure-module.html">csb.bio.structure</a></strong>: <em class="summary">3D and secondary structure APIs.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.bio.utils-module.html">csb.bio.utils</a></strong>: <em class="summary">Computational utility functions.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.build-module.html">csb.build</a></strong>: <em class="summary">CSB build related tools and programs.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.core-module.html">csb.core</a></strong>: <em class="summary">Generic containers, data structures and language extensions.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.io-module.html">csb.io</a></strong>: <em class="summary">Generic I/O utility objects.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.io.plots-module.html">csb.io.plots</a></strong>: <em class="summary">Plotting facilities, based on Python's MPL library.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.io.tsv-module.html">csb.io.tsv</a></strong>: <em class="summary">Read, query and update textual tables via flexible SQL interface.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.numeric-module.html">csb.numeric</a></strong>: <em class="summary">Low level numeric / math utility functions.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.numeric.integrators-module.html">csb.numeric.integrators</a></strong>: <em class="summary">provides various integration schemes and an abstract gradient 
        class.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics-module.html">csb.statistics</a></strong>: <em class="summary">Statistics root package.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.ars-module.html">csb.statistics.ars</a></strong>: <em class="summary">Adaptive Rejection Sampling (ARS)</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.maxent-module.html">csb.statistics.maxent</a></strong>: <em class="summary">A Maximum-Entropy model for backbone torsion angles.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.mixtures-module.html">csb.statistics.mixtures</a></strong>: <em class="summary">Mixture models for multi-dimensional data.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf-module.html">csb.statistics.pdf</a></strong>: <em class="summary">Probability density functions.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.pdf.parameterized-module.html">csb.statistics.pdf.parameterized</a></strong>: <em class="summary">Probability density functions with support for shared and computed 
        parameters.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.rand-module.html">csb.statistics.rand</a></strong>: <em class="summary">Random number generators</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers-module.html">csb.statistics.samplers</a></strong>: <em class="summary">Defines abstract samplers.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc-module.html">csb.statistics.samplers.mc</a></strong>: <em class="summary">Abstract Monte Carlo samplers.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.multichain-module.html">csb.statistics.samplers.mc.multichain</a></strong>: <em class="summary">Implements several extended-ensemble Monte Carlo sampling 
        algorithms.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.neqsteppropagator-module.html">csb.statistics.samplers.mc.neqsteppropagator</a></strong>: <em class="summary">Propagator class employing stepwise trajectories as used in the 
        NCMC algorithm (Nilmeier et al., &quot;Nonequilibrium candidate 
        Monte Carlo is an efficient tool for equilibrium simulation&quot;, 
        PNAS 2011)</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.propagators-module.html">csb.statistics.samplers.mc.propagators</a></strong>: <em class="summary">Provides various deterministic and stochastic propagators.</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.samplers.mc.singlechain-module.html">csb.statistics.samplers.mc.singlechain</a></strong>: <em class="summary">Various Monte Carlo equilibrium sampling algorithms, which simulate
        only one Markov chain.</em>    </li>
    </ul>
    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statistics.scalemixture-module.html">csb.statistics.scalemixture</a></strong>: <em class="summary">Approximation of a distribution as a mixture of gaussians with a 
        zero mean but different sigmas</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.statmech-module.html">csb.statmech</a></strong>: <em class="summary">Methods for statistical mechanics</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.statmech.ensembles-module.html">csb.statmech.ensembles</a></strong>: <em class="summary">Statistical Ensembles</em>    </li>
    <li> <strong class="uidlink"><a href="csb.statmech.wham-module.html">csb.statmech.wham</a></strong>: <em class="summary">Estimate the free energy and density of states from tempered 
        ensembles using histogram re-weighting.</em>    </li>
    </ul>
    </li>
    <li> <strong class="uidlink"><a href="csb.test-module.html">csb.test</a></strong>: <em class="summary">This is a top level package, hosting the entire CSB test framework.</em>
    <ul>
    <li> <strong class="uidlink"><a href="csb.test.app-module.html">csb.test.app</a></strong>: <em class="summary">CSB Test Runner app.</em>    </li>
    </ul>
    </li>
    </ul>
    </li>
</ul>
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="csb-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th bgcolor="#70b0f0" class="navbar-select"
          >&nbsp;&nbsp;&nbsp;Trees&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >CSB v1.2.2.614</th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Fri Dec 27 22:25:15 2013
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
