
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>PGLCNV</title>
<meta http-equiv="Content-Language" content="English" />
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<link rel="stylesheet" type="text/css" href="style.css" media="screen" />
</head>
<body>

<div id="wrap">
<div id="top"></div>
<div id="content">

<div class="header">
<img src="leaf.png">
<br><br><br><br><br>
</div>	

<div class="breadcrumbs"><h1>PGLCNV</h1></div><div class="middle"><h2>Project Summary</h2><div style="width: 90%"><b>Description:</b> This is a real world application of Leaf. It is the protocol for a CNV study in a rare disease. Details
can't be disclosed since the study is still unpublished at the moment of publishing its protocol as a Leaf
example. For the same reason, <b>links to output files are intentionally not active!</b>.<br><b>Number of nodes:</b> 40<br><b>Number of <i>[F]</i>-nodes:</b> 21<br><b>Total number of output files:</b> 122<br><b>Total size of output files:</b> 3.48G<br><b>Total CPU time required:</b> 01:03:31.05<br></div><br><h2>Protocol map</h2><div align = "center"><IMG SRC="PGLCNV.gif" USEMAP="#G" />
</div><map id="G" name="G">
<area shape="rect" id="exportComRegGen" href="#exportComRegGen_details" title="&lt;TABLE&gt;" alt="" coords="1435,913,1500,923"/>
<area shape="rect" id="exportComRegGen" href="CompareRegGenes.txt" title="Open output file" alt="" coords="1502,913,1567,923"/>
<area shape="rect" id="exportMergedRegs" href="#exportMergedRegs_details" title="&lt;TABLE&gt;" alt="" coords="1618,757,1670,766"/>
<area shape="rect" id="exportMergedRegs" href="mergedRegsTB.tsv" title="Open output file" alt="" coords="1672,757,1723,766"/>
<area shape="rect" id="exportIntersMat_genes" href="#exportIntersMat_genes_details" title="&lt;TABLE&gt;" alt="" coords="1778,402,1841,411"/>
<area shape="rect" id="exportIntersMat_genes" href="verbose_inters_mat.txt" title="Open output file" alt="" coords="1843,402,1905,411"/>
<area shape="rect" id="joinPennCNVout" href="#joinPennCNVout_details" title="&lt;TABLE&gt;" alt="" coords="475,582,526,591"/>
<area shape="rect" id="joinPennCNVout" href="pglfeb.csv_allNO14.out" title="Open output file" alt="" coords="527,582,578,591"/>
<area shape="rect" id="addGeneInfo" href="#addGeneInfo_details" title="&lt;TABLE&gt;" alt="" coords="631,633,673,642"/>
<area shape="rect" id="addGeneInfo" href="pglfeb.csv_allNO14.out.genes" title="Open output file" alt="" coords="675,633,717,642"/>
<area shape="rect" id="makeBed" href="#makeBed_details" title="&lt;TABLE&gt;" alt="" coords="621,563,673,572"/>
<area shape="rect" id="makeBed" href="pglfeb.csv.bed" title="Open output file" alt="" coords="675,563,726,572"/>
<area shape="rect" id="exportPerRegion" href="#exportPerRegion_details" title="&lt;TABLE&gt;" alt="" coords="1273,768,1319,777"/>
<area shape="rect" id="exportPerRegion" href="perRegion.csv" title="Open output file" alt="" coords="1321,768,1367,777"/>
<area shape="rect" id="exportToCircos" href="#exportToCircos_details" title="&lt;TABLE&gt;" alt="" coords="1448,773,1500,782"/>
<area shape="rect" id="readFile" href="#readFile_details" title="&lt;TABLE&gt;" alt="" coords="770,633,830,642"/>
<area shape="rect" id="sampleSheet" href="#sampleSheet_details" title="&lt;TABLE&gt;" alt="" coords="17,515,68,524"/>
<area shape="rect" id="sampleSheet" href="sample_sheet.tsv" title="Open output file" alt="" coords="70,515,120,524"/>
<area shape="rect" id="getSampleNames" href="#getSampleNames_details" title="&lt;TABLE&gt;" alt="" coords="1271,505,1319,514"/>
<area shape="rect" id="prepareInput" href="#prepareInput_details" title="&lt;TABLE&gt;" alt="" coords="169,584,219,593"/>
<area shape="rect" id="prepareInput" href="#prepareInput_details" title="Multiple outputs: click to go to details" alt="" coords="221,584,271,593"/>
<area shape="rect" id="reformat2" href="#reformat2_details" title="&lt;TABLE&gt;" alt="" coords="1095,774,1150,783"/>
<area shape="rect" id="manualClean" href="#manualClean_details" title="&lt;TABLE&gt;" alt="" coords="935,605,993,614"/>
<area shape="rect" id="distMatGfx" href="#distMatGfx_details" title="&lt;TABLE&gt;" alt="" coords="1782,493,1841,502"/>
<area shape="rect" id="distMatGfx" href="#distMatGfx_details" title="Multiple outputs: click to go to details" alt="" coords="1843,493,1901,502"/>
<area shape="rect" id="intersectTBRegs" href="#intersectTBRegs_details" title="&lt;TABLE&gt;" alt="" coords="1448,632,1500,641"/>
<area shape="rect" id="exportMergedRegsTB" href="#exportMergedRegsTB_details" title="&lt;TABLE&gt;" alt="" coords="1611,627,1670,636"/>
<area shape="rect" id="exportMergedRegsTB" href="mergedRegsTB.tsv" title="Open output file" alt="" coords="1672,627,1730,636"/>
<area shape="rect" id="mergeFragments" href="#mergeFragments_details" title="&lt;TABLE&gt;" alt="" coords="1610,697,1670,706"/>
<area shape="rect" id="intersectRegs" href="#intersectRegs_details" title="&lt;TABLE&gt;" alt="" coords="1446,702,1500,712"/>
<area shape="rect" id="countCNVlenghts" href="#countCNVlenghts_details" title="&lt;TABLE&gt;" alt="" coords="1435,843,1500,852"/>
<area shape="rect" id="genoTypeCheck" href="#genoTypeCheck_details" title="&lt;TABLE&gt;" alt="" coords="314,657,372,666"/>
<area shape="rect" id="getGeneNames" href="#getGeneNames_details" title="&lt;TABLE&gt;" alt="" coords="1273,215,1319,224"/>
<area shape="rect" id="intersMat_genes" href="#intersMat_genes_details" title="&lt;TABLE&gt;" alt="" coords="1612,407,1670,416"/>
<area shape="rect" id="clustergram" href="#clustergram_details" title="&lt;TABLE&gt;" alt="" coords="1783,72,1841,81"/>
<area shape="rect" id="clustergram" href="#clustergram_details" title="Multiple outputs: click to go to details" alt="" coords="1843,72,1900,81"/>
<area shape="rect" id="exportCNVDiffMat" href="#exportCNVDiffMat_details" title="&lt;TABLE&gt;" alt="" coords="1790,224,1841,233"/>
<area shape="rect" id="exportCNVDiffMat" href="CNVgenes.csv" title="Open output file" alt="" coords="1843,224,1893,233"/>
<area shape="rect" id="CNVDiffMat" href="#CNVDiffMat_details" title="&lt;TABLE&gt;" alt="" coords="1441,151,1500,160"/>
<area shape="rect" id="exportCNVlenghts" href="#exportCNVlenghts_details" title="&lt;TABLE&gt;" alt="" coords="1619,902,1670,911"/>
<area shape="rect" id="exportCNVlenghts" href="cnv_lenghts.tsv" title="Open output file" alt="" coords="1672,902,1722,911"/>
<area shape="rect" id="exportFigCNVlenghts" href="#exportFigCNVlenghts_details" title="&lt;TABLE&gt;" alt="" coords="1612,842,1670,851"/>
<area shape="rect" id="exportFigCNVlenghts" href="#exportFigCNVlenghts_details" title="Multiple outputs: click to go to details" alt="" coords="1672,842,1729,851"/>
<area shape="rect" id="addStats" href="#addStats_details" title="&lt;TABLE&gt;" alt="" coords="1269,311,1319,320"/>
<area shape="rect" id="CNVboolVects" href="#CNVboolVects_details" title="&lt;TABLE&gt;" alt="" coords="1439,432,1500,441"/>
<area shape="rect" id="exportDistMats" href="#exportDistMats_details" title="&lt;TABLE&gt;" alt="" coords="1782,579,1841,589"/>
<area shape="rect" id="exportDistMats" href="#exportDistMats_details" title="Multiple outputs: click to go to details" alt="" coords="1843,579,1902,589"/>
<area shape="rect" id="reformat" href="#reformat_details" title="&lt;TABLE&gt;" alt="" coords="1095,605,1150,614"/>
<area shape="rect" id="CompareRegionAndGenes" href="#CompareRegionAndGenes_details" title="&lt;TABLE&gt;" alt="" coords="1248,675,1319,684"/>
<area shape="rect" id="exportPerRegion_clean" href="#exportPerRegion_clean_details" title="&lt;TABLE&gt;" alt="" coords="1257,605,1319,614"/>
<area shape="rect" id="exportPerRegion_clean" href="perRegion_clean.csv" title="Open output file" alt="" coords="1321,605,1383,614"/>
<area shape="rect" id="samplesDistMats" href="#samplesDistMats_details" title="&lt;TABLE&gt;" alt="" coords="1614,538,1670,547"/>
<area shape="rect" id="LogR_BAF_FileName" href="#LogR_BAF_FileName_details" title="&lt;TABLE&gt;" alt="" coords="11,595,68,604"/>
<area shape="rect" id="LogR_BAF_FileName" href="pglfeb.csv" title="Open output file" alt="" coords="70,595,125,604"/>
<area shape="rect" id="PennCNV" href="#PennCNV_details" title="&lt;TABLE&gt;" alt="" coords="316,577,372,586"/>
<area shape="rect" id="PennCNV" href="#PennCNV_details" title="Multiple outputs: click to go to details" alt="" coords="374,577,430,586"/>
<area shape="rect" id="exportFigExtRegs" href="#exportFigExtRegs_details" title="&lt;TABLE&gt;" alt="" coords="1792,678,1841,687"/>
<area shape="rect" id="exportFigExtRegs" href="extRegs.pdf" title="Open output file" alt="" coords="1843,678,1891,687"/>
<area shape="rect" id="exportMergedFragments" href="#exportMergedFragments_details" title="&lt;TABLE&gt;" alt="" coords="1774,748,1841,757"/>
<area shape="rect" id="exportMergedFragments" href="extRegions.tsv" title="Open output file" alt="" coords="1843,748,1909,757"/>
<area shape="rect" id="computeFisher" href="#computeFisher_details" title="&lt;TABLE&gt;" alt="" coords="1617,151,1670,160"/>
</map>
<h2>Nodes details<h2>
<h3><a name="exportComRegGen_details">exportComRegGen</a></h3>
Exports CompareRegionAndGenes output to tsv format<br>Output file: <a href="CompareRegGenes.txt">CompareRegGenes.txt</a><br><br><b>Last build was:</b> Sun Jan  8 03:47:10 2012.<br><b>Required CPU time:</b> 00:00:0.08.<br><div class="code"><pre>def exportComRegGen(data):<br>    """Exports CompareRegionAndGenes output to tsv format"""<br>    f = open('CompareRegGenes.txt','w')<br><br>    f.write('Gene\tSample\ts/t\tReg. start\tReg. end\tReg. size\tTanscript. start\t Transcript. end\tTranscript. size\tCode start\t Code end\tCode size\tInters. to Transcr. %\t Inters. to Code %')<br>    <br>    for line in data:<br>        if line[0]!='NOT_FOUND':<br>            for item in line:<br>                if type(item)==float:<br>                    <br>                    f.write(str(item)+'\t')<br>            f.write('\n')<br>    return 'CompareRegGenes.txt'<br></pre></div><hr><h3><a name="exportMergedRegs_details">exportMergedRegs</a></h3>
Exports to tsv file<br>Output file: <a href="mergedRegsTB.tsv">mergedRegsTB.tsv</a><br><br><b>Last build was:</b> Sun Jan  8 03:47:16 2012.<br><b>Required CPU time:</b> 00:00:0.13.<br><div class="code"><pre>def exportMergedRegs(mergedRegs):<br>    """Exports to tsv file"""<br>    f = open('mergedRegsTB.tsv', 'w')<br><br>    f.write('B/T\tChr\tReg. Start\tReg. End\tNum Samples\tCN\t"Genes"\n')<br><br>    for key in sorted(mergedRegs.keys()):<br>        for i in range(0, len(mergedRegs[key][0])):<br>            try:<br>                f.write(('B' if key[0] == 's' else 'T') +'\t'+<br>                        key[1]+'\t'+<br>                        str(mergedRegs[key][0][i]).strip('[]').replace(', ','\t')+'\t'+<br>                        str(mergedRegs[key][1][i])+'\t'+<br>                        str(mergedRegs[key][3][i]).<br>                        replace('[','').<br>                        replace(']','').<br>                        replace("'", '')+'\t'+<br>                        str(mergedRegs[key][2][i]).replace('set','').<br>                        replace(')','').<br>                        replace('(','').<br>                        replace('[','').<br>                        replace(']','').<br>                        replace("'", '')+'\n'<br>                        )<br>            except: import pdb; pdb.set_trace()<br><br>    return 'mergedRegsTB.tsv'<br></pre></div><hr><h3><a name="exportIntersMat_genes_details">exportIntersMat_genes</a></h3>
Exports the output of intersMat_genes to csv format<br>Output file: <a href="verbose_inters_mat.txt">verbose_inters_mat.txt</a><br><br><b>Last build was:</b> Sun Jan  8 03:47:20 2012.<br><b>Required CPU time:</b> 00:00:0.08.<br><div class="code"><pre>def exportIntersMat_genes(genes, samples, inters_names):<br>    """Exports the output of intersMat_genes to csv format"""<br>    f = open('verbose_inters_mat.txt', 'w')<br>    f.write(' ,'+','.join(map(lambda x:x[0]+x[1], samples)) + '\n')<br>    for idx1 in range(0,len(inters_names)):<br>        f.write(samples[idx1][0]+samples[idx1][1]+',')<br>        for idx2 in range(0,len(inters_names)):<br>            f.write(str(list(inters_names[idx1][idx2])).strip('[]').replace('\'','').replace(',','') + ',')<br>        f.write('\n')<br>    return 'verbose_inters_mat.txt'<br></pre></div><hr><h3><a name="joinPennCNVout_details">joinPennCNVout</a></h3>
Puts together the different ouptuts (1 per sample) of PennCNV<br>Output file: <a href="pglfeb.csv_allNO14.out">pglfeb.csv_allNO14.out</a><br><br><b>Last build was:</b> Sun Jan  8 03:45:59 2012.<br><b>Required CPU time:</b> 00:00:0.11.<br><div class="code"><pre>def joinPennCNVout(inputfiles):<br>    """Puts together the different ouptuts (1 per sample) of PennCNV"""<br>    #log_ofile = open(fname + '_all.log', 'a')<br>    if os.path.exists(fname + '_allNO14.out'):<br>        os.remove(fname+'_allNO14.out')<br>    out_ofile = open(fname + '_allNO14.out', 'a')<br>    for ifile in inputfiles:<br>        #log_ifname = ifile + '.log'<br>        #log_ofile.write(open(log_ifname, 'r').read())<br>        #if not '14B' in ifile and not '5B' in ifile and not '34' in ifile and not '33' in ifile:<br>        #if not '14' in ifile and not '21' in ifile and not '14B' in ifile and not '5B' in ifile:<br>        out_ofile.write(open(ifile, 'r').read())<br>    out_ofile.close()<br>    return fname + '_allNO14.out'<br></pre></div><hr><h3><a name="exportPerRegion_details">exportPerRegion</a></h3>
Exports to tsv file<br>Output file: <a href="perRegion.csv">perRegion.csv</a><br><br><b>Last build was:</b> Sun Jan  8 03:51:58 2012.<br><b>Required CPU time:</b> 00:00:0.26.<br><div class="code"><pre>def exportPerRegion(data):<br>    """Exports to tsv file"""<br>    f = open('perRegion.csv','w')<br>    sdict = idToNameDict()<br>    f.write(data[0].replace('\t',',')+'\n')<br><br>    for line in data[1:]:<br>        l = line.replace(',','.').split()<br>        l[6] = sdict[l[6]]<br>        l = str(l).replace('\'', '').strip('[]')<br>        f.write(l+'\n')<br>    return 'perRegion.csv'<br></pre></div><hr><h3><a name="addGeneInfo_details">addGeneInfo</a></h3>
Annotate regions with intersecting genes<br>Output file: <a href="pglfeb.csv_allNO14.out.genes">pglfeb.csv_allNO14.out.genes</a><br><br><b>Last build was:</b> Sun Jan  8 03:46:51 2012.<br><b>Required CPU time:</b> 00:00:51.95.<br><div class="code"><pre>def addGeneInfo(fname):<br>    """Annotate regions with intersecting genes"""<br>    #print 'SKIPPING'<br>    #return fname + '.genes'<br>    os.system('perl ' + penncnv_path + '/scan_region.pl "' + fname + '" -refgene ' + annot_path + '/refGene.txt -reflink ' + annot_path + '/refLink.txt > "' + fname + '.genes1"')<br>    os.system('perl ' + penncnv_path + '/scan_region.pl "' + fname + '.genes1" -refgene ' + annot_path + '/refGene.txt -reflink ' + annot_path + '/refLink.txt -expandleft 5m > "' + fname + '.genes2"')<br>    os.system('perl ' + penncnv_path + '/scan_region.pl "' + fname + '.genes2" -refgene ' + annot_path + '/refGene.txt -reflink ' + annot_path + '/refLink.txt -expandright 5m > "' + fname + '.genes"')<br>    os.system('rm "' + fname + '".genes1')<br>    os.system('rm "' + fname + '".genes2')<br><br>    #os.system('perl ' + penncnv_path + '/detect_cnv.pl -cctest -pfb ' + penncnv_path + '/lib/ho1v1.hg18.pfb -phenofile phenotype -cnv "' + fname + '" > "' + fname + '".CC')<br>    return fname + '.genes'<br></pre></div><hr><h3><a name="sampleSheet_details">sampleSheet</a></h3>
The name of the filename containing sample IDs and names<br>Output file: <a href="sample_sheet.tsv">sample_sheet.tsv</a><br><br><b>Last build was:</b> Sun Jan  8 03:37:36 2012.<br><b>Required CPU time:</b> 00:00:0.03.<br><div class="code"><pre>def sampleSheet():<br>    """The name of the filename containing sample IDs and names"""<br><br>    return('sample_sheet.tsv')<br></pre></div><hr><h3><a name="readFile_details">readFile</a></h3>
From now on the input does not come from files<br><br><b>Last build was:</b> Sun Jan  8 03:46:51 2012.<br><b>Required CPU time:</b> 00:00:0.06.<br><div class="code"><pre>def readFile(fname):<br>    """From now on the input does not come from files"""<br>    return open(fname).readlines()<br></pre></div><hr><h3><a name="distMatGfx_details">distMatGfx</a></h3>
Produces an MDS visualization of the output of samplesDistMats<br>Output files: <a href="t_tani_distrib.pdf">t_tani_distrib.pdf</a> , <a href="t_tani_mds.pdf">t_tani_mds.pdf</a> .<br><br><b>Last build was:</b> Sun Jan  8 03:51:56 2012.<br><b>Required CPU time:</b> 00:00:0.25.<br><div class="code"><pre>def distMatGfx(samples, mats):<br>    """Produces an MDS visualization of the output of samplesDistMats"""<br>    tani = mats[2]<br>    test_couples=[['13t','14t'], ['33t1','33t2'], ['34t','34tL'], ['5t', '5Bt']]<br>    #s = getSampleNames()<br>    s = samples<br>    <br>    tani_t = list()<br>    tani_t_mat = list()<br>    tids = list()<br>    print 'Adding background distances :'<br>    for idx1, s1 in enumerate(s):<br>        matrow = list()<br>        if 't' in s1:<br>            tids.append(s1)<br>            for idx2, s2 in enumerate(s):<br>                if 't' in s2:<br>                    matrow.append(tani[idx1,idx2])<br>                    if s1 != s2:<br>                        tani_t.append(tani[idx1,idx2])<br>                        #print s[idx1] + '-' + s[idx2]+',',<br>            tani_t_mat.append(matrow)<br>    #print<br><br>    #hist(tani_t,30)<br><br><br>    test_dists=list()<br>    sdict = idToNameDict()<br><br>    for idx, item in enumerate(test_couples):<br>        if item[0] not in s or item[1] not in s:<br>            continue<br>        s1 = s.index(item[0])<br>        s2 = s.index(item[1])<br>        d = tani[s1,s2]<br>        test_dists.append(d)<br>        #annotate('('+sdict[item[0]]+','+sdict[item[1]]+')', [d,10], [d,10*idx+20])<br>        print d<br><br><br>    mds=r.cmdscale(array(tani_t_mat))<br>    #savefig('t_tani_distrib.pdf')<br><br>    #figure()<br>    #plot(mds[:,0], mds[:,1], '.')<br>    for idx, item in enumerate(tids):<br>        pass<br>        #annotate(sdict[item], [mds[idx,0], mds[idx,1]])<br>    #savefig('t_tani_mds.pdf')<br><br>    return ['t_tani_distrib.pdf', 't_tani_mds.pdf']<br></pre></div><hr><h3><a name="intersectTBRegs_details">intersectTBRegs</a></h3>
Intersects all CNV T regions against all CNV B regions<br><br><b>Last build was:</b> Sun Jan  8 19:44:57 2012.<br><b>Required CPU time:</b> 00:00:7.36.<br><div class="code"><pre>def intersectTBRegs(getSampleNames, manualClean):<br>    """Intersects all CNV T regions against all CNV B regions"""<br>    f,h = PennCNVParser()<br>    manualClean = [f.split(l) for l in manualClean]<br>    num = dict()<br>    num['s'] = sum(['s' in s for s in getSampleNames])<br>    num['t'] = sum(['t' in s for s in getSampleNames])<br><br>    regsets = dict()<br>    chrs = set([l[1] for l in manualClean])<br><br>    for samp in getSampleNames:<br>        for chrom in chrs:            <br>            sub = [l for l in manualClean if l[7]==samp and l[1] == chrom]<br>            st = 'B' if 's' in samp else 'T'<br>            regset = [[int(l[2]),int(l[3]), l[10], l[6], st] for l in sub]<br><br>            if not chrom in regsets.keys():<br>                regsets[chrom] = list()<br>            if not sub == []:<br>                regsets[chrom].append(regset)<br><br>    a = dict()<br>    for chrom in chrs:<br> #       if chrom == '13':<br> #           import pdb; pdb.set_trace()<br>        a[chrom]=intersRegs([sorted(s) for s in regsets[chrom]])<br>    return a<br></pre></div><hr><h3><a name="exportToCircos_details">exportToCircos</a></h3>
Exports to a file suitable for use with Circos<br><br><b>Last build was:</b> Sun Jan  8 03:51:59 2012.<br><b>Required CPU time:</b> 00:00:0.02.<br><div class="code"><pre>def exportToCircos(samples, x):<br>    """Exports to a file suitable for use with Circos"""<br>    return "tobefixed!"<br>    x=[l.split('\t') for l in x][1:]<br>    x=[[l[0],l[1],l[2],l[5],l[6]] for l in x]<br>    fnames = list()<br>    for sample in samples:<br>        fname = 'circosInput_' + sample + '.txt'<br>        fnames.append(fname)<br>        f = open(fname, 'w')<br>        subx = [l[0:4] for l in x if l[4]==sample]<br>        f.writelines("hs%s\n" % str(item).strip('[]').replace('\'','').replace(',','') for item in subx)<br>        <br>    return(fnames)<br></pre></div><hr><h3><a name="intersectRegs_details">intersectRegs</a></h3>
Intersects all CNV regions of T and B samples separately<br><br><b>Last build was:</b> Sun Jan  8 03:47:15 2012.<br><b>Required CPU time:</b> 00:00:5.19.<br><div class="code"><pre>def intersectRegs(getSampleNames, manualClean):<br>    """Intersects all CNV regions of T and B samples separately"""<br>    f,h = PennCNVParser()<br>    manualClean = [f.split(l) for l in manualClean]<br>    num = dict()<br>    num['s'] = sum(['s' in s for s in getSampleNames])<br>    num['t'] = sum(['t' in s for s in getSampleNames])<br><br>    regsets = dict()<br>    chrs = set([l[1] for l in manualClean])<br><br>    for samp in getSampleNames:<br>        for chrom in chrs:<br>            sub = [l for l in manualClean if l[7]==samp and l[1] == chrom]<br>            regset = [[int(l[2]),int(l[3]), l[10], l[6]] for l in sub]<br>            if 's' in samp:<br>                if not ('s',chrom) in regsets.keys():<br>                    regsets['s', chrom] = list()<br>                regsets['s',chrom].append(regset)<br>            else:<br>                if not ('t',chrom) in regsets.keys():<br>                    regsets['t', chrom] = list()<br>                regsets['t',chrom].append(regset)<br><br>    a = dict()<br>    for chrom in chrs:<br>        a['s',chrom]=intersRegs([sorted(s) for s in regsets['s',chrom]])<br>        a['t',chrom]=intersRegs([sorted(s) for s in regsets['t',chrom]])<br>    return a<br></pre></div><hr><h3><a name="reformat2_details">reformat2</a></h3>
Dummy filter to skip manual clean phase<br><br><b>Last build was:</b> Sun Jan  8 03:51:58 2012.<br><b>Required CPU time:</b> 00:00:1.83.<br><div class="code"><pre>def reformat2(ifile):<br>    'Dummy filter to skip manual clean phase'<br><br>    f, header = PennCNVParser()<br><br>    ofile = list()<br>    ofile.append('\t'.join(header))<br><br>    p = getPairings()<br><br>    for line in ifile:<br>        newline = (f.split(line)[1:-1])<br>        try:<br>            t=newline.pop(10)<br>        except IndexError:<br>            print line<br>        <br>        <br>        if newline[6] in [x[0] for x in p]:<br>            newline.insert(7,'s')<br>        else:<br>            newline.insert(7,'t')<br><br>        genes = newline[10].split(',')<br>        for gene in genes:<br>            newline[10]=gene        <br>            ofile.append('\t'.join(newline))<br>            <br>    return ofile<br></pre></div><hr><h3><a name="genoTypeCheck_details">genoTypeCheck</a></h3>
Builds a matrix of % of genotype differences between each couple of samples<br><br><b>Last build was:</b> Sun Jan  8 19:42:16 2012.<br><b>Required CPU time:</b> 00:02:57.52.<br><div class="code"><pre>def genoTypeCheck(fnames):<br>    """Builds a matrix of % of genotype differences between each couple of samples"""<br>    data=[]<br>    for ifile in fnames:<br>        curgtype = list()<br>        cursamp = open(ifile).readlines()<br>        for line in cursamp:<br>            line = line.split('\t')<br>            curgtype.append(line[3])<br>        data.append(curgtype)<br><br>    <br>    data=array(data)<br>    differences=list()<br>    f = open('gtypecheck_small.txt', 'w')<br>    sdict = idToNameDict()<br><br>    for sample1 in data:<br>        templist = list()<br>        for sample2 in data:<br>            templist.append(float(sum(array(sample1)!=array(sample2))) / len(sample1) * 100)<br>            try:<br>                print sdict[sample1[0].replace('.GType','')], 'vs', sdict[sample2[0].replace('.GType','')],':','\t', templist[-1], '\t'<br><br>            #print templist[-1]<br><br>                if templist[-1] < 20 and sdict[sample1[0].replace('.GType','')] != sdict[sample2[0].replace('.GType','')]:<br>                    f.write(sdict[sample1[0].replace('.GType','')] + ' vs ' + sdict[sample2[0].replace('.GType','')] + '\t' + str(templist[-1]) + '\n')<br>        <br>                differences.append(templist)<br>    <br>                f = open('gtypecheck.txt', 'w')<br>                for idx, item in enumerate(differences):<br>                    f.write(sdict[data[idx][0].replace('.GType','')]+'  ')<br>                    f.write('\n')<br>                for idx, item in enumerate(differences):<br>                        f.write('  ' + sdict[data[idx][0].replace('.GType','')])<br>                        f.writelines(str(array(item)).strip('[]').replace('\n','')+'\n')<br><br>            except:<br>                print 'Skipping couple: ' + sample1[0] +', '+sample2[0]<br><br>    return differences<br></pre></div><hr><h3><a name="getGeneNames_details">getGeneNames</a></h3>
Extract the names of all annotated genes<br><br><b>Last build was:</b> Sun Jan  8 03:47:16 2012.<br><b>Required CPU time:</b> 00:00:0.05.<br><div class="code"><pre>def getGeneNames(f):<br>    """Extract the names of all annotated genes"""<br>    data=list()<br>    for line in f:<br>        nline=line[0:-1]<br>        t = nline.split()<br>        data.extend(t[7].split(','))<br>    data=array(data)   <br>    <br>    return unique(data)<br></pre></div><hr><h3><a name="countCNVlenghts_details">countCNVlenghts</a></h3>
Computes percentage of genome affected by CNV<br><br><b>Last build was:</b> Sun Jan  8 19:43:38 2012.<br><b>Required CPU time:</b> 00:00:0.36.<br><div class="code"><pre>def countCNVlenghts(getSampleNames, reformat):<br>    """Computes percentage of genome affected by CNV"""<br>    di = idToNameDict()<br>    y=[l.split('\t') for l in reformat]<br>    glength=float(2867188341)<br>    res = []<br>    <br>    for s in getSampleNames:<br>        try:<br>            sub=[l for l in y if l[6]==s]<br>            lens = [int(l[2])-int(l[1]) for l in sub]<br>            res.append([di[s] , str(sum(lens)) , str(100*(sum(lens) / glength))])<br>        except:<br>            import pdb; pdb.set_trace()<br>            <br>    for st in ['s', 't']:<br>        sub = [l for l in y if l[7]==st]<br>        lens = [int(l[2])-int(l[1]) for l in sub]<br>        if st=='s':<br>            label = 'Blood'<br>        else:<br>            label = 'Tumor'<br>        print label + ' average: ' + str(sum(lens) /<br>                                         (len(getSampleNames)/float(2))) +'\t' + str(100 * (sum(lens)/ (len(getSampleNames)/float(2)) / glength) )<br><br>    return res<br></pre></div><hr><h3><a name="addStats_details">addStats</a></h3>
Adds descriptive statistics<br><br><b>Last build was:</b> Sun Jan  8 03:47:17 2012.<br><b>Required CPU time:</b> 00:00:0.99.<br><div class="code"><pre>def addStats(fname):<br>    """Adds descriptive statistics"""<br>    data=list()<br>    f = fname<br>    for line in f:<br>        line=line[0:-1]<br>        data.append(line.split('\t'))<br>    data=array(data)<br>    <br>    genes = unique(data[:,10])<br><br>    p = getPairings()<br><br>    s=list()<br>    for samp in p[0]:<br>        for raw in data:<br>            if raw[6] == samp:<br>                s.append(raw)<br>    s=array(s)<br><br><br>    t=list()<br><br>    for samp in p[1]:<br>        for raw in data:<br>            if raw[6] == samp:<br>                t.append(raw)<br>    t=array(t)<br><br>    gene_count = dict()<br><br>    for gene in genes:<br>        if gene == 'NOT_FOUND':<br>            gene_count[gene]=(0,0,0,0)<br>        else:<br>            s_where = s[nonzero(s[:,10]==gene)]<br>            t_where = t[nonzero(t[:,10]==gene)]<br>            s_count = sum(s[:,10]==gene)<br>            t_count = sum(t[:,10]==gene)<br>            s_diffcount = len(unique(s_where[:,6]))<br>            t_diffcount = len(unique(t_where[:,6]))<br>            gene_count[gene]=(s_count, t_count, s_diffcount, t_diffcount)<br>        <br>    of=list()<br>    for idx, line in enumerate(data):        <br>        if idx==0:<br>            of.append('\t'.join(line) + '\ts hits \tt hits\thits diff.\ts un. hits\tt un. hits\t un. hits diff.')<br>        else:<br>            currinfo = gene_count[line[10]]<br>            of.append('\t'.join(line)+\<br>                    '\t'+str(currinfo[0])+\<br>                    '\t'+str(currinfo[1])+\<br>                    '\t'+str(abs(currinfo[0]-currinfo[1]))+\<br>                    '\t'+str(currinfo[2])+\<br>                    '\t'+str(currinfo[3])+\<br>                    '\t'+str(abs(currinfo[2]-currinfo[3])))<br>                     #'\t'+str(pvalue(currinfo[2], 8-currinfo[2], currinfo[3], 8-currinfo[3]).two_tail)+'\n')<br>                     <br>    return of<br></pre></div><hr><h3><a name="exportCNVlenghts_details">exportCNVlenghts</a></h3>
[documentation not available]<br>Output file: <a href="cnv_lenghts.tsv">cnv_lenghts.tsv</a><br><br><b>Last build was:</b> Sun Jan  8 19:43:38 2012.<br><b>Required CPU time:</b> 00:00:0.02.<br><div class="code"><pre>def exportCNVlenghts(countCNVl):<br>    f = open('cnv_lenghts.tsv', 'w')<br>    for line in countCNVl:<br>        f.writelines(str(line).strip('[]').replace(',','\t').replace("'",'')+'\n')<br>    return 'cnv_lenghts.tsv'<br></pre></div><hr><h3><a name="exportDistMats_details">exportDistMats</a></h3>
Exports the output of samplesDistMats to csv format<br>Output files: <a href="inters_mat.txt">inters_mat.txt</a> , <a href="notIntersMat.txt">notIntersMat.txt</a> , <a href="tanimotoMat.txt">tanimotoMat.txt</a> .<br><br><b>Last build was:</b> Sun Jan  8 19:43:38 2012.<br><b>Required CPU time:</b> 00:00:0.03.<br><div class="code"><pre>def exportDistMats(samples, mats):<br>    """Exports the output of samplesDistMats to csv format"""<br>    intMat=mats[0]<br>    notIntMat=mats[1]<br>    tanimotoMat=mats[2]<br><br>    f_intMat = open('intersMat.txt','w')<br>    f_notIntMat = open('notIntersMat.txt','w')<br>    f_tanimotoMat = open('tanimotoMat.txt','w')<br><br>    f_intMat.write(' ,'+','.join(samples) + '\n')<br>    f_notIntMat.write(' ,'+','.join(samples) + '\n')<br>    f_tanimotoMat.write(' ,'+','.join(samples) + '\n')<br><br>    for idx, item in enumerate(intMat):<br>        f_intMat.write(samples[idx]+',')<br>        f_intMat.write(str(list(intMat[idx])).strip('[]').replace('.0','').replace(' ','')+'\n')<br><br>        f_notIntMat.write(samples[idx]+',')<br>        f_notIntMat.write(str(list(notIntMat[idx])).strip('[]').replace('.0','').replace(' ','')+'\n')<br><br>        f_tanimotoMat.write(samples[idx]+',')<br>        f_tanimotoMat.write(str(list(tanimotoMat[idx])).strip('[]').replace(' ','')+'\n')<br><br>    return ['inters_mat.txt', 'notIntersMat.txt', 'tanimotoMat.txt']<br></pre></div><hr><h3><a name="intersMat_genes_details">intersMat_genes</a></h3>
Produces a list of common genes between each couple of samples<br><br><b>Last build was:</b> Sun Jan  8 03:47:19 2012.<br><b>Required CPU time:</b> 00:00:0.38.<br><div class="code"><pre>def intersMat_genes(dataset, fname, genes, samples):<br>    """Produces a list of common genes between each couple of samples"""<br>    inters_names = list()<br>    for idx1, geneline1 in enumerate(dataset):<br>        inters_names_t = list()<br>        for idx2, geneline2 in enumerate(dataset):<br>            inters_names_t.append(genes[nonzero(logical_and(geneline1, geneline2))])<br>        inters_names.append(inters_names_t)<br>    return inters_names<br></pre></div><hr><h3><a name="reformat_details">reformat</a></h3>
Change format from region-centric to gene-centric<br><br><b>Last build was:</b> Sun Jan  8 03:46:56 2012.<br><b>Required CPU time:</b> 00:00:0.69.<br><div class="code"><pre>def reformat(ifile):<br>    """Change format from region-centric to gene-centric"""<br>    f, header = PennCNVParser()<br><br>    ofile = list()<br>    ofile.append('\t'.join(header))<br><br>    p = getPairings()<br><br>    for line in ifile:<br>        newline = (f.split(line)[1:-1])<br>        try:<br>            t=newline.pop(10)<br>        except IndexError:<br>            print line<br>        <br>        <br>        if newline[6] in [x[0] for x in p]:<br>            newline.insert(7,'s')<br>        else:<br>            newline.insert(7,'t')<br><br>        genes = newline[10].split(',')<br>        for gene in genes:<br>            newline[10]=gene        <br>            ofile.append('\t'.join(newline))<br>            <br>    return ofile<br></pre></div><hr><h3><a name="CNVboolVects_details">CNVboolVects</a></h3>
Creates a matrix of presence/absence of CNV in each sample for each gene (out of all the gene that are affected by CNV in at least 1 sample<br><br><b>Last build was:</b> Sun Jan  8 03:47:18 2012.<br><b>Required CPU time:</b> 00:00:0.35.<br><div class="code"><pre>def CNVboolVects(raw, allsamples):<br>    """Creates a matrix of presence/absence of CNV in each sample for each gene (out of all the gene that are affected by CNV in at least 1 sample"""<br>    data=list()<br>    f=raw<br>    for line in f:<br>        line=line[0:-1]<br>        data.append(line.split('\t'))    <br>    data=array(data)<br>    <br>    <br>    genes = unique(data[1:-1,10])<br>    hashd = dict(   zip(  genes, range( 0,len(genes) )  )   )<br>    <br><br>    #samples = getSampleNames()<br>    samples = allsamples<br>    s=data<br>    vectors = np.zeros((len(samples),len(genes)))<br>    for s_idx, sample in enumerate(samples):        <br>        sample_genes = s[s[:,6]==sample][:,10]<br>        for gene in sample_genes:<br>            vectors[s_idx][hashd[gene]]=1<br><br>    return vectors<br></pre></div><hr><h3><a name="LogR_BAF_FileName_details">LogR_BAF_FileName</a></h3>
The name of the file containing Log R Ratio and B Allele Frequency data<br>Output file: <a href="pglfeb.csv">pglfeb.csv</a><br><br><b>Last build was:</b> Sun Jan  8 03:37:36 2012.<br><b>Required CPU time:</b> 00:00:0.03.<br><div class="code"><pre>def LogR_BAF_FileName():<br>    """The name of the file containing Log R Ratio and B Allele Frequency data"""<br>    return(fname)<br></pre></div><hr><h3><a name="clustergram_details">clustergram</a></h3>
Calls R to make a clustergram<br>Output files: <a href="clustergram_001_108.png">clustergram_001_108.png</a> , <a href="clustergram_005_248.png">clustergram_005_248.png</a> , <a href="clustergram_05_1606.png">clustergram_05_1606.png</a> , <a href="heatmap_001_108.png">heatmap_001_108.png</a> , <a href="heatmap_005_248.png">heatmap_005_248.png</a> , <a href="heatmap_05_1606.png">heatmap_05_1606.png</a> .<br><br><b>Last build was:</b> Sun Jan  8 19:44:49 2012.<br><b>Required CPU time:</b> 00:00:57.72.<br><div class="code"><pre>def clustergram(diffmat, fisher, genes, samples):<br>    """Calls R to make a clustergram"""<br>    #samplesNoLetter = [item.replace('t','').replace('s','') for item in samples]<br>    #samples = [] <br>    #[samples.append(i) for i in samplesNoLetter if not i in samples]<br><br>    p=getPairings()<br>    sdict = idToNameDict()<br>    s2 = list()<br>    for i in range(0,len(p)):<br>        s2.append(sdict[p[i][0]]+' vs '+sdict[p[i][1]])<br>    samples = s2<br><br>        <br><br>    d=diffmat[0]<br>    gdict = geneChrDict()<br><br>    #import pdb; pdb.set_trace()<br>    thresholds = [0.01, 0.05, 0.5]<br>    cfnames = list()<br>    hfnames = list()<br><br>    for t in thresholds:<br>        ok_genes=list()<br>        ok_gnames=list()<br><br>        for idx, gene_row in enumerate(d):<br>            if fisher[1][idx] < t:       <br>                if str.lower(genes[idx]) != 'loc100130872-spon2' and str.lower(genes[idx]) != 'loc647859':<br>                    ok_genes.append(gene_row)<br>                    ok_gnames.append(genes[idx])<br><br>        print 'Selected genes: ' + str(len(ok_gnames))<br><br>        for idx, gene in enumerate(ok_gnames):<br>            if len(gdict[gene])==1:<br>                chr = '0'+ gdict[gene]<br>            else:<br>                chr = gdict[gene]<br>            ok_gnames[idx]= chr + ' ' + gene<br><br>        ok_genes=array(ok_genes)<br>        r.library('gplots')<br>        colors = r.redgreen(len(unique(ok_genes)))<br>        colors = colors[::-1]<br>        #colors[2]='#880000'<br>        colors[3]='#880000'<br>        #return ok_genes, samples, ok_gnames, colors<br><br>        cfnames.append('clustergram_' + str(t).replace('.','') + '_' + str(len(ok_gnames)) + '.png')<br>        hfnames.append('heatmap_' + str(t).replace('.','') + '_' + str(len(ok_gnames)) + '.png')<br><br>        callRforClustergram(cfnames[-1], ok_genes, samples, ok_gnames, colors)<br>        makeHeatmap(hfnames[-1], ok_genes, samples, ok_gnames, colors)<br>    <br><br>    return hstack([cfnames, hfnames])<br></pre></div><hr><h3><a name="PennCNV_details">PennCNV</a></h3>
Runs PennCNV to detect CNV regions<br>Output files: <a href="pglfeb.csv_11s.tab.out">pglfeb.csv_11s.tab.out</a> , <a href="pglfeb.csv_11t.tab.out">pglfeb.csv_11t.tab.out</a> , <a href="pglfeb.csv_12s.tab.out">pglfeb.csv_12s.tab.out</a> , <a href="pglfeb.csv_12t.tab.out">pglfeb.csv_12t.tab.out</a> , <a href="pglfeb.csv_13s.tab.out">pglfeb.csv_13s.tab.out</a> , <a href="pglfeb.csv_13t.tab.out">pglfeb.csv_13t.tab.out</a> , <a href="pglfeb.csv_14Bt.tab.out">pglfeb.csv_14Bt.tab.out</a> , <a href="pglfeb.csv_14s.tab.out">pglfeb.csv_14s.tab.out</a> , <a href="pglfeb.csv_19s.tab.out">pglfeb.csv_19s.tab.out</a> , <a href="pglfeb.csv_19t.tab.out">pglfeb.csv_19t.tab.out</a> , <a href="pglfeb.csv_1s.tab.out">pglfeb.csv_1s.tab.out</a> , <a href="pglfeb.csv_1t.tab.out">pglfeb.csv_1t.tab.out</a> , <a href="pglfeb.csv_20s.tab.out">pglfeb.csv_20s.tab.out</a> , <a href="pglfeb.csv_20t.tab.out">pglfeb.csv_20t.tab.out</a> , <a href="pglfeb.csv_21s.tab.out">pglfeb.csv_21s.tab.out</a> , <a href="pglfeb.csv_21t.tab.out">pglfeb.csv_21t.tab.out</a> , <a href="pglfeb.csv_2s.tab.out">pglfeb.csv_2s.tab.out</a> , <a href="pglfeb.csv_2t.tab.out">pglfeb.csv_2t.tab.out</a> , <a href="pglfeb.csv_32s.tab.out">pglfeb.csv_32s.tab.out</a> , <a href="pglfeb.csv_32t.tab.out">pglfeb.csv_32t.tab.out</a> , <a href="pglfeb.csv_33s.tab.out">pglfeb.csv_33s.tab.out</a> , <a href="pglfeb.csv_33t1.tab.out">pglfeb.csv_33t1.tab.out</a> , <a href="pglfeb.csv_34s.tab.out">pglfeb.csv_34s.tab.out</a> , <a href="pglfeb.csv_34t.tab.out">pglfeb.csv_34t.tab.out</a> , <a href="pglfeb.csv_36s.tab.out">pglfeb.csv_36s.tab.out</a> , <a href="pglfeb.csv_36t.tab.out">pglfeb.csv_36t.tab.out</a> , <a href="pglfeb.csv_37s.tab.out">pglfeb.csv_37s.tab.out</a> , <a href="pglfeb.csv_37t.tab.out">pglfeb.csv_37t.tab.out</a> , <a href="pglfeb.csv_3s.tab.out">pglfeb.csv_3s.tab.out</a> , <a href="pglfeb.csv_3t.tab.out">pglfeb.csv_3t.tab.out</a> , <a href="pglfeb.csv_43s.tab.out">pglfeb.csv_43s.tab.out</a> , <a href="pglfeb.csv_43t.tab.out">pglfeb.csv_43t.tab.out</a> , <a href="pglfeb.csv_44s.tab.out">pglfeb.csv_44s.tab.out</a> , <a href="pglfeb.csv_44t.tab.out">pglfeb.csv_44t.tab.out</a> , <a href="pglfeb.csv_45s.tab.out">pglfeb.csv_45s.tab.out</a> , <a href="pglfeb.csv_45t.tab.out">pglfeb.csv_45t.tab.out</a> , <a href="pglfeb.csv_4s.tab.out">pglfeb.csv_4s.tab.out</a> , <a href="pglfeb.csv_4t.tab.out">pglfeb.csv_4t.tab.out</a> , <a href="pglfeb.csv_5Bt.tab.out">pglfeb.csv_5Bt.tab.out</a> , <a href="pglfeb.csv_5s.tab.out">pglfeb.csv_5s.tab.out</a> , <a href="pglfeb.csv_5t.tab.out">pglfeb.csv_5t.tab.out</a> , <a href="pglfeb.csv_6s.tab.out">pglfeb.csv_6s.tab.out</a> , <a href="pglfeb.csv_6t.tab.out">pglfeb.csv_6t.tab.out</a> , <a href="pglfeb.csv_7s.tab.out">pglfeb.csv_7s.tab.out</a> , <a href="pglfeb.csv_7t.tab.out">pglfeb.csv_7t.tab.out</a> , <a href="pglfeb.csv_8s.tab.out">pglfeb.csv_8s.tab.out</a> , <a href="pglfeb.csv_8t.tab.out">pglfeb.csv_8t.tab.out</a> .<br><br><b>Last build was:</b> Sun Jan  8 19:10:54 2012.<br><b>Required CPU time:</b> 00:52:42.29.<br><div class="code"><pre>def PennCNV(inputfiles):<br>    """Runs PennCNV to detect CNV regions"""<br>    ofiles = list()<br>    for ifile in inputfiles:<br>        cmdstr = 'perl ' + penncnv_path + \<br>            '/detect_cnv.pl -test -hmm ' + \<br>            penncnv_path + '/lib/hhall.hmm -pfb ' + \<br>            penncnv_path + '/lib/ho1v1.hg18.pfb -gcmodelfile ' + \<br>            penncnv_path + '/lib/ho1v1.hg18.gcmodel ' + \<br>            ifile + \<br>            ' -log "' + ifile + '.log" ' +  \<br>            '-out "' + fname + '_' + ifile + '.out"'<br>            <br>        print 'Executing: ' + cmdstr<br>        #print 'WARNING: skipping PennCNV call! Uncomment to really call.'<br>        os.system(cmdstr)<br>        #ofiles.append(lp.newFile(fname + '_' + ifile + '.out'))<br>        ofiles.append(fname + '_' + ifile + '.out')<br>    return ofiles<br></pre></div><hr><h3><a name="getSampleNames_details">getSampleNames</a></h3>
Extracts sample names<br><br><b>Last build was:</b> Sun Jan  8 03:47:10 2012.<br><b>Required CPU time:</b> 00:00:0.03.<br><div class="code"><pre>def getSampleNames(fname):<br>    """Extracts sample names"""<br>    b=open(fname).readlines()<br>    t=list()<br>    p = getPairedSampleIDs()<br>    for l in b[1:]:<br>        s = l.split()[0]<br>        #if s in p:<br>        t.append(s)<br>    return t<br></pre></div><hr><h3><a name="CompareRegionAndGenes_details">CompareRegionAndGenes</a></h3>
Computes intersections between regions and genes<br><br><b>Last build was:</b> Sun Jan  8 03:47:09 2012.<br><b>Required CPU time:</b> 00:00:13.22.<br><div class="code"><pre>def CompareRegionAndGenes(data):<br>    """Computes intersections between regions and genes"""<br>    ref = open(penncnv_path+'/ucsc_hg18/refGene.txt').readlines()<br>    for idx, line in enumerate(ref):<br>        ref[idx] = line.split()<br>    refgenes = [ref[i][12] for i in range(0, len(ref))]<br>    <br>    res = list()<br>    for idx, line in enumerate(data[1:]):<br>        if mod(idx, 100)==0:<br>            print 'Intersecting: '+str(idx)+' of '+str(len(data)-1)<br>        line = line.split('\t')<br>        gene = line[10]<br>        try:<br>            where = refgenes.index(gene)<br>            int1 = regIntersect([int(line[1]), int(line[2])], [int(ref[where][4]), int(ref[where][5])])<br>            int2 = regIntersect([int(line[1]), int(line[2])], [int(ref[where][6]), int(ref[where][7])])<br>            #res.append(['***********************************************'])<br>            res.append([<br>                        gene, <br>                        line[6],            #sample<br>                        line[7],            #s/t<br>                        int(line[1]),       #reg start<br>                        int(line[2]),       #reg end<br>                        int(line[2])-int(line[1]), # reg size<br>                        int(ref[where][4]), #trans start<br>                        int(ref[where][5]), #trans end<br>                        int(ref[where][5])-int(ref[where][4]),<br>                        int(ref[where][6]), #code start<br>                        int(ref[where][7]), #code end<br>                        int(ref[where][7])-int(ref[where][6]),<br>                        int1,               #int trans<br>                        int2,               #int code<br>                        ])<br>        except: <br>            res.append([<br>                        gene,<br>                        line[6],<br>                        line[7],<br>                        int(line[1]),<br>                        int(line[2]),<br>                        '?', '?', '?', '?', '?', '?','?','?'<br>                        ])<br>    <br>    return res<br></pre></div><hr><h3><a name="exportFigExtRegs_details">exportFigExtRegs</a></h3>
Creates a figure for the extended regions using R<br>Output file: <a href="extRegs.pdf">extRegs.pdf</a><br><br><b>Last build was:</b> Sun Jan  8 19:45:02 2012.<br><b>Required CPU time:</b> 00:00:3.74.<br><div class="code"><pre>def exportFigExtRegs(frags):<br>    """Creates a figure for the extended regions using R"""<br>    r.library('quantsmooth')<br>    chrwidth = 10<br>    chrspacing = 30<br><br>    r.pdf(file='extRegs.pdf', height=10, width=12)<br>    #maxarmlen = max(<br>    maxlen = max([a[0][1] for a in frags['1']])<br>    r.plot(1,type="n",xlim=[0,(chrspacing+chrwidth)*(len(frags)-1)],ylim=[0, maxlen],xaxt="n",yaxt="n", ann=False, frame_plot=False, tick_number=0, xlab="Autosomes")<br><br>    keys = [int(k) for k in frags]<br>    skeys = sort(keys)<br>    tickat = list()<br>    for cidx,chrom in enumerate([str(k) for k in skeys]):<br>        offs = cidx*(chrspacing+chrwidth)<br>        #maxy = max([a[0][1] for a in frags[chrom]])<br>        tothits = sum([a[0][1]-a[0][0] for a in frags[chrom][0::2]])<br>        maxlen = max([a[0][1] for a in frags[chrom]])<br>        maxy = r.lengthChromosome(chrom, units = 'bases')[chrom]<br>        r.lines([offs, offs], [0,maxy])<br><br>        for fidx, frag in enumerate(frags[chrom]):<br>            if frag[1]=='' and frag[2]=='':<br>                continue<br>            y = (frag[0][0]+float(frag[0][1]))/2<br>            y = (y * maxy) / maxlen<br>            l = frag [0][1]-frag[0][0]+1<br>            lB = int(frag[1])<br>            lT = int(frag[2])<br>            Ts = sum([a=='T' for a in frag[3]])<br>            Bs = sum([a=='B' for a in frag[3]])<br>            #Ts_p = sum([a=='T' for idx,a in enumerate(frag[3]) if int(frag[4][idx])>2])<br>            #Ts_n = sum([a=='T' for idx,a in enumerate(frag[3]) if int(frag[4][idx])<2])        <br>            #Bs_p = sum([a=='B' for idx,a in enumerate(frag[3]) if int(frag[4][idx])>2])<br>            #Bs_n = sum([a=='B' for idx,a in enumerate(frag[3]) if int(frag[4][idx])<2])<br>            Ts_p = sum([a=='T' for idx,a in enumerate(frag[3]) if int(frag[4][idx])>2])<br>            Ts_n = sum([a=='T' for idx,a in enumerate(frag[3]) if int(frag[4][idx])<2])        <br>            Bs_p = sum([a=='B' for idx,a in enumerate(frag[3]) if int(frag[4][idx])>2])<br>            Bs_n = sum([a=='B' for idx,a in enumerate(frag[3]) if int(frag[4][idx])<2])<br>            Tgain = Ts_p - Ts_n<br>            Bgain = Bs_p - Bs_n<br>            #Tcol = 'red' if Tgain < 0 else 'green'<br>            #Bcol = 'red' if Bgain < 0 else 'green'<br>            Dcol = 'darkred' if Tgain - Bgain > 0 else 'darkgreen'<br>            <br><br>            armlen = int(abs(lT - lB))*1.5<br>            if max(lB, lT)>0:<br>                if Dcol == 'darkred':<br>                    r.lines([offs +chrwidth-2 , armlen + offs +chrwidth-2], [maxy-y,maxy-y], col=Dcol)<br>                else:<br>                    r.lines([offs -armlen -chrwidth+2, offs -chrwidth+2], [maxy-y,maxy-y], col=Dcol)<br><br>        tickat.append(offs)<br>        print chrom, tothits/float(maxy)<br>        bleachv = 1-min((tothits/float(maxy))/0.08,1)<br><br>        r.paintCytobands(chrom, pos=[offs+chrwidth/2, maxy], width=chrwidth, units="bases", orientation='v', legend=False, bleach=bleachv)<br>        <br><br>#    r.axis(at=tickat, labels=range(1,23), cex_axis=1, side=range(1,23), mgp=[3,.5,0], xlab='Autosomes', main='Genomic regions affected by CNV')<br>    r.axis(at=tickat, lty=0, labels=range(1,23), cex_axis=1, side=range(1,23), xlab='Autosomes', main='Genomic regions affected by CNV', mgp=[3,-0.8,0])<br>        <br><br>                # if Bgain:<br>                #     #r.lines([-log10(l) + offs, 0 + offs], [y,y], col=Bcol)<br>                #     r.lines([offs - lB -1, 0 + offs -1], [y,y], col=Bcol)<br>                #     r.lines([offs - lB -1, 0 + offs -1], [y,y], col=Bcol)<br>                # if Tgain:<br>                #     #r.lines([0 + offs, log10(l) + offs], [y,y], col=Tcol)<br>                #     r.lines([0 + offs +1 , lT + offs +1], [y,y], col=Tcol)<br>                #     r.lines([0 + offs +1 , lT + offs +1], [y,y], col=Tcol)<br><br><br>    r.dev_off()<br>    return('extRegs.pdf')<br></pre></div><hr><h3><a name="manualClean_details">manualClean</a></h3>
Removes a list of CNVi-related genes produced elsewhere<br><br><b>Last build was:</b> Sun Jan  8 03:46:55 2012.<br><b>Required CPU time:</b> 00:00:3.48.<br><div class="code"><pre>def manualClean(x):<br>    """Removes a list of CNVi-related genes produced elsewhere"""<br>    no=[int(n) for n in open('manuallyRemovedCNVs.txt').read().split()]<br>    cleandata = list()<br>    f = open('cleanedCNVs.txt', 'w')<br><br>    for idx, line in enumerate(x):<br>        if not (idx+1 in no):<br>            cleandata.append(line)<br>            f.write(line)<br>    return cleandata<br></pre></div><hr><h3><a name="samplesDistMats_details">samplesDistMats</a></h3>
Computes Tanimoto distance between samples<br><br><b>Last build was:</b> Sun Jan  8 03:47:21 2012.<br><b>Required CPU time:</b> 00:00:0.24.<br><div class="code"><pre>def samplesDistMats(dataset):<br>    """Computes Tanimoto distance between samples"""<br>    inters_mat = squareform(pdist(dataset, lambda x,y: sum(logical_and(x,y))))<br>    for idx in range(0,dataset.shape[0]):<br>        inters_mat[idx,idx]=sum(dataset,axis=1)[idx]    <br>        <br>    notIntersMat = squareform(pdist(dataset, lambda x,y: sum(~logical_and(x,y))))<br>    tanimotoMat = squareform(pdist(dataset, lambda x,y: 1 - sum(logical_and(x,y))/float(sum(logical_or(x,y)))  ))<br><br>    return [inters_mat, notIntersMat, tanimotoMat]<br></pre></div><hr><h3><a name="exportPerRegion_clean_details">exportPerRegion_clean</a></h3>
Dummy node to skip manual filtering phase<br>Output file: <a href="perRegion_clean.csv">perRegion_clean.csv</a><br><br><b>Last build was:</b> Sun Jan  8 19:45:02 2012.<br><b>Required CPU time:</b> 00:00:0.14.<br><div class="code"><pre>def exportPerRegion_clean(data):<br>    """Dummy node to skip manual filtering phase"""<br>    f = open('perRegion_clean.csv','w')<br>    sdict = idToNameDict()<br>    f.write(data[0].replace('\t',',')+'\n')<br><br>    for line in data[1:]:<br>        l = line.replace(',','.').split()<br>        l[6] = sdict[l[6]]<br>        l = str(l).replace('\'', '').strip('[]')<br>        f.write(l+'\n')<br>    return 'perRegion_clean.csv'<br></pre></div><hr><h3><a name="exportCNVDiffMat_details">exportCNVDiffMat</a></h3>
Exports to tsv file<br>Output file: <a href="CNVgenes.csv">CNVgenes.csv</a><br><br><b>Last build was:</b> Sun Jan  8 19:45:15 2012.<br><b>Required CPU time:</b> 00:00:12.21.<br><div class="code"><pre>def exportCNVDiffMat(cnvdiff, cnvdata, fisher, genes, samples):<br>    """Exports to tsv file"""<br>    p = getPairings()<br>    samples = list()<br>    for a in p:<br>        samples.extend([a[0], a[1]])<br>    <br>    samples = samplesToNames(samples)        <br><br>    data = list()<br>    for line in cnvdata:<br>        data.append(line.split('\t'))<br>    data = array(data)<br>    vects, vs, vt = cnvdiff[0], cnvdiff[1], cnvdiff[2]<br>    gdict = geneChrDict()<br><br>    for i in range(len(samples),0,-1):<br>        if mod(i,2)==0:<br>            samples.insert(i, samples[i-1]+'-'+samples[i-2])<br>            <br>            header =['gene', 'Chr',<br>                     'B gain', 'B loss', 'B CNV',<br>                     'T gain', 'T loss', 'T CNV', <br>                     'B CNV + T CNV',<br>                     'TvsB loss','TvsB gain', 'TvsBgain+TvsBloss', 'TvsB loss/gain p-value',<br>                     'max CNV', 'Reg. Max length',<br>                     'CNV samples (repeated if hit in multiple regions)']<br><br>    header = hstack([header, samples])<br>  <br>    fisher_abs = fisher[0]<br>    fisher_diff = fisher[1]<br><br>    f = open('CNVgenes.csv', 'w')<br>    f.write(str(header).strip('[]').replace('\n','').replace('\' \'',',').strip('\'') + '\n')<br>    for idx, vect in enumerate(vects):<br>        #gene name<br>        f.write(genes[idx])<br>        #if genes[idx]=='ACOT11':<br>        #    import pdb; pdb.set_trace()<br><br>        #chromosome<br>        f.write(',' + str(gdict[genes[idx]]))<br>        <br>        #cnv stats<br>        cnvBgain = sum(vs[idx]>2)<br>        cnvBloss = sum(vs[idx]<2)<br>        cnvB = cnvBgain + cnvBloss<br><br>        cnvTgain = sum(vt[idx]>2)<br>        cnvTloss = sum(vt[idx]<2)<br>        cnvT = cnvTgain + cnvTloss<br><br>        f.write(', ' + str(cnvBgain) + ',' + str(cnvBloss) + ',' + str(cnvB))<br>        f.write(', ' + str(cnvTgain) + ',' + str(cnvTloss) + ',' + str(cnvT))<br>        f.write(', ' + str(cnvB+cnvT))<br>#        f.write(', ' + str(cnvB) + ',' + str(cnvT) + ',' + str(cnvB+cnvT))<br><br>        #fisher on B vs T cnvs<br>        #f.write(','+str(fisher_abs[idx]))<br>        <br>        #differential cnvs stats<br>        f.write(str(',' +<br>                    str(sum(vect<0))+','+<br>                    str(sum(vect>0))+','+<br>                    str(sum(vect<0)+<br>                        sum(vect>0))+',')<br>                .replace('\n',''))<br><br>        #fisher on s vs t differential cnvs<br>        f.write(',' + str(fisher_diff[idx]) + ',')<br><br><br>        # max CNV<br>        cns = hstack([vs[idx],vt[idx]])<br>        maxcnv = cns[argmax(abs(2-cns))]<br>        f.write(str(maxcnv) + ',')<br><br>        # per gene additional info<br>        info = extractPerGeneProperties(data, genes[idx])[1:]<br>        for idx2, item in enumerate(info):<br>            if not item=='nan':<br>                if idx2 in [3,4,5]:<br>                    f.write(item.replace(',','').replace('\n','').replace('\'', '').replace('[','').replace(']','')+',')<br><br>                    <br>        #cnv per sample<br>        f.write(str(zip(vs[idx],vt[idx],vect)).strip('[]').replace('(','').replace(')','').replace('\n','').replace('.0',''))<br><br>        f.write('\n')<br><br>    return 'CNVgenes.csv'<br></pre></div><hr><h3><a name="exportMergedFragments_details">exportMergedFragments</a></h3>
Exports extended regions to tsv file<br>Output file: <a href="extRegions.tsv">extRegions.tsv</a><br><br><b>Last build was:</b> Sun Jan  8 19:45:15 2012.<br><b>Required CPU time:</b> 00:00:0.09.<br><div class="code"><pre>def exportMergedFragments(mf):<br>    '''Exports extended regions to tsv file'''<br>    f = open('extRegions.tsv', 'w')<br><br>    f.write('Chr.\t'+<br>            'ExtReg. Start\t'+<br>            'ExtReg. End\t'+<br>            'ExtReg. Length\t'+<br>            'Max hits B\t'+<br>            'Max hits T\t'+<br>            'B/T\t'+<br>            'CN\t'+<br>            '"Genes"\n')<br><br>    for chrom in mf:<br>        for extreg in mf[chrom]:<br>            f.write(<br>                #chromosome<br>                chrom+'\t'+<br>                #reg start<br>                str(extreg[0][0])+'\t'+<br>                #reg end<br>                str(extreg[0][1])+'\t'+<br>                #reg length<br>                str(extreg[0][1]-extreg[0][0]+1)+'\t'+<br>                # max hits B<br>                str(extreg[1])+'\t'+<br>                # max hits T<br>                str(extreg[2])+'\t'+<br>                #BT<br>                str(extreg[3]).strip('[]').replace("'",'').replace(',','')+'\t'+<br>                #CN<br>                str(extreg[4]).strip('[]').replace("'",'').replace(',','')+'\t'+<br>                #genes<br>                str(extreg[5]).replace('set','').replace('(','').replace(')','')<br>                .replace('[','').replace(']','').replace("'",'')+'\n'<br>                    )<br><br>    return 'extRegions.tsv'<br></pre></div><hr><h3><a name="exportMergedRegsTB_details">exportMergedRegsTB</a></h3>
Exports to tsv file<br>Output file: <a href="mergedRegsTB.tsv">mergedRegsTB.tsv</a><br><br><b>Last build was:</b> Sun Jan  8 19:45:16 2012.<br><b>Required CPU time:</b> 00:00:0.66.<br><div class="code"><pre>def exportMergedRegsTB(mergedRegs):<br>    """Exports to tsv file"""<br>    f = open('mergedRegs.tsv', 'w')<br><br>    f.write('Chr.\tReg. Start\tReg. End\tReg. Lenght\tNum Samples\tCN-B\tCN-T\tnB\tnT\t"Genes"\n')<br><br>    for key in sorted(mergedRegs.keys()):<br>        for i in range(0, len(mergedRegs[key][0])):<br>#            try:<br><br>            try:<br>                cnb = [cn for idx, cn in enumerate(mergedRegs[key][3][i])<br>                       if mergedRegs[key][4][i][idx]=='B']<br>                cnt = [cn for idx, cn in enumerate(mergedRegs[key][3][i])<br>                       if mergedRegs[key][4][i][idx]=='T']<br>            except:<br>                import pdb; pdb.set_trace()<br>            f.write(<br>                # Chr<br>                key+'\t'+<br><br>                #Reg. Start e Reg. End<br>                str(mergedRegs[key][0][i]).strip('[]').replace(', ','\t')+'\t'+<br><br>                #Reg. Length<br>                str(mergedRegs[key][0][i][1]-mergedRegs[key][0][i][0]+1)+'\t'+<br><br>                #Num Samples<br>                str(mergedRegs[key][1][i])+'\t'+<br><br>                #CN-B<br>                str(cnb).<br>                replace('[','').<br>                replace(']','').<br>                replace("'", '')+'\t'+<br><br>                #CN-T<br>                str(cnt).<br>                replace('[','').<br>                replace(']','').<br>                replace("'", '')+'\t'+<br><br>                #nB<br>                str(sum([st=='B' for st in mergedRegs[key][4][i]]))+'\t'+<br><br>                #nT<br>                str(sum([st=='T' for st in mergedRegs[key][4][i]]))+'\t'+<br><br>                #Genes<br>                str(mergedRegs[key][2][i]).replace('set','').<br>                replace(')','').<br>                replace('(','').<br>                replace('[','').<br>                replace(']','').<br>                replace("'", '')+'\n'<br>                )<br>#           except: import pdb; pdb.set_trace()<br><br>    return 'mergedRegsTB.tsv'<br></pre></div><hr><h3><a name="mergeFragments_details">mergeFragments</a></h3>
Merges all contiguous non-void region fragments<br><br><b>Last build was:</b> Sun Jan  8 19:44:58 2012.<br><b>Required CPU time:</b> 00:00:0.37.<br><div class="code"><pre>def mergeFragments(intRegs):<br>    """Merges all contiguous non-void region fragments"""<br>    extreg=dict()<br><br>    for chrom in intRegs:<br>        extreg[chrom] = list()<br>        currExtRegBT = list()<br>        currExtRegCN = list()<br>        currExtRegS = [Inf,0]<br>        currExtRegG = set()<br>        currExtRegMaxhitsB = 0<br>        currExtRegMaxhitsT = 0<br><br>        for idx, frag in enumerate(intRegs[chrom][4]):<br>            a = intRegs[chrom][0][idx][0]<br>            b = intRegs[chrom][0][idx][1]<br>            #import pdb; pdb.set_trace()<br>            if frag != []:<br>                currExtRegMaxhitsT = max(currExtRegMaxhitsT, sum([a1=='T' for a1 in frag]))<br>                currExtRegMaxhitsB = max(currExtRegMaxhitsB, sum([a1=='B' for a1 in frag]))<br>                currExtRegBT.extend(frag)<br>                currExtRegCN.extend(intRegs[chrom][3][idx])<br>                currExtRegS = [min(currExtRegS[0], a), max(currExtRegS[1],b)]<br>                currExtRegG = currExtRegG.union(intRegs[chrom][2][idx])<br>            else:<br>                extreg[chrom].append([currExtRegS, currExtRegMaxhitsB,<br>                                      currExtRegMaxhitsT, currExtRegBT, currExtRegCN, currExtRegG])<br>                extreg[chrom].append([[a,b], '', '', '', '', 'unknown'])<br>                currExtRegCN=list()<br>                currExtRegBT=list()<br>                currExtRegG = set()<br>                currExtRegS = [Inf,0]<br>                currExtRegMaxhitsT = 0<br>                currExtRegMaxhitsB = 0<br>                                <br>        # handle last fragment (will get no '[]' as stop signal)<br>        if frag != []:<br>            extreg[chrom].append([currExtRegS, currExtRegMaxhitsB,<br>                                  currExtRegMaxhitsT, currExtRegBT, currExtRegCN, currExtRegG])<br>                                <br><br>    return extreg<br></pre></div><hr><h3><a name="computeFisher_details">computeFisher</a></h3>
Tests the hyphotesis that a gene is hit by CNV in M samples out of N by chance<br><br><b>Last build was:</b> Sun Jan  8 19:43:52 2012.<br><b>Required CPU time:</b> 00:00:8.96.<br><div class="code"><pre>def computeFisher(cnvect):<br>    """Tests the hyphotesis that a gene is hit by CNV in M samples out of N by chance"""<br>    svec = cnvect[1]<br>    tvec = cnvect[2]<br>    diff = cnvect[0]<br>    numsamp = len(svec[0])<br>    fishval_diff = zeros(svec.shape[0])<br>    fishval_abs = zeros(svec.shape[0])<br><br>    for idx, row in enumerate(diff):<br>        a11 = sum(row < 0)<br>        a12 = numsamp-a11<br>        a21 = sum(row > 0)<br>        a22 = numsamp-a21<br>        #fishval[idx]=r.fisher_test(array([[a11,a12],[a21,a22]]), alternative='less')['p.value']<br>        fishval_diff[idx]=r.fisher_test(array([[a11,a12],[a21,a22]]))['p.value']<br><br>    for idx, row in enumerate(svec):<br>        a11 = sum(row != 2)<br>        a12 = numsamp-a11<br>        a21 = sum(tvec[idx] != 2)<br>        a22 = numsamp-a21<br>        fishval_abs[idx]=r.fisher_test(array([[a11,a12],[a21,a22]]))['p.value']<br><br>    return fishval_abs, fishval_diff<br></pre></div><hr><h3><a name="prepareInput_details">prepareInput</a></h3>
Converts data to a format suitable for PennCNV<br>Output files: <a href="11s.tab">11s.tab</a> , <a href="11t.tab">11t.tab</a> , <a href="12s.tab">12s.tab</a> , <a href="12t.tab">12t.tab</a> , <a href="13s.tab">13s.tab</a> , <a href="13t.tab">13t.tab</a> , <a href="14Bt.tab">14Bt.tab</a> , <a href="14s.tab">14s.tab</a> , <a href="19s.tab">19s.tab</a> , <a href="19t.tab">19t.tab</a> , <a href="1s.tab">1s.tab</a> , <a href="1t.tab">1t.tab</a> , <a href="20s.tab">20s.tab</a> , <a href="20t.tab">20t.tab</a> , <a href="21s.tab">21s.tab</a> , <a href="21t.tab">21t.tab</a> , <a href="2s.tab">2s.tab</a> , <a href="2t.tab">2t.tab</a> , <a href="32s.tab">32s.tab</a> , <a href="32t.tab">32t.tab</a> , <a href="33s.tab">33s.tab</a> , <a href="33t1.tab">33t1.tab</a> , <a href="34s.tab">34s.tab</a> , <a href="34t.tab">34t.tab</a> , <a href="36s.tab">36s.tab</a> , <a href="36t.tab">36t.tab</a> , <a href="37s.tab">37s.tab</a> , <a href="37t.tab">37t.tab</a> , <a href="3s.tab">3s.tab</a> , <a href="3t.tab">3t.tab</a> , <a href="43s.tab">43s.tab</a> , <a href="43t.tab">43t.tab</a> , <a href="44s.tab">44s.tab</a> , <a href="44t.tab">44t.tab</a> , <a href="45s.tab">45s.tab</a> , <a href="45t.tab">45t.tab</a> , <a href="4s.tab">4s.tab</a> , <a href="4t.tab">4t.tab</a> , <a href="5Bt.tab">5Bt.tab</a> , <a href="5s.tab">5s.tab</a> , <a href="5t.tab">5t.tab</a> , <a href="6s.tab">6s.tab</a> , <a href="6t.tab">6t.tab</a> , <a href="7s.tab">7s.tab</a> , <a href="7t.tab">7t.tab</a> , <a href="8s.tab">8s.tab</a> , <a href="8t.tab">8t.tab</a> .<br><br><b>Last build was:</b> Sun Jan  8 03:42:33 2012.<br><b>Required CPU time:</b> 00:04:57.23.<br><div class="code"><pre>def prepareInput(fname, ssheet):<br>    """Converts data to a format suitable for PennCNV"""<br>    ss = [l.split('\t') for l in open(ssheet).readlines()][1:]<br><br>    scolumns = range(4,200,3)<br>    #a=open(fname).readline()<br>    #samples = [line.replace('.GType','') for line in a.split(',')[3::3]]<br>    samples = [s[0] for s in ss]<br><br><br>    fnames = list()<br>    for k, sample in enumerate(samples):<br>        print 'Preparazione campione ' + sample + '...',<br>        #print 'SKIPPING!'<br>        os.system('cut -f 1-3,' + str(scolumns[k]) + '-' + str(scolumns[k]+2) + ' --delimiter , "' + fname +'" > ' + sample + '.csv')<br>        os.system('cat ' + sample + ".csv | tr ',' '\t' > " + sample + '.tab')<br>        os.system('rm ' + sample + '.csv')<br>        fnames.append(sample + '.tab')<br><br><br>    return fnames<br></pre></div><hr><h3><a name="exportFigCNVlenghts_details">exportFigCNVlenghts</a></h3>
Creates a figure of the samples based % of genome affected by CNV<br>Output files: <a href="cnvlenghts.pdf">cnvlenghts.pdf</a> , <a href="cnvlenghts_log.pdf">cnvlenghts_log.pdf</a> .<br><br><b>Last build was:</b> Sun Jan  8 19:45:16 2012.<br><b>Required CPU time:</b> 00:00:0.04.<br><div class="code"><pre>def exportFigCNVlenghts(countCNVlenghts_o):<br>    <br>    """Creates a figure of the samples based % of genome affected by CNV"""<br>    r.pdf('cnvlenghts_log.pdf')<br>    y = [int(x[1])/1000000 for x in countCNVlenghts_o if not 'B' in x[0]]<br>    l = [x[0] for x in countCNVlenghts_o if not 'B' in x[0]]<br>    yidx = sorted(range(len(y)), key = y.__getitem__)<br>    ys = sorted(y)<br>    ls = [l[x] for x in yidx]<br>    r.barplot(ys, horiz=True, names_arg=ls, log='x', cex_names=0.5, las=1)<br>    r.dev_off()<br><br>    r.pdf('cnvlenghts.pdf')<br>    y = [int(x[1]) for x in countCNVlenghts_o if not 'B' in x[0]]<br>    l = [x[0] for x in countCNVlenghts_o if not 'B' in x[0]]<br>    yidx = sorted(range(len(y)), key = y.__getitem__)<br>    ys = sorted(y)<br>    ls = [l[x] for x in yidx]<br>    r.barplot(ys, horiz=True, names_arg=ls, cex_names=0.5, las=1)<br>    #r.lines(range(0,len(ys)), r.cumsum(ys))<br>    #r.lines(r.cumsum(y), range(0,len(y)))<br>    r.dev_off()<br><br>    return ('cnvlenghts.pdf', 'cnvlenghts_log.pdf')<br></pre></div><hr><h3><a name="makeBed_details">makeBed</a></h3>
Create Bed file for use with Genome Browser<br>Output file: <a href="pglfeb.csv.bed">pglfeb.csv.bed</a><br><br><b>Last build was:</b> Sun Jan  8 19:45:16 2012.<br><b>Required CPU time:</b> 00:00:0.42.<br><div class="code"><pre>def makeBed(CNVinput):<br>    """Create Bed file for use with Genome Browser"""<br>    os.system('perl ' + penncnv_path + '/visualize_cnv.pl -format bed -output "' + fname + '.bed" "' + CNVinput + '"')<br>    return fname + '.bed'<br></pre></div><hr><h3><a name="CNVDiffMat_details">CNVDiffMat</a></h3>
Makes a matrix of B/T difference in CN between same genes of different samples<br><br><b>Last build was:</b> Sun Jan  8 19:43:41 2012.<br><b>Required CPU time:</b> 00:00:2.43.<br><div class="code"><pre>def CNVDiffMat(fname, genes, allsamples):<br>    """Makes a matrix of B/T difference in CN between same genes of different samples"""<br>    data=list()<br>    f=fname<br>    for line in f:<br>        line=line[0:-1]<br>        data.append(line.split('\t'))    <br>    data=array(data)<br><br>    hashd = dict(  zip( genes, range(0,len(genes)) )    )    <br><br>    p = getPairings()<br>    sp = [x[0] for x in p]<br>    tp = [x[1] for x in p]<br><br>    s = list()<br>    for sample in sp:<br>        for line in data:<br>            if sample == line[6]:<br>                s.append(line)<br><br>    t = list()<br>    for sample in tp:<br>        for line in data:<br>            if sample == line[6]:<br>                t.append(line)<br><br>    samples = sp<br>    s_vectors = np.zeros((len(samples),len(genes)))*nan<br>    s_vectors_min =np.zeros((len(samples),len(genes)))*nan<br>    s_vectors_num =np.zeros((len(samples),len(genes)))*nan<br>    s_vectors_cn = np.zeros((len(samples),len(genes)))*nan<br><br>    for s_idx, sample in enumerate(samples):       <br>        print s_idx, 'of', len(samples),',',<br>        x = nonzero([l[6]==sample for l in s])[0]<br>        sub_s = [s[pickme] for pickme in x]<br>        sub_s = array(sub_s)<br>        sample_genes = sub_s[:,10]<br>        <br>        for gene in unique(sample_genes):<br>            if isnan(s_vectors[s_idx][hashd[gene]]):<br>                temp = array(sub_s[sub_s[:,10]==gene][:,5], int)<br>                maxidx = argmax(temp)<br>                minidx = argmin(temp)<br>                absmax = temp[maxidx]<br>                absmin = temp[minidx]<br>                s_vectors[s_idx][hashd[gene]] = absmax<br>                s_vectors_min[s_idx][hashd[gene]] = absmin<br>                s_vectors_cn[s_idx][hashd[gene]] = assignGeneCN(temp)<br>                s_vectors_num[s_idx][hashd[gene]] = len(temp)<br>    print<br>    s_vectors[isnan(s_vectors)] = 2<br>    s_vectors_min[isnan(s_vectors_min)] = 2<br>    s_vectors_cn[isnan(s_vectors_cn)] = 2<br>    s_vectors_num[isnan(s_vectors_num)] = 0<br><br><br>    samples = tp<br>    t_vectors = np.zeros((len(samples),len(genes)))*nan<br>    t_vectors_min = np.zeros((len(samples),len(genes)))*nan<br>    t_vectors_num =np.zeros((len(samples),len(genes)))*nan<br>    t_vectors_cn = np.zeros((len(samples),len(genes)))*nan<br>    <br>    for t_idx, sample in enumerate(samples):<br>        print t_idx, 'of', len(samples),',',<br>        x = nonzero([l[6]==sample for l in t])[0]<br>        sub_t = [t[pickme] for pickme in x]<br>        sub_t = array(sub_t)<br><br>        try:<br>            sample_genes = sub_t[:,10]<br>        except IndexError:<br>                import pdb; pdb.set_trace()<br><br>        for gene in unique(sample_genes):<br>            #if gene == 'PC':<br>            #    import pdb; pdb.set_trace()<br>            if isnan(t_vectors[t_idx][hashd[gene]]):<br>                temp = array(sub_t[sub_t[:,10]==gene][:,5], int)<br>                maxidx = argmax(temp)<br>                absmax = temp[maxidx]<br>                minidx = argmin(temp)<br>                absmin = temp[minidx]<br>                t_vectors[t_idx][hashd[gene]] = absmax<br>                t_vectors_min[t_idx][hashd[gene]] = absmin<br>                t_vectors_num[t_idx][hashd[gene]] = len(temp)<br>                t_vectors_cn[t_idx][hashd[gene]] = assignGeneCN(temp)<br>    print<br>    t_vectors[isnan(t_vectors)]=2<br>    t_vectors_min[isnan(t_vectors_min)] = 2<br>    t_vectors_cn[isnan(t_vectors_cn)] = 2<br>    t_vectors_num[isnan(t_vectors_num)] = 0<br><br>    return [t_vectors_cn.T-s_vectors_cn.T, #0<br>            s_vectors_cn.T, t_vectors_cn.T, #1,2<br>            s_vectors_min.T, t_vectors_min.T, #3,4<br>            s_vectors_num.T, t_vectors_num.T, #5,6<br>            s_vectors.T, t_vectors.T] #7,8<br></pre></div><hr><div id="footer"><br><br><i>Automatically generated on Sun Jan  8 19:46:10 2012 using Leaf</i></div>
</body>
</html>
