

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Preparing raw Illumina data in different formats for use with QIIME &mdash; Homepage</title>
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.7.0-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="preparing-raw-illumina-data-in-different-formats-for-use-with-qiime">
<span id="processing-illumina-data"></span><h1>Preparing raw Illumina data in different formats for use with QIIME<a class="headerlink" href="#preparing-raw-illumina-data-in-different-formats-for-use-with-qiime" title="Permalink to this headline">¶</a></h1>
<p>This document describes how to process Illumina sequencing data with QIIME. It covers a standard workflow beginning with fastq files, and take users through one pipeline for generation of an OTU table. Due to the huge amount of data in Illumina files, it&#8217;s best to run the OTU picking through OTU table steps on a cluster. For a full HiSeq2000 run, this process can take up to 500 CPU hours.</p>
<p>This example illustrates how to use <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a> to parse your Illumina output into a format that can be used by QIIME.</p>
<div class="section" id="input-file-formats">
<h2>Input file formats<a class="headerlink" href="#input-file-formats" title="Permalink to this headline">¶</a></h2>
<p>The preferred format in QIIME for Illumina data is fastq. This format is illustrated <a class="reference internal" href="#fastq-format"><em>here</em></a>. Two other file formats can be used, but require additional steps detailed <a class="reference internal" href="#other-file-formats"><em>here</em></a>. These steps convert other common formats to fastq, allowing you to begin processing your data with <tt class="docutils literal"><span class="pre">split_libraries_fastq.py</span></tt>.</p>
<p>To get information on the full set of parameters that can be passed to <tt class="docutils literal"><span class="pre">split_libraries_fastq.py</span></tt>, review the <a class="reference external" href="../scripts/split_libraries_fastq.html">script documentation here</a>.</p>
</div>
<div class="section" id="demultiplexing-illumina-fastq-with-qiime">
<h2>Demultiplexing Illumina fastq with QIIME<a class="headerlink" href="#demultiplexing-illumina-fastq-with-qiime" title="Permalink to this headline">¶</a></h2>
<div class="section" id="demultiplexing-and-quality-filtering-one-lane-of-illumina-fastq-reads-with-qiime">
<h3>Demultiplexing and quality filtering one lane of Illumina fastq reads with QIIME<a class="headerlink" href="#demultiplexing-and-quality-filtering-one-lane-of-illumina-fastq-reads-with-qiime" title="Permalink to this headline">¶</a></h3>
<p>To demultiplex your reads you&#8217;ll need to know which of your fastq files corresponds to your barcodes, which corresponds to your amplicons, and you&#8217;ll also need a metadata mapping file (format described <a class="reference external" href="../documentation/file_formats.html#metadata-mapping-files">here</a>). To determine which are your barcode reads and which are your amplicon reads you can look at the first 4 lines of each file with the <tt class="docutils literal"><span class="pre">head</span></tt> command. For example:</p>
<div class="highlight-python"><pre>head -n 4 s_8_2_sequence.fastq</pre>
</div>
<p>Will give you the first fastq record in s_8_2_sequence.fastq. If the length of the sequence (on the second line) corresponds to the length of your barcode, then this is your barcode read file. If not, check your other fastq file. Note: due to a sequencing artifact the barcode reads may be a single base longer than your actual barcodes.</p>
<p>To demultiplex and quality filter (details on the quality filtering <a class="reference internal" href="#illumina-quality-filtering"><em>here</em></a>) your fastq data, run the following command:</p>
<div class="highlight-python"><pre>split_libraries_fastq.py -i s_8_2_sequence.fastq -o sl_out/ -b s_8_1_sequence.fastq -m s_8_map.txt</pre>
</div>
<p>In this example <tt class="docutils literal"><span class="pre">s_8_2_sequence.fastq</span></tt> contains my amplicon reads and <tt class="docutils literal"><span class="pre">s_8_1_sequence.fastq</span></tt> contains my barcode reads. My metadata mapping file is <tt class="docutils literal"><span class="pre">s_8_map.txt</span></tt>, and the output will be written to <tt class="docutils literal"><span class="pre">./sl_out/</span></tt>. This command can take tens of minutes to a couple of hours to run, depending on your computer.</p>
<p>When the command completes you&#8217;ll have three files in <tt class="docutils literal"><span class="pre">sl_out</span></tt>:</p>
<blockquote>
<ul class="simple">
<li>seqs.fna : Your demultiplexed sequences.</li>
<li>split_libraries_log.txt : contains details on how many samples were assigned to each barcode, as well as details about sequences that were quality filter. It&#8217;s a good idea to review this file after each run to confirm that the results look as expected.</li>
<li>histograms.txt : A histogram illustrating the distribution of sequence lengths after quality filtering/truncation.</li>
</ul>
</blockquote>
<p>After reviewing your split_libraries_log.txt you may choose to rerun with adjusted quality filter parameters, or continue to OTU picking. To continue to OTU picking move on to the next section.</p>
</div>
<div class="section" id="demultiplexing-and-quality-filtering-multiple-lanes-of-illumina-fastq-reads-with-qiime">
<h3>Demultiplexing and quality filtering multiple lanes of Illumina fastq reads with QIIME<a class="headerlink" href="#demultiplexing-and-quality-filtering-multiple-lanes-of-illumina-fastq-reads-with-qiime" title="Permalink to this headline">¶</a></h3>
<p>If you have multiple lanes of Illumina data that you want to demultiplex together, you&#8217;ll use a similar command as for a single lane, but will specify per-lane amplicon read, barcode read, and mapping files. To do that, call the same command with comma-separated filepaths. Note that the order of the files must correspond for each of these parameters. For example, to demultiplex data from lanes 6, 7, and 8, your command would be:</p>
<div class="highlight-python"><pre>split_libraries_fastq.py -i s_6_2_sequence.fastq,s_7_2_sequence.fastq,s_8_2_sequence.fastq -o sl_out/ -b s_6_1_sequence.fastq,s_7_1_sequence.fastq,s_8_1_sequence.fastq -m s_6_map.txt,s_7_map.txts_8_map.txt</pre>
</div>
<p>When the command completes you&#8217;ll have the same output as when running on a single lane, but note that the data in the log and histogram files are broken down by lane. All of the sequence data will be in <tt class="docutils literal"><span class="pre">seqs.fna</span></tt>.</p>
<p>Again, you may wish to review <tt class="docutils literal"><span class="pre">split_libraries_log.txt</span></tt> and adjust quality filtering parameters and rerun. When you&#8217;re satisfied, you&#8217;re read to move on to downstream analysis.</p>
</div>
</div>
<div class="section" id="using-illumina-data-in-downstream-analysis">
<h2>Using Illumina data in downstream analysis<a class="headerlink" href="#using-illumina-data-in-downstream-analysis" title="Permalink to this headline">¶</a></h2>
<p>Once you have processed your Illumina data with <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>, you can use the resulting files in all downstream QIIME scripts, including the workflow scripts.</p>
<p>A common way that we have processed Illumina data is using a closed reference OTU picking protocol, where OTUs are picked against a reference data set, and any reads that do not match a reference sequence at greater than or equal to 97% sequence identity are discarded. This has the benefit of providing a very strict quality filter (while studies to better understand quality filtering of Illumina amplicon data are in progress), but has the draw back of throwing away real sequences that do not match to know sequences.</p>
<p>To run a closed-reference OTU picking process on your demultiplexed Illumina data, you can use the <a class="reference external" href="../scripts/pick_closed_reference_otus.html">pick_closed_reference_otus.py</a> workflow. You can use your own reference database, or our greengenes OTUs reference set. To grab the latest version of these, click the &#8220;Most recent Greengenes OTUs&#8221; on the top-right corner of the <a class="reference external" href="http://blog.qiime.org">QIIME blog homepage</a>. As of this writing the latest version is 4feb2011, but this is subject to change (hence not linking it directly from this page). Download this file and unzip it. Following from the steps above you can then pick reference otus on your demutliplexed data with the following command:</p>
<div class="highlight-python"><pre>pick_closed_reference_otus.py -i sl_out/seqs.fna -o ./ucrC/ -r gg_otus_4feb2011/rep_set/gg_97_otus_4feb2011.fasta -t gg_otus_4feb2011/taxonomies/greengenes_tax.txt</pre>
</div>
<p>In this example I am picking OTUs on <tt class="docutils literal"><span class="pre">sl_out/seqs.fna</span></tt> against the greengenes 97% reference OTU collection. The output is being written to <tt class="docutils literal"><span class="pre">./ucrC/</span></tt> (which stands for uclust_ref with the -C parameter, meaning &#8216;closed reference&#8217;). Taxonomy strings from the Greengenes taxonomy is associated with each read based on their best match (determined by uclust_ref) in the reference set.</p>
<p>This step will generate an OTU table, which is the input for a lot of the analyses possible with QIIME. For example, to generate interactive 3D UniFrac PCoA plots, you would run the command:</p>
<div class="highlight-python"><pre>beta_diversity_through_plots.py -i ucrC/uclust_ref_picked_otus/otu_table.biom -o bdiv/ -t gg_otus_4feb2011/trees/gg_97_otus_4feb2011.tre -m ./s_8_map.txt</pre>
</div>
<p>Note that because we picked OTUs against a reference set, we can use the reference set phylogenetic tree for the UniFrac analysis. That is passed with <tt class="docutils literal"><span class="pre">-t</span></tt> in this example. To visualize the 3D UniFrac PCoA plots, you can open the <tt class="docutils literal"><span class="pre">bdiv/unweighted_unifrac_3d_continuous/unweighted_unifrac_pc_3D_PCoA_plots.html</span></tt> file that is generated in this analysis. This will launch the KiNG applet, and your 3D plots. These may take a little while to load depending on the quantity of data you have. (Improving these visualizations is something we&#8217;re currently working on.)</p>
</div>
<div class="section" id="processing-non-fastq-illumina-data-with-qiime">
<span id="other-file-formats"></span><h2>Processing non-fastq Illumina data with QIIME<a class="headerlink" href="#processing-non-fastq-illumina-data-with-qiime" title="Permalink to this headline">¶</a></h2>
<p>QIIME supports several formats of non-fastq data, but the strategy is to convert from these formats to fastq. For that reason your analyses will be more convenient if you can get your sequencing center to provide data in fastq format (as supported by the Illumina CASAVA software).</p>
<div class="section" id="processing-qseq-files-with-qiime">
<h3>Processing qseq files with QIIME<a class="headerlink" href="#processing-qseq-files-with-qiime" title="Permalink to this headline">¶</a></h3>
<p>You can convert qseq files to fastq files using the <a class="reference external" href="../scripts/process_qseq.html">process_qseq.py</a> script.</p>
<p>Example qseq file for amplicon read:</p>
<div class="highlight-python"><pre>M10     68      1       1       28680   29475   0       1       AACGAAAGGCAGTTTTGGAAGTAGGCGAATTAGGGTAACGCATATAGGATGCTAATACAACGTGAATGAAGTACTGCATCTATGTCACCAGCTTATTACAGCAGCTTGTCATACATGGCCGTACAGGAAACACACATCATAGCATCACACGA        BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB        0
M10     68      1       1       19607   29475   0       1       GACATAAGGGTGGTTAGTATACCGGCAAGGACGGGGTTACTAGTGACGTCCTTCCCCGTATGCCGGGCAATAATGTTTATGTTGGTTTCATGGTTTGGTCTAACTTTACCGCTACTAAATGCTGCGGATTGGTTTCGCTGAATCAGATTATT        Z__c\JQ`cc[[_[bfff[[`Qbdge_YYOOHO^cF[FUb_VHMHV`T`dBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB        1
M10     68      1       1       22962   29475   0       1       TAATCGAGCTCAACGCCCTGCATACGAAAAGACAGAATCTCTTGCAAGATGTTGGTGCGGTTAGCCAGCTGCTTATGGAAGCCAAGCATTGGGGATTGAGAAAGAGTAGAAATGCCACAAGCCTCAATAGCAGGTTTAAGAGCCTCGATACG        JJY````JO[`bab`b`bbaaaaa`\`a`OVT``]]`aa^aI\HMMMWWHHNNNGLL\`________\Z^]]^^^^^^GX]\QTXXZ[YZ^^XZ[Z^\Z^GW\^^\\^^^VZ\Y^^^^\\\\[^[\\\^VWYWWXWWZYZW^[X^\\Z^[TQ        0</pre>
</div>
<p>Example qseq file for barcode read:</p>
<div class="highlight-python"><pre>M10     68      1       1       28680   29475   0       2       ACTCACGGTATTA   \_J\Sa^Y[ZYK`   0
M10     68      1       1       19607   29475   0       2       AGACTGAGTACTA   PP\JJ\JQ`\RK^   1
M10     68      1       1       22962   29475   0       2       AGACGTGCAATTA   ^_aecceeeQ`[b   0</pre>
</div>
<p>You&#8217;ll need to know which of your reads files correspond to your barcodes and which correspond to your amplicons. To determine this you can look at the first line of representative files with the <tt class="docutils literal"><span class="pre">head</span></tt> command. For example:</p>
<div class="highlight-python"><pre>head -n 1 s_1_1_0001_qseq.txt</pre>
</div>
<p>Will give you the first line s_1_1_0001_qseq.txt. If the length of the sequence (the 9th field) corresponds to the length of your barcode, then this is your barcode read file. If not, check a qseq file corresponding to another read number (e.g., s_1_2_0001_qseq.txt). Note that due to technical artifacts you may sometimes have a single extra base here, so for a length 12 barcode your sequence may be length 13.</p>
<p>You&#8217;ll typically start here with a directory containing many qseq files. The process_qseq.py script therefore works on a directory, rather than a set of input files. In my example, the read 1 files correspond to my sequence reads and the read 2 files correspond to my barcode reads. To generate a single fastq file for the sequence reads from the qseq files, you can run the command:</p>
<div class="highlight-python"><pre>process_qseq.py -i ./ -o ./fastq/ -r 1</pre>
</div>
<p>This specifies that the qseq files are in the current directory (<tt class="docutils literal"><span class="pre">-i</span></tt>), and the fastq should be written to <tt class="docutils literal"><span class="pre">./fastq/</span></tt>. The <tt class="docutils literal"><span class="pre">-r</span> <span class="pre">1</span></tt> specifies that I want to process the read one files (i.e., my amplicon reads).</p>
<p>To generate the barcode read fastq file you can run the following command:</p>
<div class="highlight-python"><pre>process_qseq.py -i ./ -o ./fastq/ -r 2 -b 12</pre>
</div>
<p>This again specifies that the qseq files are in the current directory (<tt class="docutils literal"><span class="pre">-i</span></tt>), and the fastq should be written to <tt class="docutils literal"><span class="pre">./fastq/</span></tt>. The <tt class="docutils literal"><span class="pre">-r</span> <span class="pre">2</span></tt> specifies that I want to process the read two files (i.e., my barcode reads), and the <tt class="docutils literal"><span class="pre">-b</span> <span class="pre">12</span></tt> specifies that I only want to extract the first twelve bases of these reads.</p>
<p>Once these steps are complete you&#8217;ll have fastq files that can be passed to split_libraries_fastq.py.</p>
</div>
<div class="section" id="processing-iseq-files-with-qiime">
<h3>Processing iseq files with QIIME<a class="headerlink" href="#processing-iseq-files-with-qiime" title="Permalink to this headline">¶</a></h3>
<p>You can convert iseq files to fastq files using the <a class="reference external" href="../scripts/process_iseq.html">process_iseq.py</a> script. Determine which of the following file types you have, and call the corresponding command.</p>
<p>Example iseq with barcode in sequence (more common):</p>
<div class="highlight-python"><pre>HWI-ST753_50:6:1101:15435:9071#0/1:ACCAGACGATGCTACGGAGGGAGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCACGTAGGCGGCTTTGTAAGTTAGAGGTGAAAGCCTGGAGCTCAAC:gggggggfggdegggggggggggggggggggegggggggggegggggggeggcccccFUZSU_]]^^ggggggdggdgeeeccYacadcbeddceegggeeg
HWI-ST753_50:6:1101:15446:9128#0/1:AGCTTAACAGCTTACGTAGGGGGCAAGCGTTATCCGGAATTACTGGGTGTAAAGGGAGCGCAGACGGAGAGGCAAGTCAGCTGTGAAAACTCCAGGCTTAAC:BBBBBBBBBBBB`_```_I^HM^`__`____I^^_`_`N``_______`__`___`_\_`G_^L^^^FDJTI^^^ZW^G^BBBBBBBBBBBBBBBBBBBBBB
HWI-ST753_50:6:1101:15300:9134#0/1:ACCAGACGATGCTACGTAGGGGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGCGCGTGTAGGCGGCCAGGTAGGTCCGTTGTGAAAACTGGAGGCTTAAC:gggggggggcgcggggegggggeggfgggggggggggggggfggggggggggffMffa^cbbgggggggeggdedfb`dfeee`db^fffffge\geggdfg</pre>
</div>
<p>To generate fastq from iseq files with tweleve base barcodes contained as the first bases of the sequence, call the following command:</p>
<div class="highlight-python"><pre>process_iseq.py -i s_6_1_sequences.txt,s_7_1_sequences.txt -o ./fastq/ -b 12</pre>
</div>
<p>Example iseq with barcode in header (less common):</p>
<div class="highlight-python"><pre>HWI-6X_9267:1:1:12:410#ACAGCTA/1:TACGTAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGCATTTTAAGCCAGACGTGAAATCCCCGGGCTTAACCTGGGAACTG:abbb`aaa`^aa```ba`aaaabaaaabaaaa^[Y]^__a`abb`aaaa]Y\\_a[Y_a`a```a__]aaXT\`^\_]`a^^WSZ\JNY]^a`ORO^^`Y
HWI-6X_9267:1:1:12:1762#ACATGAT/1:GACGGAGGATGCAAGTGTTATCCGGAATCACTGGGCGTAAAGCGTCTGTAGGTTGTTTGATAAGTCAACTGTTAAATCTTGAAGCTCAACTTCAAAATCG:aaaaaaaaabaaaaa_aaaaaa`aaaaaaaa`aa``a]aa```a^a^`\```\a`^aaa_\__]]_a_``^``a^^a^b[`SJN]Y_ZZ]^W___`_^U[
HWI-6X_9267:1:1:12:1872#ACAGTTG/1:TACGGAGGGGGTTAGCGTTGTTCCGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGAAAGTTGGGGGGGAAATCCCGGGGCTCAACCCCGGACGTG:aaaaa_aaaa`[a_a`aaaa]a[MY``a\a`aaaaa_\]_\__[_]W]^[[U]aXRZ\W[J\KVTX]\YZZDVY]SUBBBBBBBBBBBBBBBBBBBBBBB</pre>
</div>
<p>To generate fastq from iseq files with six base barcodes contained in the index field of the header, call the following command:</p>
<div class="highlight-python"><pre>process_iseq.py -i s_6_1_sequences.txt,s_7_1_sequences.txt -o ./fastq/ --barcode_length 6 --barcode_in_header</pre>
</div>
<p>Note that in the second example there are actually seven bases in the index field. If only six correspond to your barcode (and the remaining bases in e.g. a technical artifact) you can specify &#8211;barcode_length 6 (as done here) to extract only the first six bases of the barcode.</p>
<p>Once these steps are complete you&#8217;ll have fastq files that can be passed to split_libraries_fastq.py.</p>
</div>
</div>
<div class="section" id="other-topics">
<h2>Other topics<a class="headerlink" href="#other-topics" title="Permalink to this headline">¶</a></h2>
<div class="section" id="example-fastq-format-qiime-default">
<span id="fastq-format"></span><h3>Example fastq format (QIIME default)<a class="headerlink" href="#example-fastq-format-qiime-default" title="Permalink to this headline">¶</a></h3>
<p>Example of amplicon read fastq:</p>
<div class="highlight-python"><pre>@M10_68:1:1:28680:29475#0/1
AACGAAAGGCAGTTTTGGAAGTAGGCGAATTAGGGTAACGCATATAGGATGCTAATACAACGTGAATGAAGTACTGCATCTATGTCACCAGCTTATTACAGCAGCTTGTCATACATGGCCGTACAGGAAACACACATCATAGCATCACACGA
+
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@M10_68:1:1:19607:29475#0/1
GACATAAGGGTGGTTAGTATACCGGCAAGGACGGGGTTACTAGTGACGTCCTTCCCCGTATGCCGGGCAATAATGTTTATGTTGGTTTCATGGTTTGGTCTAACTTTACCGCTACTAAATGCTGCGGATTGGTTTCGCTGAATCAGATTATT
+
Z__c\JQ`cc[[_[bfff[[`Qbdge_YYOOHO^cF[FUb_VHMHV`T`dBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@M10_68:1:1:22962:29475#0/1
TAATCGAGCTCAACGCCCTGCATACGAAAAGACAGAATCTCTTGCAAGATGTTGGTGCGGTTAGCCAGCTGCTTATGGAAGCCAAGCATTGGGGATTGAGAAAGAGTAGAAATGCCACAAGCCTCAATAGCAGGTTTAAGAGCCTCGATACG
+
JJY````JO[`bab`b`bbaaaaa`\`a`OVT``]]`aa^aI\HMMMWWHHNNNGLL\`________\Z^]]^^^^^^GX]\QTXXZ[YZ^^XZ[Z^\Z^GW\^^\\^^^VZ\Y^^^^\\\\[^[\\\^VWYWWXWWZYZW^[X^\\Z^[TQ</pre>
</div>
<p>Example of corresponding barcode read fastq:</p>
<div class="highlight-python"><pre>@M10_68:1:1:28680:29475#0/2
ACTCACGGTATT
+
\_J\Sa^Y[ZYK
@M10_68:1:1:19607:29475#0/2
AGACTGAGTACT
+
PP\JJ\JQ`\RK
@M10_68:1:1:22962:29475#0/2
AGACGTGCAATT
+
^_aecceeeQ`[</pre>
</div>
</div>
<div class="section" id="quality-filtering-of-illumina-data-with-qiime">
<span id="illumina-quality-filtering"></span><h3>Quality filtering of Illumina data with QIIME<a class="headerlink" href="#quality-filtering-of-illumina-data-with-qiime" title="Permalink to this headline">¶</a></h3>
<p>A sequence is discarded if any of the following conditions are met:</p>
<blockquote>
<ul class="simple">
<li>The sequence contains one or more <tt class="docutils literal"><span class="pre">N</span></tt> bases, corresponding to ambiguous base calls (adjustable with the -n parameter).</li>
<li>The high-quality region of the sequence is less than 75 bases long (adjustable with the <tt class="docutils literal"><span class="pre">-p</span></tt> parameter), where high-quality regions is defined a stretch of bases containing no more than 1 (adjustable with the <tt class="docutils literal"><span class="pre">-r</span></tt> parameter) quality character less than <tt class="docutils literal"><span class="pre">B</span></tt> (i.e., any of <tt class="docutils literal"><span class="pre">&#64;</span></tt>, <tt class="docutils literal"><span class="pre">A</span></tt>, and <tt class="docutils literal"><span class="pre">B</span></tt> are considered to be low quality scores, adjustable with the <tt class="docutils literal"><span class="pre">-q</span></tt> parameter). In other words, with the default parameter settings, the read is truncated at the base preceding the first low quality stretch, and the truncated sequence must be greater than or equal to 75 bases long to be retained.</li>
<li>If barcode error correction is disabled and the barcode is not an exact match to a barcode in the mapping file (to disable this, pass <tt class="docutils literal"><span class="pre">-u</span></tt>, which will cause the resulting sequences to be store with sample ID <tt class="docutils literal"><span class="pre">Unassigned</span></tt>.)</li>
<li>If barcode error correction is enabled and the barcode is not correctable or within <tt class="docutils literal"><span class="pre">max_barcode_errors</span></tt> (specified on the command line with <tt class="docutils literal"><span class="pre">--max_barcode_errors</span></tt>) of a good barcode.</li>
</ul>
</blockquote>
</div>
<div class="section" id="processing-paired-end-read-data-with-qiime">
<h3>Processing paired-end read data with QIIME<a class="headerlink" href="#processing-paired-end-read-data-with-qiime" title="Permalink to this headline">¶</a></h3>
<p>QIIME can be used to parse single-end or paired-end read data from the Illumina platform. The downstream support for analysis of paired-end read data is currently limited. The parsed output is in standard fasta format, so all scripts (such as align_seqs.py and assign_taxonomy.py) can read it. However because there may be a &#8216;big gap&#8217; between the 5&#8217; and 3&#8217; reads if the primers are distant in the sequence, or conversely because the reads may overlap if the primers are close, the reads are written to separate fasta files in separate runs. It is up to the user to merge these into a single file depending on how they wish to process the data (e.g., assemble over-lapping reads into contigs).</p>
<p>If specific use cases become popular we will likely add support for them in QIIME. If you&#8217;re interested in getting a specific workflow implemented you can contact us on the <a class="reference external" href="http://forum.qiime.org">QIIME Forum</a>. The information we&#8217;ll be interested in is an explanation of the workflow, and evidence that using paired-end reads improves results over using single-end reads alone. We are happy to share raw paired-end read Illumina data to facilitate such analyses.</p>
<p>To analyze demultiplex paired-end read data, run the split_libraries_fastq.py script on each read file independently. You can use the <tt class="docutils literal"><span class="pre">--rev_comp</span></tt> option on the reverse (3 prime) reads to reverse complement the reads so they&#8217;ll be in the same orientation as the forward (5 prime) reads, if this is desired.</p>
</div>
<div class="section" id="barcode-decoding">
<h3>Barcode decoding<a class="headerlink" href="#barcode-decoding" title="Permalink to this headline">¶</a></h3>
<p>QIIME currently supports length 12 golay error-correcting barcodes as the default barcode type for <tt class="docutils literal"><span class="pre">split_libraries_fastq.py</span></tt>. Non-golay codes of length 12, or barcodes of other lengths can be used by passing the length of the barcode with the <tt class="docutils literal"><span class="pre">--barcode_type</span></tt> option. No error recovery will be attempted in this case.</p>
<p>Barcode decoding slows down demultiplexing by a factor of approximately two (determined on a single HiSeq lane), but will be affected by how many barcodes actually need to be decoded. On this same HiSeq lane, barcode decoding recovered approximately 700,000 sequences (of approximately 60,000,000 input sequences) which had erroneous, correctable barcodes.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Preparing raw Illumina data in different formats for use with QIIME</a><ul>
<li><a class="reference internal" href="#input-file-formats">Input file formats</a></li>
<li><a class="reference internal" href="#demultiplexing-illumina-fastq-with-qiime">Demultiplexing Illumina fastq with QIIME</a><ul>
<li><a class="reference internal" href="#demultiplexing-and-quality-filtering-one-lane-of-illumina-fastq-reads-with-qiime">Demultiplexing and quality filtering one lane of Illumina fastq reads with QIIME</a></li>
<li><a class="reference internal" href="#demultiplexing-and-quality-filtering-multiple-lanes-of-illumina-fastq-reads-with-qiime">Demultiplexing and quality filtering multiple lanes of Illumina fastq reads with QIIME</a></li>
</ul>
</li>
<li><a class="reference internal" href="#using-illumina-data-in-downstream-analysis">Using Illumina data in downstream analysis</a></li>
<li><a class="reference internal" href="#processing-non-fastq-illumina-data-with-qiime">Processing non-fastq Illumina data with QIIME</a><ul>
<li><a class="reference internal" href="#processing-qseq-files-with-qiime">Processing qseq files with QIIME</a></li>
<li><a class="reference internal" href="#processing-iseq-files-with-qiime">Processing iseq files with QIIME</a></li>
</ul>
</li>
<li><a class="reference internal" href="#other-topics">Other topics</a><ul>
<li><a class="reference internal" href="#example-fastq-format-qiime-default">Example fastq format (QIIME default)</a></li>
<li><a class="reference internal" href="#quality-filtering-of-illumina-data-with-qiime">Quality filtering of Illumina data with QIIME</a></li>
<li><a class="reference internal" href="#processing-paired-end-read-data-with-qiime">Processing paired-end read data with QIIME</a></li>
<li><a class="reference internal" href="#barcode-decoding">Barcode decoding</a></li>
</ul>
</li>
</ul>
</li>
</ul>


	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
	<li><a href="../index.html">Home</a><br /></li>
	<li><a href="../install/index.html">Install</a><br /></li>
	<li><a href="../documentation/index.html">Documentation</a><br /></li>
	<li><a href="index.html">Tutorials</a><br /></li>
	<li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	<li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.pycogent.org" target="_blank"><img src="../_static/PoweredByPycogent.jpg" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2012, QIIME Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.4.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>