Deleted: app/bin/runner
===================================================================
--- app/bin/runner	2013-05-31 16:43:09 UTC (rev 2958)
+++ app/bin/runner	2013-05-31 17:54:22 UTC (rev 2959)
@@ -1,4 +0,0 @@
-#!/usr/bin/env python
-import sys
-from app.cli import main
-sys.exit(main())

Modified: app/setup.py
===================================================================
--- app/setup.py	2013-05-31 16:43:09 UTC (rev 2958)
+++ app/setup.py	2013-05-31 17:54:22 UTC (rev 2959)
@@ -80,7 +80,7 @@
   'Shapely                  == 1.2.17',
 
   # message queue
-  # 'my-app                   == x.x.x',
+  'my-app                   == 1.2.3',
   'APScheduler              == 2.1.0',
   'kombu                    == 2.5.10',
   'amqp                     == 1.0.11',

Modified: app/lib/python/MANIFEST
===================================================================
--- app/lib/python/MANIFEST	2013-05-31 16:43:09 UTC (rev 2958)
+++ app/lib/python/MANIFEST	2013-05-31 17:54:22 UTC (rev 2959)
@@ -68,4 +68,5 @@
 https://pypi.python.org/packages/source/G/GeoAlchemy2/GeoAlchemy2-0.2.tar.gz#md5=14d73d09cdc47e3ed92e804883e61218
 https://pypi.python.org/packages/source/t/transaction/transaction-1.4.1.zip#md5=8db2680bc0f999219861a67b8f335a88
 https://pypi.python.org/packages/source/p/psycopg2/psycopg2-2.5.tar.gz#md5=facd82faa067e99b80146a0ee2f842f6
-https://pypi.python.org/packages/source/p/pyramid/pyramid-1.4.1.tar.gz#md5=044d42f609d567d7db2948a03fffcf7c
\ No newline at end of file
+https://pypi.python.org/packages/source/p/pyramid/pyramid-1.4.1.tar.gz#md5=044d42f609d567d7db2948a03fffcf7c
+https://pypi.python.org/packages/source/m/my_app/my_app-1.2.3.tar.gz
\ No newline at end of file

Added: app/lib/python/my_app-1.2.3.tar.gz
===================================================================
(Binary files differ)


Property changes on: app/lib/python/my_app-1.2.3.tar.gz
___________________________________________________________________
Added: svn:mime-type
   + application/octet-stream

