

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>QIIME scripts &mdash; Homepage</title>
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.7.0-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="qiime-scripts">
<span id="index"></span><h1>QIIME scripts<a class="headerlink" href="#qiime-scripts" title="Permalink to this headline">Â¶</a></h1>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="add_alpha_to_mapping_file.html"><em>add_alpha_to_mapping_file.py</em> &#8211; Add alpha diversity data to a metadata mapping file</a></li>
<li class="toctree-l1"><a class="reference internal" href="add_qiime_labels.html"><em>add_qiime_labels.py</em> &#8211; Takes a directory, a metadata mapping file, and a column name that contains the fasta file names that SampleIDs are associated with, combines all files that have valid fasta extensions into a single fasta file, with valid QIIME fasta labels.</a></li>
<li class="toctree-l1"><a class="reference internal" href="adjust_seq_orientation.html"><em>adjust_seq_orientation.py</em> &#8211; Get the reverse complement of all sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="align_seqs.html"><em>align_seqs.py</em> &#8211; Align sequences using a variety of alignment methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="alpha_diversity.html"><em>alpha_diversity.py</em> &#8211; Calculate alpha diversity on each sample in an otu table, using a variety of alpha diversity metrics</a></li>
<li class="toctree-l1"><a class="reference internal" href="alpha_diversity_metrics.html"><em>alpha_diversity_metrics</em> &#8211; List of available metrics</a></li>
<li class="toctree-l1"><a class="reference internal" href="alpha_rarefaction.html"><em>alpha_rarefaction.py</em> &#8211; A workflow script for performing alpha rarefaction</a></li>
<li class="toctree-l1"><a class="reference internal" href="ampliconnoise.html"><em>ampliconnoise.py</em> &#8211; Run AmpliconNoise</a></li>
<li class="toctree-l1"><a class="reference internal" href="assign_taxonomy.html"><em>assign_taxonomy.py</em> &#8211; Assign taxonomy to each sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_diversity.html"><em>beta_diversity.py</em> &#8211; Calculate beta diversity (pairwise sample dissimilarity) on one or many otu tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_diversity_metrics.html"><em>beta_diversity_metrics</em> &#8211; List of available metrics</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_diversity_through_plots.html"><em>beta_diversity_through_plots.py</em> &#8211; A workflow script for computing beta diversity distance matrices and generating PCoA plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_significance.html"><em>beta_significance.py</em> &#8211; This script runs any of a set of common tests to determine if a sample is statistically significantly different from another sample</a></li>
<li class="toctree-l1"><a class="reference internal" href="blast_wrapper.html"><em>blast_wrapper.py</em> &#8211; Blast Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="categorized_dist_scatterplot.html"><em>categorized_dist_scatterplot.py</em> &#8211; makes a figure representing average distances between samples, broken down by categories. I call it a &#8216;categorized distance scatterplot&#8217;</a></li>
<li class="toctree-l1"><a class="reference internal" href="check_id_map.html"><em>check_id_map.py</em> &#8211; Checks user&#8217;s metadata mapping file for required data, valid format</a></li>
<li class="toctree-l1"><a class="reference internal" href="clean_raxml_parsimony_tree.html"><em>clean_raxml_parsimony_tree.py</em> &#8211; Remove duplicate tips from Raxml Tree</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster_quality.html"><em>cluster_quality.py</em> &#8211; compute the quality of a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="collate_alpha.html"><em>collate_alpha.py</em> &#8211; Collate alpha diversity results</a></li>
<li class="toctree-l1"><a class="reference internal" href="compare_3d_plots.html"><em>compare_3d_plots.py</em> &#8211; Plot several PCoA files on the same 3D plot</a></li>
<li class="toctree-l1"><a class="reference internal" href="compare_alpha_diversity.html"><em>compare_alpha_diversity.py</em> &#8211; This script compares alpha diversities based on a two-sample t-test using either parametric or non-parametric (Monte Carlo) methods.</a></li>
<li class="toctree-l1"><a class="reference internal" href="compare_categories.html"><em>compare_categories.py</em> &#8211; Analyzes statistical significance of sample groupings using distance matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="compare_distance_matrices.html"><em>compare_distance_matrices.py</em> &#8211; Computes Mantel correlation tests between sets of distance matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="compare_taxa_summaries.html"><em>compare_taxa_summaries.py</em> &#8211; Compares taxa summary files</a></li>
<li class="toctree-l1"><a class="reference internal" href="compute_core_microbiome.html"><em>compute_core_microbiome.py</em> &#8211; Identify the core microbiome.</a></li>
<li class="toctree-l1"><a class="reference internal" href="conditional_uncovered_probability.html"><em>conditional_uncovered_probability.py</em> &#8211; Calculate the conditional uncovered probability on each sample in an otu table.</a></li>
<li class="toctree-l1"><a class="reference internal" href="consensus_tree.html"><em>consensus_tree.py</em> &#8211; This script outputs a majority consensus tree given a collection of input trees.</a></li>
<li class="toctree-l1"><a class="reference internal" href="convert_fastaqual_fastq.html"><em>convert_fastaqual_fastq.py</em> &#8211; From a FASTA file and a matching QUAL file, generates a FASTQ file. From FASTQ file generates FASTA file and  matching QUAL file.</a></li>
<li class="toctree-l1"><a class="reference internal" href="convert_fastaqual_to_fastq.html"><em>convert_fastaqual_to_fastq.py</em> &#8211; From a FASTA file and a matching QUAL file, generates a minimal FASTQ file.</a></li>
<li class="toctree-l1"><a class="reference internal" href="core_diversity_analyses.html"><em>core_diversity_analyses.py</em> &#8211; A workflow for running a core set of QIIME diversity analyses.</a></li>
<li class="toctree-l1"><a class="reference internal" href="count_seqs.html"><em>count_seqs.py</em> &#8211;</a></li>
<li class="toctree-l1"><a class="reference internal" href="cytoscape_usage.html"><em>cytoscape_usage</em> &#8211; Visualizing Results with Cytoscape</a></li>
<li class="toctree-l1"><a class="reference internal" href="demultiplex_fasta.html"><em>demultiplex_fasta.py</em> &#8211; Demultiplex fasta data according to barcode sequences or data supplied in fasta labels.</a></li>
<li class="toctree-l1"><a class="reference internal" href="denoise_wrapper.html"><em>denoise_wrapper.py</em> &#8211; Denoise a flowgram file</a></li>
<li class="toctree-l1"><a class="reference internal" href="denoiser.html"><em>denoiser.py</em> &#8211; Remove noise from  454 sequencing data</a></li>
<li class="toctree-l1"><a class="reference internal" href="denoiser_preprocess.html"><em>denoiser_preprocess.py</em> &#8211; Run phase of denoiser algorithm: prefix clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="denoiser_worker.html"><em>denoiser_worker.py</em> &#8211; Start a denoiser worker process</a></li>
<li class="toctree-l1"><a class="reference internal" href="detrend.html"><em>detrend.py</em> &#8211; Detrend Principal Coordinates</a></li>
<li class="toctree-l1"><a class="reference internal" href="dissimilarity_mtx_stats.html"><em>dissimilarity_mtx_stats.py</em> &#8211; Calculate mean, median and standard deviation from a set of distance matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="distance_matrix_from_mapping.html"><em>distance_matrix_from_mapping.py</em> &#8211; Calculate the pairwise dissimilarity on one column of a mappping file</a></li>
<li class="toctree-l1"><a class="reference internal" href="exclude_seqs_by_blast.html"><em>exclude_seqs_by_blast.py</em> &#8211; Exclude contaminated sequences using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="extract_seqs_by_sample_id.html"><em>extract_seqs_by_sample_id.py</em> &#8211; Extract sequences based on the SampleID</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_alignment.html"><em>filter_alignment.py</em> &#8211; Filter sequence alignment by removing highly variable regions</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_distance_matrix.html"><em>filter_distance_matrix.py</em> &#8211; Filter a distance matrix to contain only a specified set of samples.</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_fasta.html"><em>filter_fasta.py</em> &#8211; This script can be applied to remove sequences from a fasta or fastq file based on input criteria.</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_otus_by_sample.html"><em>filter_otus_by_sample.py</em> &#8211; Filter OTU mapping file and sequences by SampleIDs</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_otus_from_otu_table.html"><em>filter_otus_from_otu_table.py</em> &#8211; Filter OTUs from an OTU table based on their observation counts or identifier.</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_samples_from_otu_table.html"><em>filter_samples_from_otu_table.py</em> &#8211; Filters samples from an OTU table on the basis of the number of observations in that sample, or on the basis of sample metadata. Mapping file can also be filtered to the resulting set of sample ids.</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_taxa_from_otu_table.html"><em>filter_taxa_from_otu_table.py</em> &#8211; Filter taxa from an OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_tree.html"><em>filter_tree.py</em> &#8211; This script prunes a tree based on a set of tip names</a></li>
<li class="toctree-l1"><a class="reference internal" href="fix_arb_fasta.html"><em>fix_arb_fasta.py</em> &#8211; Reformat ARB FASTA files</a></li>
<li class="toctree-l1"><a class="reference internal" href="identify_chimeric_seqs.html"><em>identify_chimeric_seqs.py</em> &#8211; Identify chimeric sequences in input FASTA file</a></li>
<li class="toctree-l1"><a class="reference internal" href="identify_missing_files.html"><em>identify_missing_files.py</em> &#8211; This script checks for the existence of expected files in parallel runs.</a></li>
<li class="toctree-l1"><a class="reference internal" href="inflate_denoiser_output.html"><em>inflate_denoiser_output.py</em> &#8211; Inflate denoiser results so they can be passed directly to OTU pickers.</a></li>
<li class="toctree-l1"><a class="reference internal" href="insert_seqs_into_tree.html"><em>insert_seqs_into_tree.py</em> &#8211; Tree Insertion</a></li>
<li class="toctree-l1"><a class="reference internal" href="jackknifed_beta_diversity.html"><em>jackknifed_beta_diversity.py</em> &#8211; A workflow script for performing jackknifed UPGMA clustering and build jackknifed 2d and 3D PCoA plots.</a></li>
<li class="toctree-l1"><a class="reference internal" href="load_remote_mapping_file.html"><em>load_remote_mapping_file.py</em> &#8211; Downloads and saves a remote mapping file</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_2d_plots.html"><em>make_2d_plots.py</em> &#8211; Make 2D PCoA Plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_3d_plots.html"><em>make_3d_plots.py</em> &#8211; Make 3D PCoA plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_bipartite_network.html"><em>make_bipartite_network.py</em> &#8211; This script makes a bipartite network connecting samples to observations. It is most suitable for visualization with cytoscape.</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_bootstrapped_tree.html"><em>make_bootstrapped_tree.py</em> &#8211; Make bootstrapped tree</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_distance_boxplots.html"><em>make_distance_boxplots.py</em> &#8211; Creates boxplots to compare distances between categories</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_distance_comparison_plots.html"><em>make_distance_comparison_plots.py</em> &#8211; Creates plots comparing distances between sample groupings</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_distance_histograms.html"><em>make_distance_histograms.py</em> &#8211; Make distance histograms</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_fastq.html"><em>make_fastq.py</em> &#8211; Make FASTQ file for ERA submission from paired FASTA and QUAL files</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_library_id_lists.html"><em>make_library_id_lists.py</em> &#8211; Make library id lists</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_otu_heatmap.html"><em>make_otu_heatmap.py</em> &#8211; Make heatmap of OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_otu_heatmap_html.html"><em>make_otu_heatmap_html.py</em> &#8211; Make heatmap of OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_otu_network.html"><em>make_otu_network.py</em> &#8211; Make an OTU network and calculate statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_otu_table.html"><em>make_otu_table.py</em> &#8211; Make OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_per_library_sff.html"><em>make_per_library_sff.py</em> &#8211; Make per-library sff files from ID lists</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_phylogeny.html"><em>make_phylogeny.py</em> &#8211; Make Phylogeny</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_prefs_file.html"><em>make_prefs_file.py</em> &#8211; Generate preferences file</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_qiime_py_file.html"><em>make_qiime_py_file.py</em> &#8211; Create python file</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_rarefaction_plots.html"><em>make_rarefaction_plots.py</em> &#8211; Generate Rarefaction Plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_tep.html"><em>make_tep.py</em> &#8211; Makes TopiaryExplorer project file</a></li>
<li class="toctree-l1"><a class="reference internal" href="map_reads_to_reference.html"><em>map_reads_to_reference.py</em> &#8211;  Script for performing assignment of reads against a reference database</a></li>
<li class="toctree-l1"><a class="reference internal" href="merge_mapping_files.html"><em>merge_mapping_files.py</em> &#8211; Merge mapping files</a></li>
<li class="toctree-l1"><a class="reference internal" href="merge_otu_maps.html"><em>merge_otu_maps.py</em> &#8211; Merge OTU mapping files</a></li>
<li class="toctree-l1"><a class="reference internal" href="merge_otu_tables.html"><em>merge_otu_tables.py</em> &#8211; Merge two or more OTU tables into a single OTU table.</a></li>
<li class="toctree-l1"><a class="reference internal" href="multiple_rarefactions.html"><em>multiple_rarefactions.py</em> &#8211; Perform multiple subsamplings/rarefactions on an otu table</a></li>
<li class="toctree-l1"><a class="reference internal" href="multiple_rarefactions_even_depth.html"><em>multiple_rarefactions_even_depth.py</em> &#8211; Perform multiple rarefactions on a single otu table, at one depth of sequences/sample</a></li>
<li class="toctree-l1"><a class="reference internal" href="neighbor_joining.html"><em>neighbor_joining.py</em> &#8211; Build a neighbor joining tree comparing samples</a></li>
<li class="toctree-l1"><a class="reference internal" href="nmds.html"><em>nmds.py</em> &#8211; Nonmetric Multidimensional Scaling (NMDS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="otu_category_significance.html"><em>otu_category_significance.py</em> &#8211; OTU significance and co-occurence analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_align_seqs_pynast.html"><em>parallel_align_seqs_pynast.py</em> &#8211; Parallel sequence alignment using PyNAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_alpha_diversity.html"><em>parallel_alpha_diversity.py</em> &#8211; Parallel alpha diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_assign_taxonomy_blast.html"><em>parallel_assign_taxonomy_blast.py</em> &#8211; Parallel taxonomy assignment using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_assign_taxonomy_rdp.html"><em>parallel_assign_taxonomy_rdp.py</em> &#8211; Parallel taxonomy assignment using RDP</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_beta_diversity.html"><em>parallel_beta_diversity.py</em> &#8211; Parallel beta diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_blast.html"><em>parallel_blast.py</em> &#8211; Parallel BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_identify_chimeric_seqs.html"><em>parallel_identify_chimeric_seqs.py</em> &#8211; Parallel chimera detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_map_reads_to_reference.html"><em>parallel_map_reads_to_reference.py</em> &#8211;</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_multiple_rarefactions.html"><em>parallel_multiple_rarefactions.py</em> &#8211; Parallel multiple file rarefaction</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_pick_otus_blast.html"><em>parallel_pick_otus_blast.py</em> &#8211; Parallel pick otus using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_pick_otus_trie.html"><em>parallel_pick_otus_trie.py</em> &#8211; Parallel pick otus using a trie</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_pick_otus_uclust_ref.html"><em>parallel_pick_otus_uclust_ref.py</em> &#8211; Parallel pick otus using uclust_ref</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_pick_otus_usearch61_ref.html"><em>parallel_pick_otus_usearch61_ref.py</em> &#8211; Parallel pick otus using usearch_ref</a></li>
<li class="toctree-l1"><a class="reference internal" href="pick_closed_reference_otus.html"><em>pick_closed_reference_otus.py</em> &#8211; Closed-reference OTU picking/Shotgun UniFrac workflow.</a></li>
<li class="toctree-l1"><a class="reference internal" href="pick_de_novo_otus.html"><em>pick_de_novo_otus.py</em> &#8211; A workflow for de novo OTU picking, taxonomy assignment, phylogenetic tree construction, and OTU table construction.</a></li>
<li class="toctree-l1"><a class="reference internal" href="pick_open_reference_otus.html"><em>pick_open_reference_otus.py</em> &#8211;</a></li>
<li class="toctree-l1"><a class="reference internal" href="pick_otus.html"><em>pick_otus.py</em> &#8211; OTU picking</a></li>
<li class="toctree-l1"><a class="reference internal" href="pick_rep_set.html"><em>pick_rep_set.py</em> &#8211; Pick representative set of sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot_rank_abundance_graph.html"><em>plot_rank_abundance_graph.py</em> &#8211; plot rank-abundance curve</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot_semivariogram.html"><em>plot_semivariogram.py</em> &#8211; Fits a model between two distance matrices and plots the result</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot_taxa_summary.html"><em>plot_taxa_summary.py</em> &#8211; Make taxaonomy summary charts based on taxonomy assignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="poller.html"><em>poller.py</em> &#8211; Poller for parallel QIIME scripts.</a></li>
<li class="toctree-l1"><a class="reference internal" href="poller_example.html"><em>poller_example.py</em> &#8211; Create python file</a></li>
<li class="toctree-l1"><a class="reference internal" href="preferences_file.html"><em>preferences_file</em> &#8211; Example of a prefs file</a></li>
<li class="toctree-l1"><a class="reference internal" href="principal_coordinates.html"><em>principal_coordinates.py</em> &#8211; Principal Coordinates Analysis (PCoA)</a></li>
<li class="toctree-l1"><a class="reference internal" href="print_metadata_stats.html"><em>print_metadata_stats.py</em> &#8211; Count the number of samples associated to a category value</a></li>
<li class="toctree-l1"><a class="reference internal" href="print_qiime_config.html"><em>print_qiime_config.py</em> &#8211; Print out the qiime config settings.</a></li>
<li class="toctree-l1"><a class="reference internal" href="process_iseq.html"><em>process_iseq.py</em> &#8211; Given a directory of per-swath qseq files, this script generates a single fastq per lane.</a></li>
<li class="toctree-l1"><a class="reference internal" href="process_qseq.html"><em>process_qseq.py</em> &#8211; Given a directory of per-swath qseq files, this script generates a single fastq per lane.</a></li>
<li class="toctree-l1"><a class="reference internal" href="process_sff.html"><em>process_sff.py</em> &#8211; Convert sff to FASTA and QUAL files</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality_scores_plot.html"><em>quality_scores_plot.py</em> &#8211; Generates histograms of sequence quality scores and number of nucleotides recorded at a particular index</a></li>
<li class="toctree-l1"><a class="reference internal" href="relatedness.html"><em>relatedness.py</em> &#8211; Calculate NRI (net relatedness index) and NTI (nearest taxon index) using the formulas from Phylocom 4.2/3.41 and Webb 2002.</a></li>
<li class="toctree-l1"><a class="reference internal" href="shared_phylotypes.html"><em>shared_phylotypes.py</em> &#8211; Compute shared OTUs between all pairs of samples</a></li>
<li class="toctree-l1"><a class="reference internal" href="simsam.html"><em>simsam.py</em> &#8211; Simulate samples for each sample in an OTU table, using a phylogenetic tree.</a></li>
<li class="toctree-l1"><a class="reference internal" href="single_rarefaction.html"><em>single_rarefaction.py</em> &#8211; Perform rarefaction on an otu table</a></li>
<li class="toctree-l1"><a class="reference internal" href="sort_otu_table.html"><em>sort_otu_table.py</em> &#8211; Script for sorting the sample IDs in an OTU table based on a specified value in a mapping file.</a></li>
<li class="toctree-l1"><a class="reference internal" href="split_fasta_on_sample_ids.html"><em>split_fasta_on_sample_ids.py</em> &#8211; Split a single post-split_libraries.py fasta file into per-sample fasta files.</a></li>
<li class="toctree-l1"><a class="reference internal" href="split_libraries.html"><em>split_libraries.py</em> &#8211; Split libraries according to barcodes specified in mapping file</a></li>
<li class="toctree-l1"><a class="reference internal" href="split_libraries_fastq.html"><em>split_libraries_fastq.py</em> &#8211; This script performs demultiplexing of Fastq sequence data where barcodes and sequences are contained in two separate fastq files (common on Illumina runs).</a></li>
<li class="toctree-l1"><a class="reference internal" href="split_otu_table.html"><em>split_otu_table.py</em> &#8211; Split in a single OTU table into one OTU table per value in a specified field of the mapping file.</a></li>
<li class="toctree-l1"><a class="reference internal" href="split_otu_table_by_taxonomy.html"><em>split_otu_table_by_taxonomy.py</em> &#8211; Script to split a single OTU table into multiple tables based on the taxonomy at some user-specified depth.</a></li>
<li class="toctree-l1"><a class="reference internal" href="start_parallel_jobs.html"><em>start_parallel_jobs.py</em> &#8211; Starts multiple jobs in parallel on multicore or multiprocessor systems.</a></li>
<li class="toctree-l1"><a class="reference internal" href="start_parallel_jobs_sc.html"><em>start_parallel_jobs_sc.py</em> &#8211; Starts parallel jobs on Sun GridEngine queueing systems.</a></li>
<li class="toctree-l1"><a class="reference internal" href="start_parallel_jobs_torque.html"><em>start_parallel_jobs_torque.py</em> &#8211; Starts multiple jobs in parallel on torque/qsub based multiprocessor systems.</a></li>
<li class="toctree-l1"><a class="reference internal" href="submit_to_mgrast.html"><em>submit_to_mgrast.py</em> &#8211; This script submits a FASTA file to MG-RAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="subsample_fasta.html"><em>subsample_fasta.py</em> &#8211; Randomly subsample sequences from a given fasta file</a></li>
<li class="toctree-l1"><a class="reference internal" href="summarize_otu_by_cat.html"><em>summarize_otu_by_cat.py</em> &#8211; Summarize an OTU table by a single column in the mapping file.</a></li>
<li class="toctree-l1"><a class="reference internal" href="summarize_taxa.html"><em>summarize_taxa.py</em> &#8211; Summarize taxa and store results in a new table or appended to an existing mapping file.</a></li>
<li class="toctree-l1"><a class="reference internal" href="summarize_taxa_through_plots.html"><em>summarize_taxa_through_plots.py</em> &#8211; A workflow script for performing taxonomy summaries and plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="supervised_learning.html"><em>supervised_learning.py</em> &#8211; Run supervised classification using OTUs as predictors and a mapping file category as class labels.</a></li>
<li class="toctree-l1"><a class="reference internal" href="transform_coordinate_matrices.html"><em>transform_coordinate_matrices.py</em> &#8211; Transform 2 coordinate matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="tree_compare.html"><em>tree_compare.py</em> &#8211; Compare jackknifed/bootstrapped trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="trflp_file_to_otu_table.html"><em>trflp_file_to_otu_table.py</em> &#8211; Convert TRFLP text file to an OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="trim_sff_primers.html"><em>trim_sff_primers.py</em> &#8211; Trim sff primers</a></li>
<li class="toctree-l1"><a class="reference internal" href="truncate_fasta_qual_files.html"><em>truncate_fasta_qual_files.py</em> &#8211; Generates filtered fasta and quality score files by truncating at the specified base position.</a></li>
<li class="toctree-l1"><a class="reference internal" href="truncate_reverse_primer.html"><em>truncate_reverse_primer.py</em> &#8211; Takes a demultiplexed fasta file, finds a specified reverse primer sequence, and truncates this primer and subsequent sequences following the reverse primer.</a></li>
<li class="toctree-l1"><a class="reference internal" href="unweight_fasta.html"><em>unweight_fasta.py</em> &#8211; Transform fasta files with abundance weighting into unweighted</a></li>
<li class="toctree-l1"><a class="reference internal" href="upgma_cluster.html"><em>upgma_cluster.py</em> &#8211; Build a UPGMA tree comparing samples</a></li>
<li class="toctree-l1"><a class="reference internal" href="validate_demultiplexed_fasta.html"><em>validate_demultiplexed_fasta.py</em> &#8211; Checks a fasta file to verify if it has  been properly demultiplexed, i.e., it is in QIIME compatible format.</a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    

	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
	<li><a href="../index.html">Home</a><br /></li>
	<li><a href="../install/index.html">Install</a><br /></li>
	<li><a href="../documentation/index.html">Documentation</a><br /></li>
	<li><a href="../tutorials/index.html">Tutorials</a><br /></li>
	<li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	<li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.pycogent.org" target="_blank"><img src="../_static/PoweredByPycogent.jpg" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2012, QIIME Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.4.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>