

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pick_otus.py – OTU picking &mdash; Homepage</title>
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.7.0-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="pick-otus"></span><div class="section" id="pick-otus-py-otu-picking">
<span id="index-0"></span><h1><em>pick_otus.py</em> &#8211; OTU picking<a class="headerlink" href="#pick-otus-py-otu-picking" title="Permalink to this headline">¶</a></h1>
<p><strong>Description:</strong></p>
<p>The OTU picking step assigns similar sequences to operational taxonomic units, or OTUs, by clustering sequences based on a user-defined similarity threshold. Sequences which are similar at or above the threshold level are taken to represent the presence of a taxonomic unit (e.g., a genus, when the similarity threshold is set at 0.94) in the sequence collection.</p>
<p>Currently, the following clustering methods have been implemented in QIIME:</p>
<ol class="arabic simple">
<li>cd-hit (Li &amp; Godzik, 2006; Li, Jaroszewski, &amp; Godzik, 2001), which applies a &#8220;longest-sequence-first list removal algorithm&#8221; to cluster sequences.</li>
<li>blast (Altschul, Gish, Miller, Myers, &amp; Lipman, 1990), which compares and clusters each sequence against a reference database of sequences.</li>
<li>Mothur (Schloss et al., 2009), which requires an input file of aligned sequences.  The input file of aligned sequences may be generated from an input file like the one described below by running <a class="reference external" href="./align_seqs.html">align_seqs.py</a>.  For the Mothur method, the clustering algorithm may be specified as nearest-neighbor, furthest-neighbor, or average-neighbor.  The default algorithm is furthest-neighbor.</li>
<li>prefix/suffix [Qiime team, unpublished], which will collapse sequences which are identical in their first and/or last bases (i.e., their prefix and/or suffix). The prefix and suffix lengths are provided by the user and default to 50 each.</li>
<li>Trie [Qiime team, unpublished], which collapsing identical sequences and sequences which are subsequences of other sequences.</li>
<li>uclust (Edgar, RC 2010), creates &#8220;seeds&#8221; of sequences which generate clusters based on percent identity.</li>
<li>uclust_ref (Edgar, RC 2010), as uclust, but takes a reference database to use as seeds.  New clusters can be toggled on or off.</li>
<li>usearch (Edgar, RC 2010, version v5.2.236), creates &#8220;seeds&#8221; of sequences which generate clusters based on percent identity, filters low abundance clusters, performs de novo and reference based chimera detection.</li>
<li>usearch_ref (Edgar, RC 2010, version v5.2.236), as usearch, but takes a reference database to use as seeds.  New clusters can be toggled on or off.</li>
</ol>
<p>Quality filtering pipeline with usearch 5.X is described as usearch_qf &#8220;usearch quality filter&#8221;, described here: <a class="reference external" href="http://qiime.org/tutorials/usearch_quality_filter.html">http://qiime.org/tutorials/usearch_quality_filter.html</a></p>
<ol class="arabic simple" start="8">
<li>usearch61 (Edgar, RC 2010, version v6.1.544), creates &#8220;seeds&#8221; of sequences which generate clusters based on percent identity.</li>
<li>usearch61_ref (Edgar, RC 2010, version v6.1.544), as usearch61, but takes a reference database to use as seeds.  New clusters can be toggled on or off.</li>
</ol>
<p>Chimera checking with usearch 6.X is implemented in <a class="reference external" href="./identify_chimeric_seqs.html">identify_chimeric_seqs.py</a>.  Chimera checking should be done first with usearch 6.X, and the filtered resulting fasta file can then be clustered.</p>
<p>The primary inputs for <a class="reference external" href="./pick_otus.html">pick_otus.py</a> are:</p>
<ol class="arabic simple">
<li>A FASTA file containing sequences to be clustered</li>
<li>An OTU threshold (default is 0.97, roughly corresponding to species-level OTUs);</li>
<li>The method to be applied for clustering sequences into OTUs.</li>
</ol>
<p><a class="reference external" href="./pick_otus.html">pick_otus.py</a> takes a standard fasta file as input.</p>
<p><strong>Usage:</strong> <tt class="file docutils literal"><span class="pre">pick_otus.py</span> <span class="pre">[options]</span></tt></p>
<p><strong>Input Arguments:</strong></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><strong>[REQUIRED]</strong></p>
<dl class="docutils">
<dt>-i, <cite>-</cite>-input_seqs_filepath</dt>
<dd>Path to input sequences file</dd>
</dl>
<p><strong>[OPTIONAL]</strong></p>
<dl class="last docutils">
<dt>-m, <cite>-</cite>-otu_picking_method</dt>
<dd>Method for picking OTUs.  Valid choices are: mothur, trie, uclust_ref, usearch, usearch_ref, blast, usearch61, usearch61_ref, prefix_suffix, cdhit, uclust. The mothur method requires an input file of aligned sequences.  usearch will enable the usearch quality filtering pipeline. [default: uclust]</dd>
<dt>-c, <cite>-</cite>-clustering_algorithm</dt>
<dd>Clustering algorithm for mothur otu picking method.  Valid choices are: furthest, nearest, average. [default: furthest]</dd>
<dt>-M, <cite>-</cite>-max_cdhit_memory</dt>
<dd>Maximum available memory to cd-hit-est (via the program&#8217;s -M option) for cdhit OTU picking method (units of Mbyte) [default: 400]</dd>
<dt>-o, <cite>-</cite>-output_dir</dt>
<dd>Path to store result file [default: ./&lt;OTU_METHOD&gt;_picked_otus/]</dd>
<dt>-r, <cite>-</cite>-refseqs_fp</dt>
<dd>Path to reference sequences to search against when using -m blast, -m uclust_ref, -m usearch_ref, or -m usearch61_ref [default: None]</dd>
<dt>-b, <cite>-</cite>-blast_db</dt>
<dd>Pre-existing database to blast against when using -m blast [default: None]</dd>
<dt><cite>-</cite>-min_aligned_percent</dt>
<dd>Minimum percent of query sequence that can be aligned to consider a hit (BLAST OTU picker only) [default: 0.5]</dd>
<dt>-s, <cite>-</cite>-similarity</dt>
<dd>Sequence similarity threshold (for blast, cdhit, uclust, uclust_ref, usearch, usearch_ref, usearch61, or usearch61_ref) [default: 0.97]</dd>
<dt>-e, <cite>-</cite>-max_e_value</dt>
<dd>Max E-value when clustering with BLAST [default: 1e-10]</dd>
<dt>-q, <cite>-</cite>-trie_reverse_seqs</dt>
<dd>Reverse seqs before picking OTUs with the Trie OTU picker for suffix (rather than prefix) collapsing [default: False]</dd>
<dt>-n, <cite>-</cite>-prefix_prefilter_length</dt>
<dd>Prefilter data so seqs with identical first prefix_prefilter_length are automatically grouped into a single OTU.  This is useful for large sequence collections where OTU picking doesn&#8217;t scale well [default: None; 100 is a good value]</dd>
<dt>-t, <cite>-</cite>-trie_prefilter</dt>
<dd>Prefilter data so seqs which are identical prefixes of a longer seq are automatically grouped into a single OTU; useful for large sequence collections where OTU picking doesn&#8217;t scale well [default: False]</dd>
<dt>-p, <cite>-</cite>-prefix_length</dt>
<dd>Prefix length when using the prefix_suffix otu picker; WARNING: CURRENTLY DIFFERENT FROM prefix_prefilter_length (-n)! [default: 50]</dd>
<dt>-u, <cite>-</cite>-suffix_length</dt>
<dd>Suffix length when using the prefix_suffix otu picker [default: 50]</dd>
<dt>-z, <cite>-</cite>-enable_rev_strand_match</dt>
<dd>Enable reverse strand matching for uclust, uclust_ref, usearch, usearch_ref, usearch61, or usearch61_ref otu picking, will double the amount of memory used. [default: False]</dd>
<dt>-D, <cite>-</cite>-suppress_presort_by_abundance_uclust</dt>
<dd>Suppress presorting of sequences by abundance when picking OTUs with uclust or uclust_ref [default: False]</dd>
<dt>-A, <cite>-</cite>-optimal_uclust</dt>
<dd>Pass the &#8211;optimal flag to uclust for uclust otu picking. [default: False]</dd>
<dt>-E, <cite>-</cite>-exact_uclust</dt>
<dd>Pass the &#8211;exact flag to uclust for uclust otu picking. [default: False]</dd>
<dt>-B, <cite>-</cite>-user_sort</dt>
<dd>Pass the &#8211;user_sort flag to uclust for uclust otu picking. [default: False]</dd>
<dt>-C, <cite>-</cite>-suppress_new_clusters</dt>
<dd>Suppress creation of new clusters using seqs that don&#8217;t match reference when using -m uclust_ref, -m usearch61_ref, or -m usearch_ref [default: False]</dd>
<dt><cite>-</cite>-max_accepts</dt>
<dd>Max_accepts value to uclust, uclust_ref, usearch61, and usearch61_ref.  By default, will use value suggested by method (uclust: 20, usearch61: 1) [default: default]</dd>
<dt><cite>-</cite>-max_rejects</dt>
<dd>Max_rejects value for uclust, uclust_ref, usearch61, and usearch61_ref.  With default settings, will use value recommended by clustering method used (uclust: 500, usearch61: 8 for usearch_fast_cluster option, 32 for reference and smallmem options) [default: default]</dd>
<dt><cite>-</cite>-stepwords</dt>
<dd>Stepwords value to uclust and uclust_ref [default: 20]</dd>
<dt><cite>-</cite>-word_length</dt>
<dd>Word length value for uclust, uclust_ref, and usearch, usearch_ref, usearch61, and usearch61_ref. With default setting, will use the setting recommended by the method (uclust: 12, usearch: 64, usearch61: 8).  int value can be supplied to override this setting. [default: default]</dd>
<dt><cite>-</cite>-uclust_otu_id_prefix</dt>
<dd>OTU identifier prefix (string) for the de novo uclust OTU picker and for new clusters when uclust_ref is used without -C [default: denovo, OTU ids are ascending integers]</dd>
<dt><cite>-</cite>-suppress_uclust_stable_sort</dt>
<dd>Don&#8217;t pass &#8211;stable-sort to uclust [default: False]</dd>
<dt><cite>-</cite>-suppress_uclust_prefilter_exact_match</dt>
<dd>Don&#8217;t collapse exact matches before calling uclust [default: False]</dd>
<dt>-d, <cite>-</cite>-save_uc_files</dt>
<dd>Enable preservation of intermediate uclust (.uc) files that are used to generate clusters via uclust.  Also enables preservation of all intermediate files created by usearch  and usearch61. [default: True]</dd>
<dt>-j, <cite>-</cite>-percent_id_err</dt>
<dd>Percent identity threshold for cluster error detection with usearch. [default: 0.97]</dd>
<dt>-g, <cite>-</cite>-minsize</dt>
<dd>Minimum cluster size for size filtering with usearch. [default: 4]</dd>
<dt>-a, <cite>-</cite>-abundance_skew</dt>
<dd>Abundance skew setting for de novo chimera detection with usearch. [default: 2.0]</dd>
<dt>-f, <cite>-</cite>-db_filepath</dt>
<dd>Reference database of fasta sequences for reference based chimera detection with usearch. [default: None]</dd>
<dt><cite>-</cite>-perc_id_blast</dt>
<dd>Percent ID for mapping OTUs created by usearch back to original sequence IDs [default: 0.97]</dd>
<dt><cite>-</cite>-de_novo_chimera_detection</dt>
<dd>Deprecated:  de novo chimera detection performed by default, pass &#8211;suppress_de_novo_chimera_detection to disable. [default: None]</dd>
<dt>-k, <cite>-</cite>-suppress_de_novo_chimera_detection</dt>
<dd>Suppress de novo chimera detection in usearch. [default: False]</dd>
<dt><cite>-</cite>-reference_chimera_detection</dt>
<dd>Deprecated:  Reference based chimera detection performed by default, pass &#8211;supress_reference_chimera_detection to disable [default: None]</dd>
<dt>-x, <cite>-</cite>-suppress_reference_chimera_detection</dt>
<dd>Suppress reference based chimera detection in usearch. [default: False]</dd>
<dt><cite>-</cite>-cluster_size_filtering</dt>
<dd>Deprecated, cluster size filtering enabled by default, pass &#8211;suppress_cluster_size_filtering to disable.  [default: None]</dd>
<dt>-l, <cite>-</cite>-suppress_cluster_size_filtering</dt>
<dd>Suppress cluster size filtering in usearch.  [default: False]</dd>
<dt><cite>-</cite>-remove_usearch_logs</dt>
<dd>Disable creation of logs when usearch is called.  Up to nine logs are created, depending on filtering steps enabled.  [default: False]</dd>
<dt><cite>-</cite>-derep_fullseq</dt>
<dd>Dereplication of full sequences, instead of subsequences. Faster than the default &#8211;derep_subseqs in usearch. [default: False]</dd>
<dt>-F, <cite>-</cite>-non_chimeras_retention</dt>
<dd>Selects subsets of sequences detected as non-chimeras to retain after de novo and reference based chimera detection.  Options are intersection or union.  union will retain sequences that are flagged as non-chimeric from either filter, while intersection will retain only those sequences that are flagged as non-chimeras from both detection methods. [default: union]</dd>
<dt><cite>-</cite>-minlen</dt>
<dd>Minimum length of sequence allowed for usearch, usearch_ref, usearch61, and usearch61_ref. [default: 64]</dd>
<dt><cite>-</cite>-usearch_fast_cluster</dt>
<dd>Use fast clustering option for usearch or usearch61_ref with new clusters.  &#8211;enable_rev_strand_match can not be enabled with this option, and the only valid option for usearch61_sort_method is &#8216;length&#8217;.  This option uses more memory than the default option for de novo clustering. [default: False]</dd>
<dt><cite>-</cite>-usearch61_sort_method</dt>
<dd>Sorting method for usearch61 and usearch61_ref.  Valid options are abundance, length, or None.  If the &#8211;usearch_fast_cluster option is enabled, the only sorting method allowed in length. [default: abundance]</dd>
<dt><cite>-</cite>-sizeorder</dt>
<dd>Enable size based preference in clustering with usearch61. Requires that &#8211;usearch61_sort_method be abundance. [default: False]</dd>
</dl>
</div>
<p><strong>Output:</strong></p>
<p>The output consists of two files (i.e. seqs_otus.txt and seqs_otus.log). The .txt file is composed of tab-delimited lines, where the first field on each line corresponds to an (arbitrary) cluster identifier, and the remaining fields correspond to sequence identifiers assigned to that cluster. Sequence identifiers correspond to those provided in the input FASTA file.  Usearch (i.e. usearch quality filter) can additionally have log files for each intermediate call to usearch.</p>
<p>Example lines from the resulting .txt file:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="31%" />
<col width="31%" />
<col width="31%" />
</colgroup>
<tbody valign="top">
<tr><td>0</td>
<td>seq1</td>
<td>seq5</td>
<td>&nbsp;</td>
</tr>
<tr><td>1</td>
<td>seq2</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr><td>2</td>
<td>seq3</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr><td>3</td>
<td>seq4</td>
<td>seq6</td>
<td>seq7</td>
</tr>
</tbody>
</table>
<p>This result implies that four clusters were created based on 7 input sequences. The first cluster (cluster id 0) contains two sequences, sequence ids seq1 and seq5; the second cluster (cluster id 1) contains one sequence, sequence id seq2; the third cluster (cluster id 2) contains one sequence, sequence id seq3, and the final cluster (cluster id 3) contains three sequences, sequence ids seq4, seq6, and seq7.</p>
<p>The resulting .log file contains a list of parameters passed to the <a class="reference external" href="./pick_otus.html">pick_otus.py</a> script along with the output location of the resulting .txt file.</p>
<p><strong>Example (uclust method, default):</strong></p>
<p>Using the seqs.fna file generated from <a class="reference external" href="./split_libraries.html">split_libraries.py</a> and outputting the results to the directory &#8220;picked_otus_default/&#8221;, while using default parameters (0.97 sequence similarity, no reverse strand matching):</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -o picked_otus_default</pre>
</div>
<p>To change the percent identity to a lower value, such as 90%, and also enable reverse strand matching, the command would be the following:</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -o picked_otus_90_percent_rev/ -s 0.90 -z</pre>
</div>
<p><strong>Uclust Reference-based OTU picking example:</strong></p>
<p>uclust_ref can be passed via -m to pick OTUs against a reference set where sequences within the similarity threshold to a reference sequence will cluster to an OTU defined by that reference sequence, and sequences outside of the similarity threshold to a reference sequence will form new clusters. OTU identifiers will be set to reference sequence identifiers when sequences cluster to reference sequences, and &#8216;qiime_otu_&lt;integer&gt;&#8217; for new OTUs. Creation of new clusters can be suppressed by passing -C, in which case sequences outside of the similarity threshold to any reference sequence will be listed as failures in the log file, and not included in any OTU.</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -r refseqs.fasta -m uclust_ref --uclust_otu_id_prefix qiime_otu_</pre>
</div>
<p><strong>Example (cdhit method):</strong></p>
<p>Using the seqs.fna file generated from <a class="reference external" href="./split_libraries.html">split_libraries.py</a> and outputting the results to the directory &#8220;cdhit_picked_otus/&#8221;, while using default parameters (0.97 sequence similarity, no prefix filtering):</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -m cdhit -o cdhit_picked_otus/</pre>
</div>
<p>Currently the cd-hit OTU picker allows for users to perform a pre-filtering step, so that highly similar sequences are clustered prior to OTU picking. This works by collapsing sequences which begin with an identical n-base prefix, where n is specified by the -n parameter. A commonly used value here is 100 (e.g., -n 100). So, if using this filter with -n 100, all sequences which are identical in their first 100 bases will be clustered together, and only one representative sequence from each cluster will be passed to cd-hit. This is used to greatly decrease the run-time of cd-hit-based OTU picking when working with very large sequence collections, as shown by the following command:</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -m cdhit -o cdhit_picked_otus_filter/ -n 100</pre>
</div>
<p>Alternatively, if the user would like to collapse identical sequences, or those which are subsequences of other sequences prior to OTU picking, they can use the trie prefiltering (&#8220;-t&#8221;) option as shown by the following command.</p>
<p>Note: It is highly recommended to use one of the prefiltering methods when analyzing large datasets (&gt;100,000 seqs) to reduce run-time.</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -m cdhit -o cdhit_picked_otus_trie_prefilter/ -t</pre>
</div>
<p><strong>BLAST OTU-Picking Example:</strong></p>
<p>OTUs can be picked against a reference database using the BLAST OTU picker. This is useful, for example, when different regions of the SSU RNA have sequenced and a sequence similarity based approach like cd-hit therefore wouldn&#8217;t work. When using the BLAST OTU picking method, the user must supply either a reference set of sequences or a reference database to compare against. The OTU identifiers resulting from this step will be the sequence identifiers in the reference database. This allows for use of a pre-existing tree in downstream analyses, which again is useful in cases where different regions of the 16s gene have been sequenced.</p>
<p>The following command can be used to blast against a reference sequence set, using the default E-value and sequence similarity (0.97) parameters:</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -o blast_picked_otus/ -m blast -r refseqs.fasta</pre>
</div>
<p>If you already have a pre-built BLAST database, you can pass the database prefix as shown by the following command:</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -o blast_picked_otus_prebuilt_db/ -m blast -b refseqs.fasta</pre>
</div>
<p>If the user would like to change the sequence similarity (&#8220;-s&#8221;) and/or the E-value (&#8220;-e&#8221;) for the blast method, they can use the following command:</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -o blast_picked_otus_90_percent/ -m blast -r refseqs.fasta -s 0.90 -e 1e-30</pre>
</div>
<p><strong>Prefix-suffix OTU Picking Example:</strong></p>
<p>OTUs can be picked by collapsing sequences which begin and/or end with identical bases (i.e., identical prefixes or suffixes).  This OTU picker is currently likely to be of limited use on its own, but will be very useful in collapsing very similar sequences in a chained OTU picking strategy that is currently in development. For example, the user will be able to pick OTUs with this method, followed by representative set picking, and then re-pick OTUs on their representative set. This will allow for highly similar sequences to be collapsed, followed by running a slower OTU picker. This ability to chain OTU pickers is not yet supported in QIIME. The following command illustrates how to pick OTUs by collapsing sequences which are identical in their first 50 and last 25 bases:</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -o prefix_suffix_picked_otus/ -m prefix_suffix -p 50 -u 25</pre>
</div>
<p><strong>Mothur OTU Picking Example:</strong></p>
<p>The Mothur program (<a class="reference external" href="http://www.mothur.org/">http://www.mothur.org/</a>) provides three clustering algorithms for OTU formation: furthest-neighbor (complete linkage), average-neighbor (group average), and nearest-neighbor (single linkage). Details on the algorithms may be found on the Mothur website and publications (Schloss et al., 2009). However, the running times of Mothur&#8217;s clustering algorithms scale with the number of sequences squared, so the program may not be feasible for large data sets.</p>
<p>The following command may be used to create OTUs based on a furthest-neighbor algorithm (the default setting) using aligned sequences as input:</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.aligned.fna -o mothur_picked_otus/ -m mothur</pre>
</div>
<p>If you prefer to use a nearest-neighbor algorithm instead, you may specify this with the &#8216;-c&#8217; flag:</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.aligned.fna -o mothur_picked_otus_nn/ -m mothur -c nearest</pre>
</div>
<p>The sequence similarity parameter may also be specified. For example, the following command may be used to create OTUs at the level of 90% similarity:</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.aligned.fna -o mothur_picked_otus_90_percent/ -m mothur -s 0.90</pre>
</div>
<p><strong>usearch :</strong></p>
<p>Usearch (<a class="reference external" href="http://www.drive5.com/usearch/">http://www.drive5.com/usearch/</a>) provides clustering, chimera checking, and quality filtering. The following command specifies a minimum cluster size of 2 to be used during cluster size filtering:</p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -m usearch --word_length 64 --db_filepath refseqs.fasta -o usearch_qf_results/ --minsize 2</pre>
</div>
<p><strong>usearch example where reference-based chimera detection is disabled, and minimum cluster size filter is reduced from default (4) to 2:</strong></p>
<div class="highlight-python"><pre>pick_otus.py -i seqs.fna -m usearch --word_length 64 --suppress_reference_chimera_detection --minsize 2 -o usearch_qf_results_no_ref_chim_detection/</pre>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    

	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
	<li><a href="../index.html">Home</a><br /></li>
	<li><a href="../install/index.html">Install</a><br /></li>
	<li><a href="../documentation/index.html">Documentation</a><br /></li>
	<li><a href="../tutorials/index.html">Tutorials</a><br /></li>
	<li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	<li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.pycogent.org" target="_blank"><img src="../_static/PoweredByPycogent.jpg" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2012, QIIME Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.4.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>