

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>split_libraries.py – Split libraries according to barcodes specified in mapping file &mdash; Homepage</title>
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.7.0-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="split-libraries"></span><div class="section" id="split-libraries-py-split-libraries-according-to-barcodes-specified-in-mapping-file">
<span id="index-0"></span><h1><em>split_libraries.py</em> &#8211; Split libraries according to barcodes specified in mapping file<a class="headerlink" href="#split-libraries-py-split-libraries-according-to-barcodes-specified-in-mapping-file" title="Permalink to this headline">¶</a></h1>
<p><strong>Description:</strong></p>
<p>Since newer sequencing technologies provide many reads per run (e.g. the 454 GS FLX Titanium series can produce 400-600 million base pairs with 400-500 base pair read lengths) researchers are now finding it useful to combine multiple samples into a single 454 run. This multiplexing is achieved through the application of a pyrosequencing-tailored nucleotide barcode design (described in (Parameswaran et al., 2007)). By assigning individual, unique sample specific barcodes, multiple sequencing runs may be performed in parallel and the resulting reads can later be binned according to sample. The script split_libraries.py performs this task, in addition to several quality filtering steps including user defined cut-offs for: sequence lengths; end-trimming; minimum quality score. To summarize, by using the fasta, mapping, and quality files, the program split_libraries.py will parse sequences that meet user defined quality thresholds and then rename each read with the appropriate Sample ID, thus formatting the sequence data for downstream analysis. If a combination of different sequencing technologies are used in any particular study, split_libraries.py can be used to perform the quality-filtering for each library individually and the output may then be combined.</p>
<p>Sequences from samples that are not found in the mapping file (no corresponding barcode) and sequences without the correct primer sequence will be excluded. Additional scripts can be used to exclude sequences that match a given reference sequence (e.g. the human genome; <a class="reference external" href="./exclude_seqs_by_blast.html">exclude_seqs_by_blast.py</a>) and/or sequences that are flagged as chimeras (<a class="reference external" href="./identify_chimeric_seqs.html">identify_chimeric_seqs.py</a>).</p>
<p><strong>Usage:</strong> <tt class="file docutils literal"><span class="pre">split_libraries.py</span> <span class="pre">[options]</span></tt></p>
<p><strong>Input Arguments:</strong></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><strong>[REQUIRED]</strong></p>
<dl class="docutils">
<dt>-m, <cite>-</cite>-map</dt>
<dd>Name of mapping file. NOTE: Must contain a header line indicating SampleID in the first column and BarcodeSequence in the second, LinkerPrimerSequence in the third.</dd>
<dt>-f, <cite>-</cite>-fasta</dt>
<dd>Names of fasta files, comma-delimited</dd>
</dl>
<p><strong>[OPTIONAL]</strong></p>
<dl class="last docutils">
<dt>-q, <cite>-</cite>-qual</dt>
<dd>Names of qual files, comma-delimited [default: None]</dd>
<dt>-r, <cite>-</cite>-remove_unassigned</dt>
<dd>DEPRECATED: pass &#8211;retain_unassigned_reads to keep unassigned reads  [default: None]</dd>
<dt>-l, <cite>-</cite>-min-seq-length</dt>
<dd>Minimum sequence length, in nucleotides [default: 200]</dd>
<dt>-L, <cite>-</cite>-max-seq-length</dt>
<dd>Maximum sequence length, in nucleotides [default: 1000]</dd>
<dt>-t, <cite>-</cite>-trim-seq-length</dt>
<dd>Calculate sequence lengths after trimming primers and barcodes [default: False]</dd>
<dt>-s, <cite>-</cite>-min-qual-score</dt>
<dd>Min average qual score allowed in read [default: 25]</dd>
<dt>-k, <cite>-</cite>-keep-primer</dt>
<dd>Do not remove primer from sequences</dd>
<dt>-B, <cite>-</cite>-keep-barcode</dt>
<dd>Do not remove barcode from sequences</dd>
<dt>-a, <cite>-</cite>-max-ambig</dt>
<dd>Maximum number of ambiguous bases [default: 6]</dd>
<dt>-H, <cite>-</cite>-max-homopolymer</dt>
<dd>Maximum length of homopolymer run [default: 6]</dd>
<dt>-M, <cite>-</cite>-max-primer-mismatch</dt>
<dd>Maximum number of primer mismatches [default: 0]</dd>
<dt>-b, <cite>-</cite>-barcode-type</dt>
<dd>Barcode type, hamming_8, golay_12, variable_length (will disable any barcode correction if variable_length set), or a number representing the length of the barcode, such as -b 4.  [default: golay_12]</dd>
<dt>-o, <cite>-</cite>-dir-prefix</dt>
<dd>Directory prefix for output files [default: .]</dd>
<dt>-e, <cite>-</cite>-max-barcode-errors</dt>
<dd>Maximum number of errors in barcode [default: 1.5]</dd>
<dt>-n, <cite>-</cite>-start-numbering-at</dt>
<dd>Seq id to use for the first sequence [default: 1]</dd>
<dt><cite>-</cite>-retain_unassigned_reads</dt>
<dd>Retain sequences which are Unassigned in the output sequence file[default: False]</dd>
<dt>-c, <cite>-</cite>-disable_bc_correction</dt>
<dd>Disable attempts to find nearest corrected barcode.  Can improve performance. [default: False]</dd>
<dt>-w, <cite>-</cite>-qual_score_window</dt>
<dd>Enable sliding window test of quality scores.  If the average score of a continuous set of w nucleotides falls below the threshold (see -s for default), the sequence is discarded. A good value would be 50. 0 (zero) means no filtering. Must pass a .qual file (see -q parameter) if this functionality is enabled.  Default behavior for this function is to truncate the sequence at the beginning of the poor quality window, and test for minimal length (-l parameter) of the resulting sequence. [default: 0]</dd>
<dt>-g, <cite>-</cite>-discard_bad_windows</dt>
<dd>If the qual_score_window option (-w) is enabled, this will override the default truncation behavior and discard any sequences where a bad window is found.  [default: False]</dd>
<dt>-p, <cite>-</cite>-disable_primers</dt>
<dd>Disable primer usage when demultiplexing.  Should be enabled for unusual circumstances, such as analyzing Sanger sequence data generated with different primers.  [default: False]</dd>
<dt>-z, <cite>-</cite>-reverse_primers</dt>
<dd>Enable removal of the reverse primer and any subsequence sequence from the end of each read.  To enable this, there has to be a &#8220;ReversePrimer&#8221; column in the mapping file. Primers a required to be in IUPAC format and written in the 5&#8217; to  3&#8217; direction.  Valid options are &#8216;disable&#8217;, &#8216;truncate_only&#8217;, and &#8216;truncate_remove&#8217;.  &#8216;truncate_only&#8217; will remove the primer and subsequent sequence data from the output read and will not alter output of sequences where the primer cannot be found. &#8216;truncate_remove&#8217; will flag sequences where the primer cannot be found to not be written and will record the quantity of such failed sequences in the log file. [default: disable]</dd>
<dt><cite>-</cite>-reverse_primer_mismatches</dt>
<dd>Set number of allowed mismatches for reverse primers (option -z). [default: 0]</dd>
<dt>-d, <cite>-</cite>-record_qual_scores</dt>
<dd>Enables recording of quality scores for all sequences that are recorded.  If this option is enabled, a file named seqs_filtered.qual will be created in the output directory, and will contain the same sequence IDs in the seqs.fna file and sequence quality scores matching the bases present in the seqs.fna file. [default: False]</dd>
<dt>-i, <cite>-</cite>-median_length_filtering</dt>
<dd>Disables minimum and maximum sequence length filtering, and instead calculates the median sequence length and filters the sequences based upon the number of median absolute deviations specified by this parameter.  Any sequences with lengths outside the number of deviations will be removed. [default: None]</dd>
<dt>-j, <cite>-</cite>-added_demultiplex_field</dt>
<dd>Use -j to add a field to use in the mapping file as an additional demultiplexing option to the barcode.  All combinations of barcodes and the values in these fields must be unique. The fields must contain values that can be parsed from the fasta labels such as &#8220;plate=R_2008_12_09&#8221;.  In this case, &#8220;plate&#8221; would be the column header and &#8220;R_2008_12_09&#8221; would be the field data (minus quotes) in the mapping file.  To use the run prefix from the fasta label, such as &#8220;&gt;FLP3FBN01ELBSX&#8221;, where &#8220;FLP3FBN01&#8221; is generated from the run ID, use &#8220;-j run_prefix&#8221; and set the run prefix to be used as the data under the column headerr &#8220;run_prefix&#8221;.  [default: None]</dd>
<dt>-x, <cite>-</cite>-truncate_ambi_bases</dt>
<dd>Enable to truncate at the first &#8220;N&#8221; character encountered in the sequences.  This will disable testing for ambiguous bases (-a option) [default: False]</dd>
</dl>
</div>
<p><strong>Output:</strong></p>
<p>Three files are generated by split_libraries.py:</p>
<ol class="arabic simple">
<li>.fna file (e.g. seqs.fna) - This is a FASTA file containing all sequences which meet the user-defined parameters, where each sequence identifier now contains its corresponding sample id from mapping file.</li>
<li>histograms.txt- This contains the counts of sequences with a particular length.</li>
<li>split_library_log.txt - This file contains a summary of the split_libraries.py analysis. Specifically, this file includes information regarding the number of sequences that pass quality control (number of seqs written) and how these are distributed across the different samples which, through the use of bar-coding technology, would have been pooled into a single 454 run. The number of sequences that pass quality control will depend on length restrictions, number of ambiguous bases, max homopolymer runs, barcode check, etc. All of these parameters are summarized in this file. If raw sequences do not meet the specified quality thresholds they will be omitted from downstream analysis. Since we never see a perfect 454 sequencing run, the number of sequences written should always be less than the number of raw sequences. The number of sequences that are retained for analysis will depend on the quality of the 454 run itself in addition to the default data filtering thresholds in the split_libraries.py script. The default parameters (minimum quality score = 25, minimum/maximum length = 200/1000, no ambiguous bases allowed, no mismatches allowed in primer sequence) can be adjusted to meet the user&#8217;s needs.</li>
</ol>
<p><strong>Standard Example:</strong></p>
<p>Using a single 454 run, which contains a single FASTA, QUAL, and mapping file while using default parameters and outputting the data into the Directory &#8220;Split_Library_Output&#8221;:</p>
<div class="highlight-python"><pre>split_libraries.py -m Mapping_File.txt -f 1.TCA.454Reads.fna -q 1.TCA.454Reads.qual -o Split_Library_Output/</pre>
</div>
<p><strong>Multiple FASTA and QUAL Files Example:</strong></p>
<p>For the case where there are multiple FASTA and QUAL files, the user can run the following comma-separated command as long as there are not duplicate barcodes listed in the mapping file:</p>
<div class="highlight-python"><pre>split_libraries.py -m Mapping_File.txt -f 1.TCA.454Reads.fna,2.TCA.454Reads.fna -q 1.TCA.454Reads.qual,2.TCA.454Reads.qual -o Split_Library_Output_comma_separated/</pre>
</div>
<p><strong>Duplicate Barcode Example:</strong></p>
<p>An example of this situation would be a study with 1200 samples. You wish to have 400 samples per run, so you split the analysis into three runs and reuse barcoded primers (you only have 600). After initial analysis you determine a small subset is underrepresented (&lt;500 sequences per samples) and you boost the number of sequences per sample for this subset by running a fourth run. Since the same sample IDs are in more than one run, it is likely that some sequences will be assigned the same unique identifier by split_libraries.py when it is run separately on the four different runs, each with their own barcode file. This will cause a problem in file concatenation of the four different runs into a single large file. To avoid this, you can use the &#8216;-n&#8217; parameter which defines a start index for split_libraries.py. From experience, most FLX runs (when combining both files for a single plate) will have 350,000 to 650,000 sequences. Thus, if Run 1 for split_libraries.py uses &#8216;-n 1000000&#8217;, Run 2 uses &#8216;-n 2000000&#8217;, etc., then you are guaranteed to have unique identifiers after concatenating the results of multiple FLX runs. With newer technologies you will just need to make sure that your start index spacing is greater than the potential number of sequences.</p>
<p>To run split_libraries.py, you will need two or more (depending on the number of times the barcodes were reused) separate mapping files (one for each Run, for example one for Run1 and another one for Run2), then you can run split_libraries.py using the FASTA and mapping file for Run1 and FASTA and mapping file for Run2. Once you have run split libraries on each file independently, you can concatenate (e.g. using the &#8216;cat&#8217; command) the sequence files that were generated by split_libraries.py. You can also concatenate the mapping files, since the barcodes are not necessary for downstream analyses, unless the same sample IDs are found in multiple mapping files.</p>
<p>Run split_libraries.py on Run 1:</p>
<div class="highlight-python"><pre>split_libraries.py -m Mapping_File.txt -f 1.TCA.454Reads.fna -q 1.TCA.454Reads.qual -o Split_Library_Run1_Output/ -n 1000000</pre>
</div>
<p>Run split_libraries.py on Run 2. The resulting FASTA files from Run 1 and Run 2 can then be concatenated using the &#8216;cat&#8217; command (e.g. cat Split_Library_Run1_Output/seqs.fna Split_Library_Run2_Output/seqs.fna &gt; Combined_seqs.fna) and used in downstream analyses.</p>
<div class="highlight-python"><pre>split_libraries.py -m Mapping_File.txt -f 2.TCA.454Reads.fna -q 2.TCA.454Reads.qual -o Split_Library_Run2_Output/ -n 2000000</pre>
</div>
<p><strong>Barcode Decoding Example:</strong></p>
<p>The standard barcode types supported by split_libraries.py are golay (Length: 12 NTs) and hamming (Length: 8 NTs). For situations where the barcodes are of a different length than golay and hamming, the user can define a generic barcode type &#8220;-b&#8221; as an integer, where the integer is the length of the barcode used in the study.</p>
<p>Note: When analyzing large datasets (&gt;100,000 seqs), users may want to use a generic barcode type, even for length 8 and 12 NTs, since the golay and hamming decoding processes can be computationally intensive, which causes the script to run slow. Barcode correction can be disabled with the -c option if desired.</p>
<p>For the case where the 8 base pair barcodes were used, you can use the following command:</p>
<div class="highlight-python"><pre>split_libraries.py -m Mapping_File_8bp_barcodes.txt -f 1.TCA.454Reads.fna  -q 1.TCA.454Reads.qual -o split_Library_output_8bp/ -b 8</pre>
</div>
<p><strong>Linkers and Primers:</strong></p>
<p>The linker and primer sequence (or all the degenerate possibilities) are associated with each barcode from the mapping file. If a barcode cannot be identified, all the possible primers in the mapping file are tested to find a matching sequence. Using truncated forms of the same primer can lead to unexpected results for rare circumstances where the barcode cannot be identified and the sequence following the barcode matches multiple primers.</p>
<p>In many cases, sequence reads are long enough to sequence through the reverse primer and sequencing adapter.  To remove these primers and all following sequences, the -z option can be used.  By default, this option is set to &#8216;disable&#8217;.  If it is set to &#8216;truncate_only&#8217;, split_libraries will trim the primer and any sequence following it if the primer is found.  If the &#8216;truncate_remove&#8217; option is set, split_libraries.py will trim the primer if found, and will not write the sequence if the primer is not found. The allowed mismatches for the reverse primer are set with the &#8211;reverse_primer_mismatches parameter (default 0).  To use reverse primer removal, one must include a &#8216;ReversePrimer&#8217; column in the mapping file, with the reverse primer recorded in the 5&#8217; to 3&#8217; orientation.</p>
<p>Example reverse primer removal, where primers are trimmed if found, and sequence is written unchanged if not found.  Mismatches are increased to 1 from the default 0:</p>
<div class="highlight-python"><pre>split_libraries.py -m Mapping_File_reverse_primer.txt -f 1.TCA.454Reads.fna -q 1.TCA.454Reads.qual -o split_libraries_output_revprimer/ --reverse_primer_mismatches 1 -z truncate_only</pre>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    

	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
	<li><a href="../index.html">Home</a><br /></li>
	<li><a href="../install/index.html">Install</a><br /></li>
	<li><a href="../documentation/index.html">Documentation</a><br /></li>
	<li><a href="../tutorials/index.html">Tutorials</a><br /></li>
	<li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	<li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.pycogent.org" target="_blank"><img src="../_static/PoweredByPycogent.jpg" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2012, QIIME Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.4.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>